---
title: "Data description"
output: html_notebook
---

# Setup

```{r setup}
require(knitr)
knitr::opts_knit$set(root.dir = file.path(getwd(), '..'))
```

```{r}
library(Biostrings)
library(eulerr) # For Euler plots
library(here)
library(phyloseq)
library(tidyverse)
library(scales) # For comma in plot labels
```

```{r}
# Plotting themes
theme_set(theme_bw() +
               theme(
                    axis.text = element_text(size = 12),
                    axis.title.x = ggtext::element_markdown(size = 14,
                                                            face = 'bold'),
                    axis.title.y = ggtext::element_markdown(size = 14,
                                                            face = 'bold'),
                    legend.title = element_text(size = 12,
                                                face = 'bold'),
                    strip.text = element_text(size = 12,
                                              face = 'bold')
               )
)
```

```{r}
study.color <- c('#4e79a7', # Weight Loss
                 '#f28e2b') # Healthy Donor
```

# Read in data

## Metabarcoding
### trnL
```{r}
ps.mb.plant <- 
     here('data',
          'processed',
          'phyloseq',
          'combined',
          'ps_mb_trnL.rds') %>% 
     readRDS()

ps.mb.plant
```

### 12SV5
```{r}
ps.mb.animal <- 
     here('data',
          'processed',
          'phyloseq',
          'combined',
          'ps_mb_12SV5.rds') %>% 
     readRDS()

ps.mb.animal
```

## Metaproteomics

```{r}
# Phyloseq
ps.protein <- 
     here('data',
          'processed',
          'phyloseq',
          'combined',
          'ps_mp.rds') %>% 
     readRDS()

ps.protein
```

```{r}
# Protein detail
proteins.df <- 
     here('data',
          'processed',
          '5FDR_1UP_both datasets.rds') %>% 
     readRDS()
```

## Menu

```{r}
ps.menu <- 
     here('data', 
          'processed', 
          'phyloseq',
          'combined',
          'ps_menu.rds') %>% 
     readRDS()

ps.menu
```

## References

### Metaproteomic database

```{r }
# Read in the first sheet of the file using read_excel function
db <- readxl::read_excel(here('data', 
                              'processed',
                              'DietaryDB_ForDavidLab.xlsx'), 
                         sheet = 1, 
                         n_max = 547) # Exclude annotations at end of file
head(db)
```

### trnL

```{r}
ref.trnL <- 
     here('..',
          'food-dbs',
          'data',
          'processed',
          'dada2-compatible',
          'trnL',
          'trnLGH.fasta') %>% 
     readDNAStringSet()

ref.trnL
```


### 12SV5

```{r}
ref.12SV5 <- 
     here('..',
          'food-dbs',
          'data',
          'processed',
          'dada2-compatible',
          '12SV5',
          '12SV5_taxonomy.fasta') %>% 
     readDNAStringSet()

ref.12SV5
```

# Pre-process

## Metaproteomic reference

```{r }
# Manual updates
# Specify scientific name for tilapia
db$scientific_name[db$scientific_name == 'Tilapia'] <- 
     'Oreochromis niloticus'
# Correct misspelling for mallard duck
db$scientific_name[db$scientific_name == 'Anas platyrhynchose'] <- 
     'Anas platyrhynchos'
# Switch var. to subsp. for durum
db$scientific_name[db$scientific_name == 'Triticum turgidum var. durum'] <-
     'Triticum turgidum subsp. durum'
db$scientific_name[db$scientific_name == 'Triticum dicoccon'] <- 
     'Triticum dicoccum'
# Correct spelling of water buffalo
db$scientific_name[db$scientific_name == 'Bubalis bubalis'] <- 
     'Bubalus bubalis'
# Think something weird about formatting of "x" in Musa x paradisiaca, not being recognized
db$scientific_name[db$common_name == 'plantain'] <- 
     'Musa x paradisiaca'
# Same for Mentha x piperita
db$scientific_name[db$common_name == 'peppermint'] <- 
     'Mentha x piperita'
# Add hybrid designation for grapefruit
db$scientific_name[db$scientific_name == 'Citrus paradisi'] <- 
     'Citrus x paradisi'
# Remove zante currant, which is redundant with V. vinifera
db <- filter(db, common_name != 'zante currant')
db$scientific_name[db$scientific_name == 'Mercenaria mercenaria, others'] <-
     'Mercenaria mercenaria'
db$scientific_name[db$scientific_name == 'Polystichum munitum and others'] <-
     'Polystichum munitum'
# Query resulted in hits for var. capitata only
db$scientific_name[db$scientific_name == 'Brassica oleracea var. capitata f. alba'] <- 
     'Brassica oleracea var. capitata'
# Switch fennel to its synonym used by NCBI taxonomy
db$scientific_name[db$scientific_name == 'Foeniculum vulgare'] <- 
     'Anethum foeniculum'
# Correct name for dragonfruit
db$scientific_name[db$scientific_name == 'Hylocereus undatus'] <- 
     'Selenicereus undatus'
# Update name for nori
db$scientific_name[db$scientific_name == 'Pyropia yezoensis'] <- 
     'Neopyropia yezoensis'
# Subspecies designation for maple (syrup)
db$scientific_name[db$scientific_name == 'Acer nigrum'] <- 
     'Acer saccharum subsp. nigrum'
``` 

## Metabarcoding

### Unidentified reads

#### trnL

```{r}
# What proportion of overall dataset?
MButils::percent_unassigned(ps.mb.plant)
```

```{r}
# Per sample?
trnL.no.id <- 
     MButils::percent_unassigned(ps.mb.plant, by_sample = TRUE) %>% 
     data.frame(pct_no_id = .) %>% 
     rownames_to_column(var = 'sample') %>% 
     separate(sample,
              into = c('subj', 'date'),
              sep = '_') %>% 
     mutate(study = ifelse(subj == '908',
                           yes = 'Healthy Donor',
                           no = 'Weight Loss'))

trnL.no.id
```
```{r}
median(trnL.no.id$pct_no_id, na.rm = TRUE)
```

```{r}
range(trnL.no.id$pct_no_id, na.rm = TRUE)
```

Small enough to safely drop: remove these from object.

```{r}
ps.mb.plant <- subset_taxa(ps.mb.plant, !is.na(phylum))
ps.mb.plant

rm(trnL.no.id)
```

#### 12SV5

```{r}
# What proportion of each sample?
MButils::percent_unassigned(ps.mb.animal)
```

Here, may want to instead ask what we can't assign to the family level.

```{r}
ps.mb.animal@tax_table@.Data %>% 
     data.frame() %>% 
     rownames_to_column(var = 'asv') %>% 
     select(family:species)
```
Nothing: everything has a genus or species assignment. 

### Filter to >=5 counts 

Impose >= 5 count/taxon filter. 

```{r}
# trnL
asvtab.mb.plant <- otu_table(ps.mb.plant)@.Data
sum(asvtab.mb.plant)

asvtab.mb.plant[asvtab.mb.plant < 5] <- 0
sum(asvtab.mb.plant)

# Replace in phyloseq object
otu_table(ps.mb.plant) <- otu_table(asvtab.mb.plant,
                                    taxa_are_rows = FALSE)
```

```{r}
# 12SV5
asvtab.mb.animal <- otu_table(ps.mb.animal)@.Data
sum(asvtab.mb.animal)

asvtab.mb.animal[asvtab.mb.animal < 5] <- 0
sum(asvtab.mb.animal)

# Replace in phyloseq object
otu_table(ps.mb.animal) <- otu_table(asvtab.mb.animal,
                                    taxa_are_rows = FALSE)
```

### Add lag day

```{r echo = FALSE}
# trnL
samdf.mb.plant <- 
     sample_data(ps.mb.plant) %>% 
     as('data.frame') %>% 
     rownames_to_column(var = 'row')

# Label each stool sample with its two preceding days
for (delta in 1:2){
     varname <- paste0('delta', delta)
     samdf.mb.plant <- mutate(samdf.mb.plant, 
                             !!varname := paste(subj, date - delta, 
                                                sep = '_'))
}

# Replace in phyloseq object
sample_data(ps.mb.plant) <- column_to_rownames(samdf.mb.plant, var = 'row')
```

```{r echo = FALSE}
# 12SV5
samdf.mb.animal <- 
     sample_data(ps.mb.animal) %>% 
     as('data.frame') %>% 
     rownames_to_column(var = 'row')

# Label each stool sample with its two preceding days
for (delta in 1:2){
     varname <- paste0('delta', delta)
     samdf.mb.animal <- mutate(samdf.mb.animal, 
                             !!varname := paste(subj, date - delta, 
                                                sep = '_'))
}

# Replace in phyloseq object
sample_data(ps.mb.animal) <- column_to_rownames(samdf.mb.animal, var = 'row')
```

### Rename

```{r}
taxa_names(ps.mb.plant) <- tax_table(ps.mb.plant)[, 'name']
taxa_names(ps.mb.animal) <- tax_table(ps.mb.animal)[, 'name']
```


## Metaproteomic phyloseq

```{r}
# Do not apply this filter for protein level analysis

# Impose >= 5 PSM/taxon filter
asvtab.protein <- otu_table(ps.protein)@.Data
sum(asvtab.protein)

asvtab.protein[asvtab.protein < 5] <- 0
sum(asvtab.protein)

# Replace in phyloseq object
otu_table(ps.protein) <- otu_table(asvtab.protein,
                                   taxa_are_rows = FALSE)
```

```{r}
# Does this remove any taxa? 
ntaxa(ps.protein)

ps.protein <- 
     ps.protein %>% 
     prune_taxa(taxa_sums(.) > 0, .) 

ntaxa(ps.protein)
```

Wowww! Lots more than I thought.  

### Add lag day

```{r echo = FALSE}
# Pull sample data for stool samples
samdf.protein <- 
     sample_data(ps.protein) %>% 
     as('data.frame') %>% 
     rownames_to_column(var = 'row')

# Label each stool sample with its two preceding days
for (delta in 1:2){
     varname <- paste0('delta', delta)
     samdf.protein <- mutate(samdf.protein, 
                             !!varname := paste(subj, date - delta, 
                                                sep = '_'))
}

# Replace in phyloseq object
sample_data(ps.protein) <- column_to_rownames(samdf.protein, var = 'row')
```

### Separate by kingdom

Need to think about how to transform data when doing a combined ordination on plant and animals (if this is even possible)???

For now I'm not certain, so splitting.
```{r}
nsamples(ps.protein)
```

```{r}
ps.protein.plant <- 
     ps.protein %>% 
     subset_taxa(kingdom == 'Viridiplantae') %>% 
     prune_samples(sample_sums(.) > 0, .)

ps.protein.plant
```

```{r}
ps.protein.animal <- 
     ps.protein %>% 
     subset_taxa(kingdom == 'Metazoa') %>% 
     prune_samples(sample_sums(.) > 0, .)
ps.protein.animal
```

```{r}
# What foods are left over?
ps.protein.other <- 
     ps.protein %>% 
     subset_taxa(!(kingdom %in% c('Viridiplantae',
                                  'Metazoa'))) 

ntaxa(ps.protein.other)
```

```{r}
ps.protein.other@tax_table@.Data %>% 
     data.frame()
```

A few fungi, seaweeds (Rhodophyta), spirulina.  Might be fun to do an independent analysis of the prevalence of these taxa. 

## Menu phyloseq

Here, what should I plot? Aggregated menu from 1-2 d before intake (to be consistent w/downstream results?)

Why don't I just start with one day before and see how it goes.

### Subset to lag -1

```{r}
nsamples(ps.menu)
```

```{r}
ps.menu <- 
     ps.menu %>% 
     # Use all possible samples here, rather than filtering for only successes
     # This occurs in either metabarcoding object (full 27), not yet filtered for success
     prune_samples(sample_names(.) %in% sample_data(ps.mb.animal)$delta1,
                   .) %>% 
     prune_taxa(taxa_sums(.) > 0, .)

ps.menu
```

### Separate by kingdom

```{r}
ps.menu.plant <- 
     ps.menu %>% 
     subset_taxa(kingdom == 'Viridiplantae') %>% 
     prune_samples(sample_sums(.) > 0, .)

ps.menu.plant
```

```{r}
ps.menu.animal <- 
     ps.menu %>% 
     subset_taxa(kingdom == 'Metazoa') %>% 
     prune_samples(sample_sums(.) > 0, .)

ps.menu.animal
```

```{r}
# Get leftovers as for metaproteomic dataset above
ps.menu.other <- 
     ps.menu %>% 
     subset_taxa(!(kingdom %in% c('Viridiplantae',
                                  'Metazoa'))) 

ps.menu.other
```

```{r}
# What foods?
ps.menu.other@tax_table@.Data %>% 
     data.frame()
```
Xanthan gum, mix of mushrooms.

# Describe

```{r}
melt.mp <- psmelt(ps.protein)
melt.mb.plant <- psmelt(ps.mb.plant)
melt.mb.animal <- psmelt(ps.mb.animal)
```

## Success-failure

```{r}
# Metabarcoding: trnL
melt.mb.plant %>% 
     group_by(Sample) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     group_by(reads_tot > 0) %>% 
     count()
```

```{r}
# Metabarcoding: 12SV5
melt.mb.animal %>% 
     group_by(Sample) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     group_by(reads_tot > 0) %>% 
     count()
```
```{r}
# Metaproteomic
melt.mp %>% 
     group_by(Sample) %>% 
     summarize(psms_tot = sum(Abundance)) %>% 
     group_by(psms_tot > 0) %>% 
     count()
```
## Counts per sample

### Metaproteomic spectra

```{r}
melt.mp %>% 
     group_by(Sample, study) %>% 
     summarize(psms_tot = sum(Abundance)) %>% 
     filter(psms_tot > 0) %>% 
     ggplot(aes(x = psms_tot, fill = study)) +
     geom_histogram(stat = 'bin',
                    binwidth = 1000,
                    boundary = 0) +
     scale_y_continuous(breaks = seq(2, 10, by = 2),
                        limits = c(0, 10)) +
     scale_x_continuous(labels = comma) +
     scale_fill_manual(values = study.color) +
     labs(x = 'PSMs', y = 'Samples (*n*)') +
     theme(legend.position = 'none',
           panel.grid.minor = element_blank())
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results',
            'manuscript',
            'supplemental',
            'description',
            paste0(plotID, '_PSM histogram.pdf')),
       height = 3, width = 3)
```

```{r}
melt.mp %>% 
     group_by(Sample) %>% 
     summarize(psms_tot = sum(Abundance)) %>% 
     filter(psms_tot > 0) %>% 
     summarize(median(psms_tot),
               range(psms_tot))
```

### Metabarcoding reads

#### trnL

```{r}
melt.mb.plant %>% 
     group_by(Sample, study) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     filter(reads_tot > 0) %>% 
     ggplot(aes(x = reads_tot, fill = study)) +
     geom_histogram(stat = 'bin',
                    binwidth = 5000,
                    boundary = 0) +
     scale_y_continuous(breaks = seq(2, 10, by = 2),
                        limits = c(0, 10)) +
     scale_x_continuous(labels = comma,
                        breaks = seq(0, 90000, 30000)) +
     scale_fill_manual(values = study.color) +
     labs(x = '*trnL* reads', y = 'Samples (*n*)') +
     theme(legend.position = 'none',
           panel.grid.minor = element_blank())
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results',
            'manuscript',
            'supplemental',
            'description',
            paste0(plotID, '_trnL read histogram.pdf')),
       height = 3, width = 3)
```

```{r}
melt.mb.plant %>% 
     group_by(Sample) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     filter(reads_tot > 0) %>% 
     summarize(median(reads_tot),
               range(reads_tot))
```
#### 12SV5

```{r}
melt.mb.animal %>% 
     group_by(Sample, study) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     filter(reads_tot > 0) %>% 
     ggplot(aes(x = reads_tot, fill = study)) +
     geom_histogram(stat = 'bin',
                    binwidth = 5000,
                    boundary = 0) +
     scale_y_continuous(breaks = seq(2, 10, by = 2),
                        limits = c(0, 10)) +
     scale_x_continuous(labels = comma,
                        limits = c(0, 100000),
                        breaks = seq(0, 90000, 30000)) +
     scale_fill_manual(values = study.color) +
     labs(x = '12SV5 reads', y = 'Samples (*n*)') +
     theme(legend.position = 'none',
           panel.grid.minor = element_blank())
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results',
            'manuscript',
            'supplemental',
            'description',
            paste0(plotID, '_12SV5 read histogram.pdf')),
       height = 3, width = 3)
```

```{r}
melt.mb.animal %>% 
     group_by(Sample) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     filter(reads_tot > 0) %>% 
     summarize(median(reads_tot),
               range(reads_tot))
```

## # reference foods

### Metaproteomics

TODO: Should likely go back and more accurately map these names to species based on entries in the "Notes" column

```{r}
db %>% 
     group_by(category) %>% 
     summarize(n_distinct(scientific_name))
```
plant* is Rhodophyta (plant)
NA is hake (animal)

### Metabarcoding

#### 12SV5 

Can just leverage taxonomy here to split apart and reduce
```{r}
# Number of sequences:
length(ref.12SV5)
```

```{r}
ref.12SV5 <- 
     ref.12SV5 %>% 
     names() %>% 
     data.frame(name = .) %>% 
     # Trim terminal semicolon
     mutate(name = gsub(pattern = ';$', 
                        replacement = '',
                        name)) %>% 
     separate(name,
              into = c('kingdom',
                       'phylum',
                       'class',
                       'order',
                       'family',
                       'genus',
                       'species',
                       'subspecies'),
              sep = ';',
              remove = FALSE) %>% 
     MButils::lowest_level()

ref.12SV5.db
```
```{r}
# Number of animals
# Note one of these is human: exclude
n_distinct(ref.12SV5.db$name) - 1
```

#### trnL

```{r}
# Number of sequences:
length(ref.trnL)
```

```{r}
# Number of plants
ref.trnL <- 
     ref.trnL %>% 
     names() %>% 
     data.frame(name = .) %>% 
     separate(name,
              into = c('accession', 
                       'name'),
              sep = ' ',
              extra = 'merge')
```

```{r}
n_distinct(ref.trnL$name)
```

## Detected foods

### Metaproteomics

```{r}
ntaxa(ps.protein)
```

What about number of unique dietary peptides?  Use intermediate processing stage of data for this (saved in "Metaproteomic phyloseq" notebook):

```{r}
str(proteins.df)
```

```{r}
dim(proteins.df)
```

```{r}
n_distinct(proteins.df$Accession)
n_distinct(proteins.df$Description)
```

```{r}
# Most prevalent foods
melt.mp %>% 
     group_by(Sample, OTU) %>% 
     summarize(prevalence_within = sum(Abundance > 5)) %>% 
     # Now group by just food, get across-sample prevalence
     group_by(OTU) %>% 
     summarize(prevalence_across = sum(prevalence_within)) %>% 
     arrange(desc(prevalence_across))
```

```{r}
# Most abundant foods
melt.mp %>% 
     filter(Abundance > 5) %>% 
     group_by(OTU) %>% 
     summarize(abundance = sum(Abundance)) %>% 
     arrange(desc(abundance))
```

### Metabarcoding

```{r}
# trnL
ps.mb.plant
```

```{r}
# Most prevalent foods
melt.mb.plant %>% 
     group_by(Sample, name) %>% 
     summarize(prevalence_within = sum(Abundance > 0)) %>% 
     # Now group by just food, get across-sample prevalence
     group_by(name) %>% 
     summarize(prevalence_across = sum(prevalence_within)) %>% 
     arrange(desc(prevalence_across))
```

```{r}
# Most abundant foods
melt.mb.plant %>% 
     group_by(name) %>% 
     summarize(abundance = sum(Abundance)) %>% 
     arrange(desc(abundance))
```

```{r}
# 12SV5
ps.mb.animal
```

```{r}
# Most prevalent foods
melt.mb.animal %>% 
     group_by(Sample, name) %>% 
     summarize(prevalence_within = sum(Abundance > 0)) %>% 
     # Now group by just food, get across-sample prevalence
     group_by(name) %>% 
     summarize(prevalence_across = sum(prevalence_within)) %>% 
     arrange(desc(prevalence_across))
```

### Intersection

What species are shared between metaproteomic, metabarcoding, and (day before sample) menu datasets?

#### Filter to only successful samples in all 3

```{r}
ps.mb.plant
ps.mb.plant <- 
     ps.mb.plant %>% 
     prune_samples(sample_sums(.) > 0, .) %>% 
     prune_taxa(taxa_sums(.) > 0, .)

ps.mb.plant
```

No similar pruning necessary for animal metabarcoding (all successful) or metaproteomic dataset (failed sample falls off when split kingdom-wise). 

```{r}
samples.plant <- 
     intersect(sample_names(ps.mb.plant),
               sample_names(ps.protein.plant))

length(samples.plant)
```

```{r}
samples.animal <- 
     intersect(sample_names(ps.mb.animal),
               sample_names(ps.protein.animal))

length(samples.animal)
```

```{r}
subset_ps <- function(ps, samples){
     ps %>% 
          prune_samples(samples, .) %>% 
          prune_taxa(taxa_sums(.) > 0, .)
}

# Now, subset each dataset (except menu-- needs to be referenced w/ delta)
ps.mb.plant <- subset_ps(ps.mb.plant, samples.plant)
ps.protein.plant <- subset_ps(ps.protein.plant, samples.plant)

ps.mb.animal <- subset_ps(ps.mb.animal, samples.animal)
ps.protein.animal <- subset_ps(ps.protein.animal, samples.animal)
```

```{r}
# Now, can subset menus
ps.menu.plant <- 
     ps.menu.plant %>% 
     prune_samples(sample_names(.) %in% 
                        c(sample_data(ps.mb.plant)$delta1,
                          sample_data(ps.mb.plant)$delta2),
                   .) %>% 
     prune_taxa(taxa_sums(.) > 0, .)

ps.menu.animal <- 
     ps.menu.animal %>% 
     prune_samples(sample_names(.) %in% 
                        c(sample_data(ps.mb.animal)$delta1,
                          sample_data(ps.mb.animal)$delta2),
                   .) %>% 
     prune_taxa(taxa_sums(.) > 0, .)
```


```{r}
# Plants
# Get union of taxa detected in the two analyses
all.plants <- 
     union(taxa_names(ps.mb.plant),
           taxa_names(ps.protein.plant)) %>% 
     union(taxa_names(ps.menu.plant)) 

length(all.plants)
```

```{r}
# Animals
all.animals <- 
     union(taxa_names(ps.mb.animal),
           taxa_names(ps.protein.animal)) %>% 
     union(taxa_names(ps.menu.animal)) 

length(all.animals)
```

```{r}
# Make dataframe for plot input
# Plants
plant.df <- 
     data.frame(taxon = all.plants) %>% 
     mutate(dna = taxon %in% taxa_names(ps.mb.plant),
            protein = taxon %in% taxa_names(ps.protein.plant),
            menu = taxon %in% taxa_names(ps.menu.plant))

# Quick checks
sum(plant.df$dna) # Should be 81
sum(plant.df$protein) # Should be 75
sum(plant.df$menu) # Should be 111 (123 for 2 days)
```

```{r}
# Make dataframe for plot input
# Animals
animal.df <- 
     data.frame(taxon = all.animals) %>% 
     mutate(dna = taxon %in% taxa_names(ps.mb.animal),
            protein = taxon %in% taxa_names(ps.protein.animal),
            menu = taxon %in% taxa_names(ps.menu.animal))

# Quick checks
sum(animal.df$dna) # Should be 11
sum(animal.df$protein) # Should be 28
sum(animal.df$menu) # Should be 19 (20 for 2 days)
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
pdf(here('results',
         'manuscript',
         'tbd',
         paste0(plotID, 
                '_Plant taxa Euler diagram across methods, with counts, 1 day prior menu.pdf')),
    height = 4, width = 4)

euler(plant.df[,c('dna', 'protein', 'menu')]) %>% 
     plot(
          fills = c('#fbb04e', '#849db1', '#7e756d'),
          labels = TRUE,
          shape = "ellipse", 
          quantities = TRUE
          )
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
pdf(here('results',
         'manuscript',
         'tbd',
         paste0(plotID, '_Animal taxa Euler diagram across methods, with counts, 1 day prior menu.pdf')),
    height = 4, width = 4)

euler(animal.df[,c('dna', 'protein', 'menu')]) %>% 
     plot(
          fills = c('#fbb04e', '#849db1', '#7e756d'),
          labels = TRUE,
          shape = "ellipse", 
          quantities = TRUE
          )
```

### Differences by dataset

#### Plant

```{r}
# In all
filter(plant.df,
       dna & protein & menu) %>% 
     pull(taxon) %>% 
     sort()
```

```{r}
# In DNA and protein, missing from menu
filter(plant.df,
       dna & protein & !menu) %>% 
     pull(taxon) %>% 
     sort()
```

Tea, chickpea, pistachio, pomegranate, sorghum, mung bean.

Confirmed each of these against menu.  Notes:
* No record of pistachio

```{r}
# In DNA and menu, missing from protein
filter(plant.df,
       dna & !protein & menu) %>% 
     pull(taxon) %>% 
     sort()
```

## Per-sample details

Overall view:
```{r echo = FALSE}
# How many food taxa consumed daily (menu)?
melt.menu <- psmelt(ps.menu)
counts.menu <- 
     melt.menu %>% 
     filter(Abundance > 0) %>% 
     group_by(subj, date, kingdom) %>% 
     count(name = 'menu_taxa')
```

```{r echo = FALSE}
# How many food taxa detected daily (metaproteomics)?
melt.protein <- psmelt(ps.protein)
counts.protein <- 
     melt.protein %>% 
     filter(Abundance > 4) %>% 
     group_by(subj, date, kingdom) %>% 
     count(name = 'protein_taxa')
```

```{r}
# Note some NA values here: what are they?
melt.protein %>% 
     filter(Abundance > 4) %>% 
     filter(is.na(kingdom)) %>% 
     pull(OTU) %>% 
     unique()
```
If one defines the kingdom Plantae to mean the Archaeplastida, the red algae will be part of that kingdom.
If Plantae are defined more narrowly, to be the Viridiplantae, then the red algae might be considered their own kingdom, or part of the kingdom Protista.

```{r echo = FALSE}
# How many food taxa detected daily (metabarcoding)?
melt.mb <- psmelt(
     merge_phyloseq(ps.mb.plant,
                    ps.mb.animal)
)
counts.mb <- 
     melt.mb %>% 
     filter(Abundance > 4) %>% 
     group_by(subj, date, phylum) %>% 
     count(name = 'dna_taxa')
```

```{r}
# Do this for only samples in common
# Though need to preserve Fungi, which aren't measured by metabarcoding
# This can be done by doing a full join
counts <- 
     counts.protein %>% 
     full_join(counts.mb) %>% 
     filter(!is.na(kingdom)) %>% 
     mutate(delta1 = date - 1) %>% 
     # Join to menu, lagged 1 day 
     # Full join to add kingdom if necessary
     full_join(counts.menu, by = c('kingdom',
                                   'subj', 
                                   'delta1' = 'date')) %>% 
     # Then remove elements that didn't join by the date
     filter(!is.na(date))

# Long version for plotting
counts.long <- pivot_longer(counts, 
                            cols = ends_with('taxa'),
                            names_to = 'dataset',
                            values_to = 'value')
```

```{r}
# Update names for plotting
counts.long$dataset <- factor(counts.long$dataset, 
                                     levels = c('menu_taxa', 
                                                'protein_taxa',
                                                'dna_taxa'), 
                                     labels = c('Recorded menu',
                                                'Metaproteomics',
                                                'Metabarcoding'))
# Refactor for plotting
counts.long$kingdom <- factor(counts.long$kingdom,
                              levels = c('Viridiplantae',
                                         'Metazoa', 
                                         'Fungi'))
```

```{r}
ggplot(counts.long, aes(x = kingdom, y = value, fill = dataset)) +
     geom_boxplot() +
     labs(title = 'Number of food taxa identified by measure',
          subtitle = '>4 PSMs per taxon filter',
          x = 'Kingdom',
          y = 'Number of taxa', fill = 'Measure') +
     ylim(0, 65) + # Leave room to add significance
     # scale_y_log10() +
     scale_fill_manual(values = c('#d7ce9f', '#849db1', '#b66353')) +
     theme_classic() +
     theme()
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
ggsave(
     here(
          'results',
          'QC',
          paste0(plotID,
                 '_Number of food taxa by measure, shared samples only.pdf'
                 )
          ),
     height = 3,
     width = 6)
```

Statistics on above plot: 

```{r}
# Repeated measures ANOVA (Viridiplantae)
# Get only plant detection events, make unique sample ID
viridiplantae <- 
     counts.long %>% 
     mutate(sample = paste(subj, date, sep = '_')) %>% 
     filter(kingdom == 'Viridiplantae') 

model <- lm(value ~ dataset + sample,
            data = viridiplantae)

analysis <- Anova(model, 
                  idata = viridiplantae,
                  idesign = ~sample)

print(analysis)
```

```{r}
# Post-hoc testing

# Unadjusted p values
pairwise.t.test(viridiplantae$value,
                viridiplantae$dataset,
                p.adjust.method = 'BH')
```

```{r}
# Repeated measures ANOVA (Metazoa)
metazoa <- 
     counts.long %>% 
     mutate(sample = paste(subj, date, sep = '_')) %>% 
     filter(kingdom == 'Metazoa') 

model <- lm(value ~ dataset + sample,
            data = metazoa)

analysis <- Anova(model, 
                  idata = metazoa,
                  idesign = ~sample)

print(analysis)
```

```{r}
# Unadjusted p values
pairwise.t.test(metazoa$value,
                metazoa$dataset,
                p.adjust.method = 'BH')
```


```{r}
# Paired t-test (Fungi)
fungi <- 
     filter(counts, kingdom == 'Fungi')

t.test(fungi$menu_taxa, fungi$protein_taxa,
       paired = TRUE)
```

Patterns of recorded foods by day of week: 

```{r echo = FALSE}
counts.long %>% 
     group_by(ID, date, dataset) %>% 
     summarize(value = sum(value, na.rm = TRUE)) %>% 
     filter(dataset == 'Recorded menu') %>% 
     ggplot(aes(x = date, y = value)) +
     geom_point(aes(color = dataset)) +
     facet_wrap(~ ID, nrow = 4, scales = 'free_x') + 
     labs(x = 'Date', y = 'Number of taxa') +
     theme_bw() +
     theme(axis.title.x = element_blank(),
           legend.title = element_blank(),
           legend.position = 'bottom')

plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
# ggsave(here('results', 
#             paste0(plotID, 
#                    '_Number of food taxa in menu by day.pdf')),
#        height = 7, width = 4)
```

