---
title: "Data description"
output: html_notebook
---

# Setup

```{r setup}
require(knitr)
knitr::opts_knit$set(root.dir = file.path(getwd(), '..'))
```

```{r}
library(Biostrings)
library(here)
library(phyloseq)
library(tidyverse)
library(scales) # For comma in plot labels
```

```{r}
# Plotting themes
theme_set(theme_bw() +
               theme(
                    axis.text = element_text(size = 12),
                    axis.title.x = ggtext::element_markdown(size = 14,
                                                            face = 'bold'),
                    axis.title.y = ggtext::element_markdown(size = 14,
                                                            face = 'bold'),
                    legend.title = element_text(size = 12,
                                                face = 'bold'),
                    strip.text = element_text(size = 12,
                                              face = 'bold')
               )
)
```

```{r}
study.color <- c('#4e79a7', # Weight Loss
                 '#f28e2b') # Healthy Donor
```

# Read in data

## Metabarcoding
### trnL
```{r}
ps.trnL <- 
     here('data',
          'processed',
          'phyloseq',
          'combined',
          'ps_mb_trnL.rds') %>% 
     readRDS()

ps.trnL
```

### 12SV5
```{r}
ps.12SV5 <- 
     here('data',
          'processed',
          'phyloseq',
          'combined',
          'ps_mb_12SV5.rds') %>% 
     readRDS()

ps.12SV5
```

## Metaproteomics

```{r}
# Phyloseq
ps.mp <- 
     here('data',
          'processed',
          'phyloseq',
          'combined',
          'ps_mp.rds') %>% 
     readRDS()

ps.mp
```

```{r}
# Protein detail
proteins.df <- 
     here('data',
          'processed',
          '5FDR_1UP_both datasets.rds') %>% 
     readRDS()
```

## References

### Metaproteomic database

```{r }
# Read in the first sheet of the file using read_excel function
db <- readxl::read_excel(here('data', 
                              'processed',
                              'DietaryDB_ForDavidLab.xlsx'), 
                         sheet = 1, 
                         n_max = 547) # Exclude annotations at end of file
head(db)
```

### trnL

```{r}
ref.trnL <- 
     here('..',
          'food-dbs',
          'data',
          'processed',
          'dada2-compatible',
          'trnL',
          'trnLGH.fasta') %>% 
     readDNAStringSet()

ref.trnL
```


### 12SV5

```{r}
ref.12SV5 <- 
     here('..',
          'food-dbs',
          'data',
          'processed',
          'dada2-compatible',
          '12SV5',
          '12SV5_taxonomy.fasta') %>% 
     readDNAStringSet()

ref.12SV5
```

# Pre-process

## Metaproteomic reference

## Reference

```{r }
# Manual updates
# Specify scientific name for tilapia
db$scientific_name[db$scientific_name == 'Tilapia'] <- 
     'Oreochromis niloticus'
# Correct misspelling for mallard duck
db$scientific_name[db$scientific_name == 'Anas platyrhynchose'] <- 
     'Anas platyrhynchos'
# Switch var. to subsp. for durum
db$scientific_name[db$scientific_name == 'Triticum turgidum var. durum'] <-
     'Triticum turgidum subsp. durum'
db$scientific_name[db$scientific_name == 'Triticum dicoccon'] <- 
     'Triticum dicoccum'
# Correct spelling of water buffalo
db$scientific_name[db$scientific_name == 'Bubalis bubalis'] <- 
     'Bubalus bubalis'
# Think something weird about formatting of "x" in Musa x paradisiaca, not being recognized
db$scientific_name[db$common_name == 'plantain'] <- 
     'Musa x paradisiaca'
# Same for Mentha x piperita
db$scientific_name[db$common_name == 'peppermint'] <- 
     'Mentha x piperita'
# Add hybrid designation for grapefruit
db$scientific_name[db$scientific_name == 'Citrus paradisi'] <- 
     'Citrus x paradisi'
# Remove zante currant, which is redundant with V. vinifera
db <- filter(db, common_name != 'zante currant')
db$scientific_name[db$scientific_name == 'Mercenaria mercenaria, others'] <-
     'Mercenaria mercenaria'
db$scientific_name[db$scientific_name == 'Polystichum munitum and others'] <-
     'Polystichum munitum'
# Query resulted in hits for var. capitata only
db$scientific_name[db$scientific_name == 'Brassica oleracea var. capitata f. alba'] <- 
     'Brassica oleracea var. capitata'
# Switch fennel to its synonym used by NCBI taxonomy
db$scientific_name[db$scientific_name == 'Foeniculum vulgare'] <- 
     'Anethum foeniculum'
# Correct name for dragonfruit
db$scientific_name[db$scientific_name == 'Hylocereus undatus'] <- 
     'Selenicereus undatus'
# Update name for nori
db$scientific_name[db$scientific_name == 'Pyropia yezoensis'] <- 
     'Neopyropia yezoensis'
# Subspecies designation for maple (syrup)
db$scientific_name[db$scientific_name == 'Acer nigrum'] <- 
     'Acer saccharum subsp. nigrum'
``` 

## Metabarcoding

### Unidentified reads

```{r}
# What proportion of overall dataset?
MButils::percent_unassigned(ps.trnL)
```

```{r}
# Per sample?
trnL.no.id <- 
     MButils::percent_unassigned(ps.trnL, by_sample = TRUE) %>% 
     data.frame(pct_no_id = .) %>% 
     rownames_to_column(var = 'sample') %>% 
     separate(sample,
              into = c('subj', 'date'),
              sep = '_') %>% 
     mutate(study = ifelse(subj == '908',
                           yes = 'Healthy Donor',
                           no = 'Weight Loss'))

trnL.no.id
```
```{r}
median(trnL.no.id$pct_no_id, na.rm = TRUE)
```

```{r}
range(trnL.no.id$pct_no_id, na.rm = TRUE)
```

Small enough to safely drop: remove these from object.

```{r}
ps.trnL <- subset_taxa(ps.trnL, !is.na(phylum))
ps.trnL

rm(trnL.no.id)
```

### 12SV5

```{r}
# What proportion of each sample?
MButils::percent_unassigned(ps.12SV5)
```

Here, may want to instead ask what we can't assign to the family level.

```{r}
ps.12SV5@tax_table@.Data %>% 
     data.frame() %>% 
     rownames_to_column(var = 'asv') %>% 
     select(family:species)
```
Nothing: everything has a genus or species assignment. 

# Describe

```{r}
melt.mp <- psmelt(ps.mp)
melt.trnL <- psmelt(ps.trnL)
melt.12SV5 <- psmelt(ps.12SV5)
```

## Success-failure

```{r}
# Metabarcoding: trnL
melt.trnL %>% 
     group_by(Sample) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     group_by(reads_tot > 0) %>% 
     count()
```

```{r}
# Metabarcoding: 12SV5
melt.12SV5 %>% 
     group_by(Sample) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     group_by(reads_tot > 0) %>% 
     count()
```
```{r}
# Metaproteomic
melt.mp %>% 
     group_by(Sample) %>% 
     summarize(psms_tot = sum(Abundance)) %>% 
     group_by(psms_tot > 0) %>% 
     count()
```
## Counts per sample

### Metaproteomic spectra

```{r}
melt.mp %>% 
     group_by(Sample, study) %>% 
     summarize(psms_tot = sum(Abundance)) %>% 
     filter(psms_tot > 0) %>% 
     ggplot(aes(x = psms_tot, fill = study)) +
     geom_histogram(stat = 'bin',
                    binwidth = 1000,
                    boundary = 0) +
     scale_y_continuous(breaks = seq(2, 10, by = 2),
                        limits = c(0, 10)) +
     scale_x_continuous(labels = comma) +
     scale_fill_manual(values = study.color) +
     labs(x = 'PSMs', y = 'Samples (*n*)') +
     theme(legend.position = 'none',
           panel.grid.minor = element_blank())
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results',
            'manuscript',
            'supplemental',
            'description',
            paste0(plotID, '_PSM histogram.pdf')),
       height = 3, width = 3)
```

```{r}
melt.mp %>% 
     group_by(Sample) %>% 
     summarize(psms_tot = sum(Abundance)) %>% 
     filter(psms_tot > 0) %>% 
     summarize(median(psms_tot),
               range(psms_tot))
```

### Metabarcoding reads

#### trnL

```{r}
melt.trnL %>% 
     group_by(Sample, study) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     filter(reads_tot > 0) %>% 
     ggplot(aes(x = reads_tot, fill = study)) +
     geom_histogram(stat = 'bin',
                    binwidth = 5000,
                    boundary = 0) +
     scale_y_continuous(breaks = seq(2, 10, by = 2),
                        limits = c(0, 10)) +
     scale_x_continuous(labels = comma,
                        breaks = seq(0, 90000, 30000)) +
     scale_fill_manual(values = study.color) +
     labs(x = '*trnL* reads', y = 'Samples (*n*)') +
     theme(legend.position = 'none',
           panel.grid.minor = element_blank())
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results',
            'manuscript',
            'supplemental',
            'description',
            paste0(plotID, '_trnL read histogram.pdf')),
       height = 3, width = 3)
```

```{r}
melt.trnL %>% 
     group_by(Sample) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     filter(reads_tot > 0) %>% 
     summarize(median(reads_tot),
               range(reads_tot))
```
#### 12SV5

```{r}
melt.12SV5 %>% 
     group_by(Sample, study) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     filter(reads_tot > 0) %>% 
     ggplot(aes(x = reads_tot, fill = study)) +
     geom_histogram(stat = 'bin',
                    binwidth = 5000,
                    boundary = 0) +
     scale_y_continuous(breaks = seq(2, 10, by = 2),
                        limits = c(0, 10)) +
     scale_x_continuous(labels = comma,
                        limits = c(0, 100000),
                        breaks = seq(0, 90000, 30000)) +
     scale_fill_manual(values = study.color) +
     labs(x = '12SV5 reads', y = 'Samples (*n*)') +
     theme(legend.position = 'none',
           panel.grid.minor = element_blank())
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results',
            'manuscript',
            'supplemental',
            'description',
            paste0(plotID, '_12SV5 read histogram.pdf')),
       height = 3, width = 3)
```

```{r}
melt.12SV5 %>% 
     group_by(Sample) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     filter(reads_tot > 0) %>% 
     summarize(median(reads_tot),
               range(reads_tot))
```

## # reference foods

### Metaproteomics

TODO: Should likely go back and more accurately map these names to species based on entries in the "Notes" column

```{r}
db %>% 
     group_by(category) %>% 
     summarize(n_distinct(scientific_name))
```
plant* is Rhodophyta (plant)
NA is hake (animal)

### Metabarcoding

#### 12SV5 

Can just leverage taxonomy here to split apart and reduce
```{r}
# Number of sequences:
length(ref.12SV5)
```

```{r}
ref.12SV5 <- 
     ref.12SV5 %>% 
     names() %>% 
     data.frame(name = .) %>% 
     # Trim terminal semicolon
     mutate(name = gsub(pattern = ';$', 
                        replacement = '',
                        name)) %>% 
     separate(name,
              into = c('kingdom',
                       'phylum',
                       'class',
                       'order',
                       'family',
                       'genus',
                       'species',
                       'subspecies'),
              sep = ';',
              remove = FALSE) %>% 
     MButils::lowest_level()

ref.12SV5.db
```
```{r}
# Number of animals
# Note one of these is human: exclude
n_distinct(ref.12SV5.db$name) - 1
```

#### trnL

```{r}
# Number of sequences:
length(ref.trnL)
```

```{r}
# Number of plants
ref.trnL <- 
     ref.trnL %>% 
     names() %>% 
     data.frame(name = .) %>% 
     separate(name,
              into = c('accession', 
                       'name'),
              sep = ' ',
              extra = 'merge')
```

```{r}
n_distinct(ref.trnL$name)
```

## Detected foods

### Metaproteomics

```{r}
ntaxa(ps.mp)
```

What about number of unique dietary peptides?  Use intermediate processing stage of data for this (saved in "Metaproteomic phyloseq" notebook):

```{r}
str(proteins.df)
```

```{r}
dim(proteins.df)
```

```{r}
n_distinct(proteins.df$Accession)
n_distinct(proteins.df$Description)
```

```{r}
# Most prevalent foods
melt.mp %>% 
     group_by(Sample, OTU) %>% 
     summarize(prevalence_within = sum(Abundance > 5)) %>% 
     # Now group by just food, get across-sample prevalence
     group_by(OTU) %>% 
     summarize(prevalence_across = sum(prevalence_within)) %>% 
     arrange(desc(prevalence_across))
```

```{r}
# Most abundant foods
melt.mp %>% 
     filter(Abundance > 5) %>% 
     group_by(OTU) %>% 
     summarize(abundance = sum(Abundance)) %>% 
     arrange(desc(abundance))
```
### Metabarcoding

```{r}
# trnL
ps.trnL
```

```{r}
# Most prevalent foods
melt.trnL %>% 
     group_by(Sample, name) %>% 
     summarize(prevalence_within = sum(Abundance > 0)) %>% 
     # Now group by just food, get across-sample prevalence
     group_by(name) %>% 
     summarize(prevalence_across = sum(prevalence_within)) %>% 
     arrange(desc(prevalence_across))
```

```{r}
# Most abundant foods
melt.trnL %>% 
     group_by(name) %>% 
     summarize(abundance = sum(Abundance)) %>% 
     arrange(desc(abundance))
```
```{r}
# 12SV5
ps.12SV5
```

```{r}
# Most prevalent foods
melt.12SV5 %>% 
     group_by(Sample, name) %>% 
     summarize(prevalence_within = sum(Abundance > 0)) %>% 
     # Now group by just food, get across-sample prevalence
     group_by(name) %>% 
     summarize(prevalence_across = sum(prevalence_within)) %>% 
     arrange(desc(prevalence_across))
```

## Detected taxa

```{r}
asvtab.protein <- otu_table(ps.protein)@.Data
asvtab.mb <- otu_table(ps.mb)@.Data
asvtab.menu <- otu_table(ps.menu)@.Data
```

What species are shared between metaproteomic, metabarcoding, and menu datasets?

### Euler plot

Note that for menu data, perhaps best not to include the full range as this might include taxa we'd never expect to see in samples.  

```{r}
# Get only those taxa in Mon-Thu menu
taxa.menu.mtwr <- 
     ps.menu %>% 
     subset_samples(day %in% c('Monday',
                       'Tuesday',
                       'Wednesday',
                       'Thursday')) %>% 
     prune_taxa(taxa_sums(.) > 0, .) %>% 
     taxa_names()
```

```{r}
# How many taxa does this remove?
ntaxa(ps.menu) - length(taxa.menu.mtwr)
```

```{r}
# Get union of taxa detected in the two analyses
all <- 
     union(taxa_names(ps.mb),
           taxa_names(ps.protein)) %>% 
     union(taxa_names(ps.menu)) # All taxa
     # union(taxa.menu.mtwr) # Mon-Thurs menus only

length(all)
```

```{r}
# Make dataframe for plot input
plot.df <- 
     data.frame(taxon = all) %>% 
     mutate(dna = taxon %in% taxa_names(ps.mb),
            protein = taxon %in% taxa_names(ps.protein),
            menu = taxon %in% taxa_names(ps.menu))
            # menu = taxon %in% taxa.menu.mtwr)

# Quick checks
sum(plot.df$dna) # Should be 78
sum(plot.df$protein) # Should be 158
sum(plot.df$menu) # Should be 151 (133 if Mon-Thurs only)
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
# pdf(here('results',
#          'QC',
#          paste0(plotID, '_Taxa Euler diagram across methods, with counts, only M-R menu.pdf')),
#     height = 4, width = 4)

euler(plot.df[,c('dna', 'protein', 'menu')]) %>% 
     plot(
          # fills = c('firebrick3', 'steelblue4', 'goldenrod2'),
          fills = c('#b66353', '#849db1', '#d7ce9f'),
          labels = TRUE,
          shape = "ellipse", 
          quantities = TRUE
          )
```

### Differences by dataset

```{r}
# In all
filter(plot.df,
       dna & protein & menu) %>% 
     pull(taxon) %>% 
     sort()
```

```{r}
# In DNA and protein, missing from menu
filter(plot.df,
       dna & protein & !menu) %>% 
     pull(taxon) %>% 
     sort()
```

Cashew, tea, goat, chickpea, hazelnut, sunflower, banana, pistachio, raspberry, sorghum.

Confirmed each of these against menu.  Notes:
* No record of other nuts: cashew, hazelnut, pistachio
* Capra hircus named as Capra aegagrus subsp. hircus: Will correct in menu
* 

## Per-sample details

Overall view:
```{r echo = FALSE}
# How many food taxa consumed daily (menu)?
melt.menu <- psmelt(ps.menu)
counts.menu <- 
     melt.menu %>% 
     filter(Abundance > 0) %>% 
     group_by(ID, date, kingdom) %>% 
     count(name = 'menu_taxa')
```

```{r echo = FALSE}
# How many food taxa detected daily (metaproteomics)?
melt.protein <- psmelt(ps.protein)
counts.protein <- 
     melt.protein %>% 
     filter(Abundance > 4) %>% 
     group_by(subj, date, kingdom) %>% 
     count(name = 'protein_taxa')
```

```{r}
# Note some NA values here: what are they?
melt.protein %>% 
     filter(Abundance > 4) %>% 
     filter(is.na(kingdom)) %>% 
     pull(OTU) %>% 
     unique()
```
If one defines the kingdom Plantae to mean the Archaeplastida, the red algae will be part of that kingdom.
If Plantae are defined more narrowly, to be the Viridiplantae, then the red algae might be considered their own kingdom, or part of the kingdom Protista.

```{r}
# Do this for only samples in common
# Though need to preserve Fungi, which aren't measured by metabarcoding
# This can be done by doing a full join
counts <- 
     counts.protein %>% 
     full_join(counts.mb) %>% 
     filter(!is.na(kingdom)) %>% 
     mutate(delta1 = date - 1) %>% 
     # Join to menu, lagged 1 day 
     # Full join to add kingdom if necessary
     full_join(counts.menu, by = c('kingdom',
                                   'subj' = 'ID', 
                                   'delta1' = 'date')) %>% 
     # Then remove elements that didn't join by the date
     filter(!is.na(date))

# Long version for plotting
counts.long <- pivot_longer(counts, 
                            cols = ends_with('taxa'),
                            names_to = 'dataset',
                            values_to = 'value')
```

```{r}
# Update names for plotting
counts.long$dataset <- factor(counts.long$dataset, 
                                     levels = c('menu_taxa', 
                                                'protein_taxa',
                                                'dna_taxa'), 
                                     labels = c('Recorded menu',
                                                'Metaproteomics',
                                                'Metabarcoding'))
# Refactor for plotting
counts.long$kingdom <- factor(counts.long$kingdom,
                              levels = c('Viridiplantae',
                                         'Metazoa', 
                                         'Fungi'))
```

```{r}
ggplot(counts.long, aes(x = kingdom, y = value, fill = dataset)) +
     geom_boxplot() +
     labs(title = 'Number of food taxa identified by measure',
          subtitle = '>4 PSMs per taxon filter',
          x = 'Kingdom',
          y = 'Number of taxa', fill = 'Measure') +
     ylim(0, 65) + # Leave room to add significance
     # scale_y_log10() +
     scale_fill_manual(values = c('#d7ce9f', '#849db1', '#b66353')) +
     theme_classic() +
     theme()
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
ggsave(
     here(
          'results',
          'QC',
          paste0(plotID,
                 '_Number of food taxa by measure, shared samples only.pdf'
                 )
          ),
     height = 3,
     width = 6)
```

Statistics on above plot: 

```{r}
# Repeated measures ANOVA (Viridiplantae)
# Get only plant detection events, make unique sample ID
viridiplantae <- 
     counts.long %>% 
     mutate(sample = paste(subj, date, sep = '_')) %>% 
     filter(kingdom == 'Viridiplantae') 

model <- lm(value ~ dataset + sample,
            data = viridiplantae)

analysis <- Anova(model, 
                  idata = viridiplantae,
                  idesign = ~sample)

print(analysis)
```

```{r}
# Post-hoc testing

# Unadjusted p values
pairwise.t.test(viridiplantae$value,
                viridiplantae$dataset,
                p.adjust.method = 'BH')
```

```{r}
# Repeated measures ANOVA (Metazoa)
metazoa <- 
     counts.long %>% 
     mutate(sample = paste(subj, date, sep = '_')) %>% 
     filter(kingdom == 'Metazoa') 

model <- lm(value ~ dataset + sample,
            data = metazoa)

analysis <- Anova(model, 
                  idata = metazoa,
                  idesign = ~sample)

print(analysis)
```

```{r}
# Unadjusted p values
pairwise.t.test(metazoa$value,
                metazoa$dataset,
                p.adjust.method = 'BH')
```


```{r}
# Paired t-test (Fungi)
fungi <- 
     filter(counts, kingdom == 'Fungi')

t.test(fungi$menu_taxa, fungi$protein_taxa,
       paired = TRUE)
```

Patterns of recorded foods by day of week: 

```{r echo = FALSE}
counts.long %>% 
     group_by(ID, date, dataset) %>% 
     summarize(value = sum(value, na.rm = TRUE)) %>% 
     filter(dataset == 'Recorded menu') %>% 
     ggplot(aes(x = date, y = value)) +
     geom_point(aes(color = dataset)) +
     facet_wrap(~ ID, nrow = 4, scales = 'free_x') + 
     labs(x = 'Date', y = 'Number of taxa') +
     theme_bw() +
     theme(axis.title.x = element_blank(),
           legend.title = element_blank(),
           legend.position = 'bottom')

plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
# ggsave(here('results', 
#             paste0(plotID, 
#                    '_Number of food taxa in menu by day.pdf')),
#        height = 7, width = 4)
```

