---
title: "Kleiner lab database testing"
output:
  html_document:
    df_print: paged
---

Here we are evaluating the preliminary data shared by the Kleiner lab at NCSU to determine if their metaproteomic database should be adjusted (e.g. filtering out homologs to human proteins, other ideas?) before continuing with analysis of the DFC samples.

# R setup

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = file.path(getwd(), '..'))
```

```{r libraries, echo=FALSE}
library(eulerr) # For Euler plots
library(here) 
library(phyloseq) 
library(RColorBrewer)
library(tidyverse) 
```

# Read in data

## Shared samples

```{r}
# Read in and name comparably to phyloseqs (SUBJ_DATE)
samples <- 
     here('data', 
          'metadata', 
          '20200723_MTA DFC samples shared.csv') %>% 
     read_csv() 
```

## Metaproteomic phyloseq

```{r}
ps.protein <- 
     here('data', 
          'processed', 
          'phyloseq',
          # '20210408_ps_mp.rds') # 1PUP analysis
          '20210912_ps_mp.rds') %>%  # 1UP analysis
     readRDS()
          
ps.protein
```

## DFC menu data phyloseq

Load in version generated in "Pre-process menu" notebook: 

```{r}
ps.menu <- 
     here('data', 
          'processed', 
          'phyloseq',
          'cleaned',
          'WORKING_ps_menu.rds') %>% 
     readRDS()

ps.menu
```

## Metabarcoding phyloseq

Load in version generated in "Pre-process metabarcoding" notebook: 

```{r}
ps.mb <- 
     here('data', 
          'processed', 
          'phyloseq',
          'cleaned',
          'WORKING_ps_mb.rds') %>% 
     readRDS()

ps.mb
```

# Pre-process

## Shared samples

```{r}
# Name comparably to phyloseq objects (SUBJ_DATE)
samples <- 
     samples %>% 
     mutate(sample.date = as.Date(sample.date, format = '%m/%d/%y'),
            name = paste(subj, sample.date, sep = '_')) %>%
     pull(name)
               
samples %>% 
     sort()
```

## Metaproteomic phyloseq

Impose PSM-based filter steps downstream (when determining what counts as a positive) so as not to alter the original object.

## Metabarcoding phyloseq

Impose count-based filter steps downstream (when determining what counts as a positive) so as not to alter the original object.

```{r}
# Filter to only the samples under consideration here
# TODO: Decide: do I want to do this??
ps.mb <- 
     ps.mb %>% 
     prune_samples(sample_names(.) %in% samples, .) %>% 
     prune_taxa(taxa_sums(.) > 0, .)

ps.mb
```

## Menu

# Exploratory comparison

## Abundances

### PSMs

```{r}
psms <- 
     ps.protein@otu_table@.Data %>% 
     as.vector() %>% 
     data.frame(psms = .) %>% 
     filter(psms > 0)

head(psms)
```
```{r}
ggplot(psms, aes(x = psms)) +
     geom_histogram(binwidth = 5,
                    boundary = 0) +
     labs(x = 'PSMs', y = 'Count') +
     theme_bw()
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
# ggsave(here('results', 'QC',
#             paste0(plotID, '_PSM count histogram, v2 analysis.pdf')),
#        height = 1)
```

### Sequence counts

```{r}
counts <- 
     ps.mb@otu_table@.Data %>% 
     as.vector() %>% 
     data.frame(counts = .) %>% 
     filter(counts > 0)

head(counts)
```

```{r}
ggplot(counts, aes(x = counts)) +
     geom_histogram(binwidth = 5,
                    boundary = 0) +
     labs(x = 'Sequence counts', y = 'Count') +
     theme_bw()
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
# ggsave(here('results', 'QC',
#             paste0(plotID, '_Sequence count histogram, v3 analysis.pdf')),
#        height = 1)
```


### Menu

```{r}
grams <- 
     ps.menu@otu_table@.Data %>% 
     as.vector() %>% 
     data.frame(grams = .) %>% 
     filter(grams > 0)

head(grams)
```

```{r}
ggplot(grams, aes(x = grams)) +
     geom_histogram(binwidth = 5,
                    boundary = 0) +
     labs(x = 'Food mass (g)', y = 'Count') +
     theme_bw()
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
# ggsave(here('results', 'QC',
#             paste0(plotID, '_Food mass histogram, v3 analysis.pdf')),
#        height = 1)
```

## Detected samples

The full set of samples shared was
```{r}
length(samples)
```

Detected by metaproteomics:
```{r}
nsamples(ps.protein)
```

Detected by metabarcoding:
```{r}
sum(samples %in% sample_names(ps.mb))
```
However, this includes both plants and animals.  Separating out, see that some are missing:
```{r}
detected.mb <- 
     ps.mb %>% 
     psmelt %>% 
     filter(kingdom %in% c('Viridiplantae', 'Metazoa')) %>% 
     group_by(subj, date, kingdom) %>% 
     summarize(detected = sum(Abundance) > 0) %>% 
     mutate(name = paste(subj, date, sep = '_')) %>% 
     ungroup() %>% 
     select(name, kingdom, detected) %>% 
     filter(name %in% samples)

detected.mb
```
```{r}
detected.mb %>% 
     group_by(kingdom, detected) %>% 
     count()
```

## Detected taxa

```{r}
asvtab.protein <- otu_table(ps.protein)@.Data
asvtab.mb <- otu_table(ps.mb)@.Data
asvtab.menu <- otu_table(ps.menu)@.Data
```

What species are shared between metaproteomic, metabarcoding, and menu datasets?

### Euler plot

Note that for menu data, perhaps best not to include the full range as this might include taxa we'd never expect to see in samples.  

```{r}
# Get only those taxa in Mon-Thu menu
taxa.menu.mtwr <- 
     ps.menu %>% 
     subset_samples(day %in% c('Monday',
                       'Tuesday',
                       'Wednesday',
                       'Thursday')) %>% 
     prune_taxa(taxa_sums(.) > 0, .) %>% 
     taxa_names()
```

```{r}
# How many taxa does this remove?
ntaxa(ps.menu) - length(taxa.menu.mtwr)
```

```{r}
# Get union of taxa detected in the two analyses
all <- 
     union(taxa_names(ps.mb),
           taxa_names(ps.protein)) %>% 
     union(taxa_names(ps.menu)) # All taxa
     # union(taxa.menu.mtwr) # Mon-Thurs menus only

length(all)
```

```{r}
# Make dataframe for plot input
plot.df <- 
     data.frame(taxon = all) %>% 
     mutate(dna = taxon %in% taxa_names(ps.mb),
            protein = taxon %in% taxa_names(ps.protein),
            menu = taxon %in% taxa_names(ps.menu))
            # menu = taxon %in% taxa.menu.mtwr)

# Quick checks
sum(plot.df$dna) # Should be 78
sum(plot.df$protein) # Should be 158
sum(plot.df$menu) # Should be 151 (133 if Mon-Thurs only)
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
# pdf(here('results',
#          'QC',
#          paste0(plotID, '_Taxa Euler diagram across methods, with counts, only M-R menu.pdf')),
#     height = 4, width = 4)

euler(plot.df[,c('dna', 'protein', 'menu')]) %>% 
     plot(
          # fills = c('firebrick3', 'steelblue4', 'goldenrod2'),
          fills = c('#b66353', '#849db1', '#d7ce9f'),
          labels = TRUE,
          shape = "ellipse", 
          quantities = TRUE
          )
```

### Differences by dataset

```{r}
# In all
filter(plot.df,
       dna & protein & menu) %>% 
     pull(taxon) %>% 
     sort()
```

```{r}
# In DNA and protein, missing from menu
filter(plot.df,
       dna & protein & !menu) %>% 
     pull(taxon) %>% 
     sort()
```

Cashew, tea, goat, chickpea, hazelnut, sunflower, banana, pistachio, raspberry, sorghum.

Confirmed each of these against menu.  Notes:
* No record of other nuts: cashew, hazelnut, pistachio
* Capra hircus named as Capra aegagrus subsp. hircus: Will correct in menu
* 

## Per-sample details

Overall view:
```{r echo = FALSE}
# How many food taxa consumed daily (menu)?
melt.menu <- psmelt(ps.menu)
counts.menu <- 
     melt.menu %>% 
     filter(Abundance > 0) %>% 
     group_by(ID, date, kingdom) %>% 
     count(name = 'menu_taxa')
```

```{r echo = FALSE}
# How many food taxa detected daily (metaproteomics)?
melt.protein <- psmelt(ps.protein)
counts.protein <- 
     melt.protein %>% 
     filter(Abundance > 4) %>% 
     group_by(subj, date, kingdom) %>% 
     count(name = 'protein_taxa')
```

```{r}
# Note some NA values here: what are they?
melt.protein %>% 
     filter(Abundance > 4) %>% 
     filter(is.na(kingdom)) %>% 
     pull(OTU) %>% 
     unique()
```
If one defines the kingdom Plantae to mean the Archaeplastida, the red algae will be part of that kingdom.
If Plantae are defined more narrowly, to be the Viridiplantae, then the red algae might be considered their own kingdom, or part of the kingdom Protista.

```{r}
# Do this for only samples in common
# Though need to preserve Fungi, which aren't measured by metabarcoding
# This can be done by doing a full join
counts <- 
     counts.protein %>% 
     full_join(counts.mb) %>% 
     filter(!is.na(kingdom)) %>% 
     mutate(delta1 = date - 1) %>% 
     # Join to menu, lagged 1 day 
     # Full join to add kingdom if necessary
     full_join(counts.menu, by = c('kingdom',
                                   'subj' = 'ID', 
                                   'delta1' = 'date')) %>% 
     # Then remove elements that didn't join by the date
     filter(!is.na(date))

# Long version for plotting
counts.long <- pivot_longer(counts, 
                            cols = ends_with('taxa'),
                            names_to = 'dataset',
                            values_to = 'value')
```

```{r}
# Update names for plotting
counts.long$dataset <- factor(counts.long$dataset, 
                                     levels = c('menu_taxa', 
                                                'protein_taxa',
                                                'dna_taxa'), 
                                     labels = c('Recorded menu',
                                                'Metaproteomics',
                                                'Metabarcoding'))
# Refactor for plotting
counts.long$kingdom <- factor(counts.long$kingdom,
                              levels = c('Viridiplantae',
                                         'Metazoa', 
                                         'Fungi'))
```

```{r}
ggplot(counts.long, aes(x = kingdom, y = value, fill = dataset)) +
     geom_boxplot() +
     labs(title = 'Number of food taxa identified by measure',
          subtitle = '>4 PSMs per taxon filter',
          x = 'Kingdom',
          y = 'Number of taxa', fill = 'Measure') +
     ylim(0, 65) + # Leave room to add significance
     # scale_y_log10() +
     scale_fill_manual(values = c('#d7ce9f', '#849db1', '#b66353')) +
     theme_classic() +
     theme()
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
ggsave(
     here(
          'results',
          'QC',
          paste0(plotID,
                 '_Number of food taxa by measure, shared samples only.pdf'
                 )
          ),
     height = 3,
     width = 6)
```

Statistics on above plot: 

```{r}
# Repeated measures ANOVA (Viridiplantae)
# Get only plant detection events, make unique sample ID
viridiplantae <- 
     counts.long %>% 
     mutate(sample = paste(subj, date, sep = '_')) %>% 
     filter(kingdom == 'Viridiplantae') 

model <- lm(value ~ dataset + sample,
            data = viridiplantae)

analysis <- Anova(model, 
                  idata = viridiplantae,
                  idesign = ~sample)

print(analysis)
```

```{r}
# Post-hoc testing

# Unadjusted p values
pairwise.t.test(viridiplantae$value,
                viridiplantae$dataset,
                p.adjust.method = 'BH')
```

```{r}
# Repeated measures ANOVA (Metazoa)
metazoa <- 
     counts.long %>% 
     mutate(sample = paste(subj, date, sep = '_')) %>% 
     filter(kingdom == 'Metazoa') 

model <- lm(value ~ dataset + sample,
            data = metazoa)

analysis <- Anova(model, 
                  idata = metazoa,
                  idesign = ~sample)

print(analysis)
```

```{r}
# Unadjusted p values
pairwise.t.test(metazoa$value,
                metazoa$dataset,
                p.adjust.method = 'BH')
```


```{r}
# Paired t-test (Fungi)
fungi <- 
     filter(counts, kingdom == 'Fungi')

t.test(fungi$menu_taxa, fungi$protein_taxa,
       paired = TRUE)
```

Patterns of recorded foods by day of week: 

```{r echo = FALSE}
counts.long %>% 
     group_by(ID, date, dataset) %>% 
     summarize(value = sum(value, na.rm = TRUE)) %>% 
     filter(dataset == 'Recorded menu') %>% 
     ggplot(aes(x = date, y = value)) +
     geom_point(aes(color = dataset)) +
     facet_wrap(~ ID, nrow = 4, scales = 'free_x') + 
     labs(x = 'Date', y = 'Number of taxa') +
     theme_bw() +
     theme(axis.title.x = element_blank(),
           legend.title = element_blank(),
           legend.position = 'bottom')

plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
# ggsave(here('results', 
#             paste0(plotID, 
#                    '_Number of food taxa in menu by day.pdf')),
#        height = 7, width = 4)
```

## Metaproteomics x menu

### Overall dataset

What species are in metaproteomic dataset that are never recorded in menu data (on any day?)

```{r}
fp <- setdiff(taxa_names(ps.protein), taxa_names(ps.menu))
length(fp)
```

```{r}
fp
```

This comparison is a bit tricky because some food taxa names aren't fully specified in our transcription of the menu.  Examples:
* 

What kingdom do these originate from?  Do we have more false-positive animals than plants, for example?

```{r}
# Pull these rows from the taxonomy table
fp.taxonomy <- taxtab.protein[fp, ]
fp.taxonomy %>% 
     data.frame() %>% 
     group_by(phylum) %>% 
     count() %>% 
     arrange(desc(n))
```

Interesting-- at least in this high-level view, it's mostly plants. 

```{r}
# What plants? Use this to ID fruits, below:
fp.taxonomy %>% 
     data.frame() %>% 
     filter(kingdom == 'Viridiplantae')
```
#### Flag foods for separate facet
```{r}
# Note that this was done for V3, which I assume is more permissive, so should work for V2? But also want to double-check.
fruits <- 
     # Joint in both
     c(
          'Ananas comosus', # Pineapple
          'Citrus reticulata', # Mandarin 
          'Citrus sinensis', # Orange
          'Citrus x paradisi', # Grapefruit
          'Fragaria x ananassa', # Strawberry
          'Malus domestica', # Apple
          'Mangifera indica', # Mango
          'Prunus avium', # Cherry
          'Prunus persica', # Peach
          'Punica granatum', # Pomegranate
          'Vitis vinifera', # Grape
          # From metaproteomics
          'Actinidia deliciosa', # Kiwi
          'Carica papaya', # Papaya
          'Citrullus lanatus', # Watermelon
          'Citrus maxima', # Pomelo
          'Musa acuminata', # Banana
          'Musa x paradisiaca', # Banana
          'Passiflora edulis', # Passionfruit
          'Psidium guajava', # Guava
          'Pyrus communis', # Pear
          'Selenicereus undatus', # Dragon fruit
          'Vitis rotundifolia', # Muscadine grape
          'Morus nigra', # Black mulberry
          # From menu
          'Ficus carica', # Fig
          'Prunus armeniaca', # Apricot
          'Vaccinium tenellum' # Blueberry
     )
            
beverages <- 
     c(
          'Camellia sinensis', # Tea
          'Coffea arabica', # Coffee
          'Coffea canephora', # Coffee
          'Humulus lupulus' # Hops (beer)
     )

other <- c('Cannabis sativa')
```

Note: had a few genus-level entries here that are written as species above instead, because there was a finer species-level entry that captured them, and it would have been hard to synchronize shift to family or genus level otherwise.  These were Coffea, Citrus, Musa, Prunus, Pyrus, and Fragaria.

```{r}
facet <- 
     c(fruits, beverages, other) %>% 
     data.frame(species = .)

# Get [family, genus] category of these foods from combined taxonomy (metaproteomics and menu)
# Join to taxonomy to get kingdom-level info
# Make overall taxonomy matrix (combined menu and metaproteomics)
taxtab <- 
     bind_rows(
          data.frame(ps.menu@tax_table),
          data.frame(ps.protein@tax_table)
          ) 

# Join level desired based on glom, above
taxtab <- 
     taxtab %>% 
     select(family, genus, species) %>% 
     distinct()

facet <- left_join(facet, taxtab)
```

Try visualizing with metacoder?

### Pairing with 1-2 days prior intake

For this broadest level comparison, let's look only at the two days prior to sampling.

```{r echo = FALSE}
# Pull sample data for stool samples
samdf.protein <- 
     sample_data(ps.protein) %>% 
     as('data.frame') %>% 
     rownames_to_column(var = 'row')

# Label each stool sample with its two preceding days
for (delta in 1:2){
     varname <- paste0('delta', delta)
     samdf.protein <- mutate(samdf.protein, 
                             !!varname := paste(subj, date - delta, 
                                                sep = '_'))
}

# Replace in phyloseq object
sample_data(ps.protein) <- column_to_rownames(samdf.protein, var = 'row')
```

```{r}
# Can't use merge_samples on phyloseq here because some samples make up >1 merge

# Do manually on ASV table
asvtab.menu <- otu_table(ps.menu)@.Data
asvtab.menu.12 <- data.frame()

for (sample in seq(dim(samdf.protein)[1])){
     # Pull menu day -1
     one.before <- asvtab.menu[samdf.protein$delta1[sample], ]
     # Pull menu day -2
     two.before <- asvtab.menu[samdf.protein$delta2[sample], ]
     # Merge and place in new, aggregated OTU table
     asvtab.menu.12 <- rbind(asvtab.menu.12,
                             one.before + two.before)
     # Update food names (only has to be done once)
     if (sample == 1){names(asvtab.menu.12) <- names(one.before + two.before)}
     # Update sample name
     row.names(asvtab.menu.12)[sample] <- samdf.protein$row[sample]
}

# Now rebuild a subsetted phyloseq object
ps.menu.12 <- phyloseq(otu_table(asvtab.menu.12, taxa_are_rows = FALSE),
                       sample_data(ps.protein), # Now this matches
                       tax_table(ps.menu)) # Menu-specific taxonomy

# Remove any taxa that aren't present any longer
ps.menu.12 <- prune_taxa(taxa_sums(ps.menu.12) > 0, ps.menu.12)
ps.menu.12
```

There are `r ntaxa(ps.menu.12)` recorded in the menu in the 2 days prior to each sample, and `r ntaxa(ps.protein)` detected by metaproteomics.

Overlap between the two? Good-- this is pretty similar in overall number to the full dataset (not filtering by menu day-- that gave 69 intersecting taxon names). 

#### Glom

```{r}
# ps.menu.12 <- tax_glom(ps.menu.12, 'genus')
# taxa_names(ps.menu.12) <- tax_table(ps.menu.12)[, 'genus']
# 
# ps.protein <- tax_glom(ps.protein, 'genus')
# taxa_names(ps.protein) <- tax_table(ps.protein)[, 'genus']
```

```{r}
# ps.menu.12 <- tax_glom(ps.menu.12, 'family')
# taxa_names(ps.menu.12) <- tax_table(ps.menu.12)[, 'family']
# 
# ps.protein <- tax_glom(ps.protein, 'family')
# taxa_names(ps.protein) <- tax_table(ps.protein)[, 'family']
```

### Paired samples, per-food

```{r}
# Extract OTU tables
asvtab.protein <- otu_table(ps.protein)@.Data
asvtab.menu.12 <- otu_table(ps.menu.12)@.Data

# Synchronize names (not all genera in common between the two)
dim(asvtab.protein)
dim(asvtab.menu.12)
length(union(colnames(asvtab.menu.12), colnames(asvtab.protein)))
```

```{r}
# Helper function to pad out columns
pad_columns <- function(x, y, fill = 0){
     # Takes two matrices and synchronizes columns across them, filling added 
     # cols with a set value
     # Returns a list of the two updated matrices
     
     # Find missing columns in both matrices
     missing.x <- setdiff(colnames(y), colnames(x))
     missing.y <- setdiff(colnames(x), colnames(y))
     
     # Pad out columns of x
     fill.x <- matrix(fill, 
                      nrow = dim(x)[1],
                      ncol = length(missing.x))
     colnames(fill.x) <- missing.x 
     x <- cbind(x, fill.x)
     
     # Pad out columns of y
     fill.y <- matrix(fill, 
                      nrow = dim(y)[1],
                      ncol = length(missing.y))
     colnames(fill.y) <- missing.y
     y <- cbind(y, fill.y)
     
     # Arrange the columns so they appear in identical order
     u <- sort(union(colnames(x), colnames(y)))
     x <- x[, u]
     y <- y[, u]
     
     list(x, y)
}
```

```{r}
padded <- pad_columns(asvtab.protein, asvtab.menu.12)
asvtab.protein <- padded[[1]]
asvtab.menu.12 <- padded[[2]]

rm(padded)
```

```{r}
# Alternate: at species level, *only* consider the intersection of the two
shared <- intersect(colnames(asvtab.menu.12), colnames(asvtab.protein))
length(shared)

# Subset to these columns only
asvtab.protein <- asvtab.protein[, shared]
asvtab.menu.12 <- asvtab.menu.12[, shared]
```

```{r}
# Confirm row and column names are equal before proceeding
all(row.names(asvtab.protein) == row.names(asvtab.menu.12))
all(colnames(asvtab.protein) == colnames(asvtab.menu.12))
```

```{r}
# Label predictions 
protein.pos <- asvtab.protein > 0 # Already filtered, above
menu.pos <- asvtab.menu.12 > 0

tp <- protein.pos & menu.pos
tn <- !protein.pos & !menu.pos
fp <- protein.pos & !menu.pos
fn <- !protein.pos & menu.pos

pred <- bind_rows(colSums(tp),
                  colSums(tn),
                  colSums(fp),
                  colSums(fn))

pred$prediction <- c('tp', 'tn', 'fp', 'fn')

pred <- select(pred, prediction, everything())
```

```{r}
# Reformat to long for visualization
pred.long <- pivot_longer(pred,
                          cols = -prediction,
                          names_to = 'taxon',
                          values_to = 'count')

# Factor data for visualization
# Calculate F-measure to be used downstream
pred.summary <- 
     pred.long %>% 
     pivot_wider(names_from = prediction, values_from = count) %>% 
     mutate(recall = tp/(tp + fn),
            precision = tp/(tp + fp),
            f_measure = (2 * recall * precision)/(recall + precision),
            pos_n = tp + fp) %>% 
     # mutate(certain = ifelse(taxon %in% facet$family,
     # mutate(certain = ifelse(taxon %in% facet$genus,
     mutate(certain = ifelse(taxon %in% unlist(facet), # At any taxonomic level
                             yes = 0,
                             no = 1)) %>% 
     arrange(desc(certain), desc(f_measure), fn, fp)

pred.summary
```
```{r}
# Join to taxonomy to get kingdom-level info
# Make overall taxonomy matrix (combined menu and metaproteomics)
taxtab <- 
     bind_rows(
          data.frame(ps.menu.12@tax_table),
          data.frame(ps.protein@tax_table)
          ) 

# Join level desired based on glom, above
taxtab <- 
     taxtab %>% 
     # select(kingdom, family) %>%
     # select(kingdom, genus) %>% 
     rownames_to_column(var = 'taxon') %>% 
     mutate(taxon = gsub('\\.+\\d+', '', taxon)) %>% 
     select(kingdom, taxon) %>% 
     distinct()

pred.summary <- left_join(pred.summary, 
                       taxtab,
                       # by = c('taxon' = 'family'))
                       # by = c('taxon' = 'genus'))
                       by = 'taxon')

# Recode NA kingdom as "Other"
pred.summary$kingdom[is.na(pred.summary$kingdom)] <- 'Other'
```

```{r}
# Incorporate into object
pred.long <- 
     pred.summary %>% 
     select(certain, taxon, kingdom) %>% 
     right_join(pred.long) %>% 
     select(certain, kingdom, taxon, everything())
```

```{r}
# Now that joins complete, add factor levels
pred.long <- 
     pred.long %>% 
     mutate(taxon = factor(taxon, levels = pred.summary$taxon),
            prediction = factor(prediction, levels = c('tp', 'tn', 'fp', 'fn'),
                                labels = c('True positive', 'True negative',
                                           'False positive', 'False negative')),
            kingdom = factor(kingdom,
                             levels = c('Viridiplantae',
                                        'Metazoa',
                                        'Fungi',
                                        'Other')))
```

```{r}
# Save for comparison to metabarcoding data
saveRDS(pred.long,
        here('data', 'processed', 'performance-metrics', 
             '20210913_Metaproteomic v. menu predictions by shared food taxa, 1-2d prior, >4 counts per taxon filter, no intake filter, v3 analysis.rds'))
```

#### Plot

```{r}
pred.long %>%
     filter(certain == 1) %>% 
     ggplot(aes(x = taxon, y = count, fill = prediction)) +
     geom_bar(position = 'stack', stat = 'identity') +
     scale_fill_manual(values = c('#1A9641','#A6D96A', # TP, TN
                                  '#D7191C', '#FDAE61')) + # FP, FN
     facet_grid(cols = vars(kingdom), 
                scales = 'free',
                space = 'free') +
     theme_classic() +
     theme(axis.text.x = element_text(angle = 20, hjust = 1),
           axis.text.y = element_blank(), 
           axis.ticks = element_blank(),
           axis.line = element_blank(),
           axis.title = element_blank(),
           legend.position = 'top',
           legend.title = element_blank(), 
           plot.margin = margin(0, 0.5, 0.5, 0.5, 'in')) 
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

width <- 
     pred.long %>% 
     filter(certain == 1) %>% 
     dim()

# ggsave(here('results', 
#             'per-food performance',
#             paste0(plotID, 
#                    '_By-food (shared taxa) predictions compared to menu, >4 PSMs per taxon filter, no intake filter, v3 analysis, certain.pdf')),
#        height = 4, width = width[1]/4 * 0.5,
#        limitsize = FALSE)
```

```{r}
# Refactor so positives are grouped:
pred.summary <- arrange(pred.summary,
                        desc(pos_n), taxon)

pred.long <- 
     pred.long %>% 
     mutate(prediction = factor(prediction, 
                                levels = c('True negative',
                                          'False negative',
                                          'True positive', 
                                          'False positive')),
            taxon = factor(taxon,
                           levels = pred.summary$taxon))
```

```{r}
pred.long %>% 
     filter(certain == 0) %>% 
     ggplot(aes(x = taxon, y = count, fill = prediction)) +
     geom_bar(position = 'stack', stat = 'identity') +
     scale_fill_manual(values = c('gray80',
                                  'gray80', 
                                  '#1A9641', 
                                  '#1A9641')) + 
     facet_grid(cols = vars(kingdom), 
                scales = 'free',
                space = 'free') +
     theme_classic() +
     theme(axis.text.x = element_text(angle = 20, hjust = 1),
           axis.text.y = element_blank(),
           axis.ticks = element_blank(),
           axis.line = element_blank(),
           axis.title = element_blank(),
           legend.position = 'top',
           legend.title = element_blank(), 
           plot.margin = margin(0, 0.5, 0, 0.5, 'in'),
           strip.text.x = element_blank()) 

```

```{r}
width <- 
     pred.long %>% 
     filter(certain == 0) %>% 
     dim()

# ggsave(here('results', 
#             'per-food performance',
#             paste0(plotID, 
#                    '_By-food (shared taxa) predictions compared to menu, >4 PSMs per taxon filter, no intake filter, v2 analysis, uncertain.pdf')),
#        height = 4, width = width[1]/4 * 0.7)
```

## Metabarcoding x menu

### Glom

```{r}
# ps.menu.12 <- tax_glom(ps.menu.12, 'genus')
# taxa_names(ps.menu.12) <- tax_table(ps.menu.12)[, 'genus']
# 
# ps.mb <- tax_glom(ps.mb, 'genus')
# taxa_names(ps.mb) <- tax_table(ps.mb)[, 'genus']
```

```{r}
ps.menu.12 <- tax_glom(ps.menu.12, 'family')
taxa_names(ps.menu.12) <- tax_table(ps.menu.12)[, 'family']

ps.mb <- tax_glom(ps.mb, 'family')
taxa_names(ps.mb) <- tax_table(ps.mb)[, 'family']
```

### Paired samples, per-food

```{r}
# Extract OTU tables
asvtab.mb <- otu_table(ps.mb)@.Data
asvtab.menu.12 <- otu_table(ps.menu.12)@.Data

# Synchronize names (not all genera in common between the two)
dim(asvtab.mb)
dim(asvtab.menu.12)
length(union(colnames(asvtab.menu.12), colnames(asvtab.mb)))
```

```{r}
# Helper function to pad out columns
pad_columns <- function(x, y, fill = 0){
     # Takes two matrices and synchronizes columns across them, filling added 
     # cols with a set value
     # Returns a list of the two updated matrices
     
     # Find missing columns in both matrices
     missing.x <- setdiff(colnames(y), colnames(x))
     missing.y <- setdiff(colnames(x), colnames(y))
     
     # Pad out columns of x
     fill.x <- matrix(fill, 
                      nrow = dim(x)[1],
                      ncol = length(missing.x))
     colnames(fill.x) <- missing.x 
     x <- cbind(x, fill.x)
     
     # Pad out columns of y
     fill.y <- matrix(fill, 
                      nrow = dim(y)[1],
                      ncol = length(missing.y))
     colnames(fill.y) <- missing.y
     y <- cbind(y, fill.y)
     
     # Arrange the columns so they appear in identical order
     u <- sort(union(colnames(x), colnames(y)))
     x <- x[, u]
     y <- y[, u]
     
     list(x, y)
}
```

```{r}
padded <- pad_columns(asvtab.mb, asvtab.menu.12)
asvtab.mb <- padded[[1]]
asvtab.menu.12 <- padded[[2]]

rm(padded)
```

```{r}
# Alternate: at species level, *only* consider the intersection of the two
shared <- intersect(colnames(asvtab.menu.12), colnames(asvtab.mb))
length(shared)

shared.samples <- intersect(row.names(asvtab.menu.12), row.names(asvtab.mb))

# Subset to these columns only
asvtab.mb <- asvtab.mb[shared.samples, shared]
asvtab.menu.12 <- asvtab.menu.12[shared.samples, shared]
```

```{r}
# Confirm row and column names are equal before proceeding
all(row.names(asvtab.mb) == row.names(asvtab.menu.12))
all(colnames(asvtab.mb) == colnames(asvtab.menu.12))
```

```{r}
# Label predictions 
protein.pos <- asvtab.mb > 0 # Already filtered, above
menu.pos <- asvtab.menu.12 > 0

tp <- protein.pos & menu.pos
tn <- !protein.pos & !menu.pos
fp <- protein.pos & !menu.pos
fn <- !protein.pos & menu.pos

pred <- bind_rows(colSums(tp),
                  colSums(tn),
                  colSums(fp),
                  colSums(fn))

pred$prediction <- c('tp', 'tn', 'fp', 'fn')

pred <- select(pred, prediction, everything())
```

```{r}
# Reformat to long for visualization
pred.long <- pivot_longer(pred,
                          cols = -prediction,
                          names_to = 'taxon',
                          values_to = 'count')

# Factor data for visualization
# Calculate F-measure to be used downstream
pred.summary <- 
     pred.long %>% 
     pivot_wider(names_from = prediction, values_from = count) %>% 
     mutate(recall = tp/(tp + fn),
            precision = tp/(tp + fp),
            f_measure = (2 * recall * precision)/(recall + precision),
            pos_n = tp + fp) %>% 
     mutate(certain = ifelse(taxon %in% facet$family,
     # mutate(certain = ifelse(taxon %in% facet$genus,
     # mutate(certain = ifelse(taxon %in% unlist(facet), # At any taxonomic level
                             yes = 0,
                             no = 1)) %>% 
     arrange(desc(certain), desc(f_measure), fn, fp)

pred.summary
```
```{r}
# Join to taxonomy to get kingdom-level info
# Make overall taxonomy matrix (combined menu and metaproteomics)
taxtab <- 
     bind_rows(
          data.frame(ps.menu.12@tax_table),
          data.frame(ps.mb@tax_table)
          ) 

# Join level desired based on glom, above
taxtab <- 
     taxtab %>% 
     select(kingdom, family) %>%
     # select(kingdom, genus) %>%
     # rownames_to_column(var = 'taxon') %>% 
     # mutate(taxon = gsub('\\.+\\d+', '', taxon)) %>% 
     # select(kingdom, taxon) %>% 
     distinct()

pred.summary <- left_join(pred.summary, 
                       taxtab,
                       by = c('taxon' = 'family'))
                       # by = c('taxon' = 'genus'))
                       # by = 'taxon')

# Recode NA kingdom as "Other"
pred.summary$kingdom[is.na(pred.summary$kingdom)] <- 'Other'
```

```{r}
# Incorporate into object
pred.long <- 
     pred.summary %>% 
     select(certain, taxon, kingdom) %>% 
     right_join(pred.long) %>% 
     select(certain, kingdom, taxon, everything())
```

```{r}
# Now that joins complete, add factor levels
pred.long <- 
     pred.long %>% 
     mutate(taxon = factor(taxon, levels = pred.summary$taxon),
            prediction = factor(prediction, levels = c('tp', 'tn', 'fp', 'fn'),
                                labels = c('True positive', 'True negative',
                                           'False positive', 'False negative')),
            kingdom = factor(kingdom,
                             levels = c('Viridiplantae',
                                        'Metazoa',
                                        'Fungi',
                                        'Other')))
```

```{r}
# Save for comparison to metabarcoding data
saveRDS(pred.long,
        here('data', 'processed', 'performance-metrics', 
             '20210913_Metabarcoding v. menu predictions by family, 1-2d prior, >4 counts per taxon filter, no intake filter.rds'))
```

#### Plot

```{r}
pred.long %>%
     filter(certain == 1) %>% 
     ggplot(aes(x = taxon, y = count, fill = prediction)) +
     geom_bar(position = 'stack', stat = 'identity') +
     scale_fill_manual(values = c('#1A9641','#A6D96A', # TP, TN
                                  '#D7191C', '#FDAE61')) + # FP, FN
     facet_grid(cols = vars(kingdom), 
                scales = 'free',
                space = 'free') +
     theme_classic() +
     theme(axis.text.x = element_text(angle = 20, hjust = 1),
           axis.text.y = element_blank(), 
           axis.ticks = element_blank(),
           axis.line = element_blank(),
           axis.title = element_blank(),
           legend.position = 'top',
           legend.title = element_blank(), 
           plot.margin = margin(0, 0.5, 0.5, 0.5, 'in')) 
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

width <- 
     pred.long %>% 
     filter(certain == 1) %>% 
     dim()

# ggsave(here('results', 
#             'per-food performance',
#             paste0(plotID, 
#                    '_Metabarcoding by-food family predictions compared to menu, >4 PSMs per taxon filter, no intake filter, certain.pdf')),
#        height = 4, width = width[1]/4 * 0.5,
#        limitsize = FALSE)
```

```{r}
# Refactor so positives are grouped:
pred.summary <- arrange(pred.summary,
                        desc(pos_n), taxon)

pred.long <- 
     pred.long %>% 
     mutate(prediction = factor(prediction, 
                                levels = c('True negative',
                                          'False negative',
                                          'True positive', 
                                          'False positive')),
            taxon = factor(taxon,
                           levels = pred.summary$taxon))
```

```{r}
pred.long %>% 
     filter(certain == 0) %>% 
     ggplot(aes(x = taxon, y = count, fill = prediction)) +
     geom_bar(position = 'stack', stat = 'identity') +
     scale_fill_manual(values = c('gray80',
                                  'gray80', 
                                  '#1A9641', 
                                  '#1A9641')) + 
     facet_grid(cols = vars(kingdom), 
                scales = 'free',
                space = 'free') +
     theme_classic() +
     theme(axis.text.x = element_text(angle = 20, hjust = 1),
           axis.text.y = element_blank(),
           axis.ticks = element_blank(),
           axis.line = element_blank(),
           axis.title = element_blank(),
           legend.position = 'top',
           legend.title = element_blank(), 
           plot.margin = margin(0, 0.5, 0, 0.5, 'in'),
           strip.text.x = element_blank()) 

```

```{r}
width <- 
     pred.long %>% 
     filter(certain == 0) %>% 
     dim()

# ggsave(here('results', 
#             'per-food performance',
#             paste0(plotID, 
#                    '_Metabarcoding by-food family predictions compared to menu, >4 PSMs per taxon filter, no intake filter, uncertain.pdf')),
#        height = 4, width = width[1]/4 * 0.7)
```