---
title: "Kleiner lab database testing"
output:
  html_document:
    df_print: paged
---

Here we are evaluating the preliminary data shared by the Kleiner lab at NCSU to determine if their metaproteomic database should be adjusted (e.g. filtering out homologs to human proteins, other ideas?) before continuing with analysis of the DFC samples.

# R setup

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = file.path(getwd(), '..'))
```

```{r libraries, echo=FALSE}
library(here) 
library(phyloseq) 
library(RColorBrewer)
library(tidyverse) 
```

## Shared samples

```{r}
# Read in and name comparably to phyloseqs (SUBJ_DATE)
samples <- 
     here('data', 'metadata', '20200723_MTA DFC samples shared.csv') %>% 
     read_csv() %>% 
     mutate(sample.date = as.Date(sample.date, format = '%m/%d/%y'),
            name = paste(subj, sample.date, sep = '_')) %>%
     pull(name)
               
samples %>% sort()
```

## Metaproteomic phyloseq

```{r}
ps.protein <- readRDS(here('data', 'processed', 'phyloseq',
                           '20210727_ps_mp.rds'))
ps.protein
```

## DFC menu data phyloseq

```{r}
ps.menu <- readRDS(here('data', 'processed', 'phyloseq',
                        '20210220_ps_menu.rds'))
ps.menu
```

## Metabarcoding phyloseq

```{r}
# Combined 12SV5 and trnL, made in combined MiniSeq notebook in DFC project folder
ps.mb <- 
     readRDS(here('data', 'processed', 'phyloseq',
                  '20210727_ps_mb.rds')) 
ps.mb
```

```{r}
# Subset to only those taxa identified to some position in the plant phylogeny
unlabeled <- 
     data.frame(ps.mb@tax_table) %>% 
     filter(is.na(superkingdom)) %>% 
     row.names()

asvtab.mb <- ps.mb@otu_table

cat('Percent of reads not assigned to a taxon:', 
    sum(asvtab.mb[, unlabeled])/sum(asvtab.mb) * 100)

ps.mb <- subset_taxa(ps.mb, !is.na(superkingdom))
```

# Exploratory comparison

## Per-sample details

Overall view:
```{r echo = FALSE}
# How many food taxa consumed daily (menu)?
melt.menu <- psmelt(ps.menu)
counts.menu <- 
     melt.menu %>% 
     filter(Abundance > 0) %>% 
     group_by(ID, date, kingdom) %>% 
     count(name = 'menu_taxa')
```

```{r echo = FALSE}
# How many food taxa detected daily (metaproteomics)?
melt.protein <- psmelt(ps.protein)
counts.protein <- 
     melt.protein %>% 
     filter(Abundance > 4) %>% 
     group_by(subj, date, kingdom) %>% 
     count(name = 'protein_taxa')
```

```{r}
# Note some NA values here: what are they?
melt.protein %>% 
     filter(Abundance > 4) %>% 
     filter(is.na(kingdom)) %>% 
     pull(OTU) %>% 
     unique()
```
All the seaweeds?  Follow up on this.

```{r}
# How many food taxa detected daily (metabarcoding)?
melt.mb <- psmelt(ps.mb)
counts.mb <- 
     melt.mb %>% 
     filter(Abundance > 0) %>% 
     group_by(subj, date, kingdom) %>% 
     count(name = 'dna_taxa')
```

```{r}
# Combine
counts <- 
     counts.menu %>% 
     left_join(counts.protein, by = c('ID' = 'subj', 'date', 'kingdom')) %>% 
     left_join(counts.mb, by = c('ID' = 'subj', 'date', 'kingdom'))
     

# Long version for plotting
counts.long <- pivot_longer(counts, 
                            cols = ends_with('taxa'),
                            names_to = 'dataset',
                            values_to = 'value')

# Update names for plotting
counts.long$dataset <- factor(counts.long$dataset, 
                              levels = c('menu_taxa', 'protein_taxa',
                                         'dna_taxa'), 
                              labels = c('Recorded menu', 'Metaproteomics',
                                         'Metabarcoding'))
```

```{r}
# Now do the same, but for only the samples in common between all 3 datasets
counts.shared <- 
     counts %>% 
     filter(((!is.na(menu_taxa) & !is.na(protein_taxa) & !is.na(dna_taxa))))

# Long version for plotting
counts.shared.long <- pivot_longer(counts.shared , 
                                   cols = ends_with('taxa'),
                                   names_to = 'dataset',
                                   values_to = 'value')

# Update names for plotting
counts.shared.long$dataset <- factor(counts.shared.long$dataset, 
                                     levels = c('menu_taxa', 
                                                'protein_taxa',
                                                'dna_taxa'), 
                                     labels = c('Recorded menu',
                                                'Metaproteomics',
                                                'Metabarcoding'))
```

```{r}
ggplot(counts.long, aes(x = kingdom, y = value, fill = dataset)) +
     geom_boxplot() +
     labs(title = 'Number of food taxa identified by source',
          subtitle = '>4 PSMs per taxa filter',
          y = 'Number of taxa', fill = 'Source') +
     theme_bw() +
     theme(axis.title.x = element_blank())

plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
# ggsave(here('results', paste0(plotID, '_Number of food taxa by source, >4 PSM filter.pdf')),
#        height = 4)
```

```{r}
ggplot(counts.shared.long, aes(x = kingdom, y = value, fill = dataset)) +
     geom_boxplot() +
     labs(title = 'Number of food taxa identified by source',
          subtitle = '>4 PSMs per taxa filter',
          y = 'Number of taxa', fill = 'Source') +
     theme_bw() +
     theme(axis.title.x = element_blank())

plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
ggsave(
     here(
          'results', 
          paste0(plotID, 
                 '_Number of food taxa by source, shared samples only.pdf'
                 )
          ),
     height = 4, 
     width = 4)
```

```{r echo = FALSE}
t.test(counts$menu_taxa, counts$protein_taxa)
```

Patterns of recorded foods by day of week: 

```{r echo = FALSE}
counts.long %>% 
     group_by(ID, date, dataset) %>% 
     summarize(value = sum(value, na.rm = TRUE)) %>% 
     filter(dataset == 'Recorded menu') %>% 
     ggplot(aes(x = date, y = value)) +
     geom_point(aes(color = dataset)) +
     facet_wrap(~ ID, nrow = 4, scales = 'free_x') + 
     labs(x = 'Date', y = 'Number of taxa') +
     theme_bw() +
     theme(axis.title.x = element_blank(),
           legend.title = element_blank(),
           legend.position = 'bottom')

plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
# ggsave(here('results', 
#             paste0(plotID, 
#                    '_Number of food taxa in menu by day.pdf')),
#        height = 7, width = 4)
```

## Metaproteomics x menu

### Apply PSM filter to phyloseq object

```{r}
## Need to be careful-- this would only apply to the following code chunk, later on can impose a threshold without altering the phyloseq object
# otutab <- otu_table(ps.protein)@.Data
# otutab[otutab <= 4] = 0
# 
# # Replace in phyloseq object
# ps.protein <- phyloseq(otu_table(otutab, taxa_are_rows = FALSE),
#                        sample_data(ps.protein),
#                        tax_table(ps.protein))
# 
# # Now remove any taxa with zero counts
# ps.protein <- prune_taxa(taxa_sums(ps.protein) > 0,
#                          ps.protein)
```

### Overall dataset

```{r}
taxtab.protein <- tax_table(ps.protein)@.Data
taxtab.menu <- tax_table(ps.menu)@.Data
```

What species are shared between metaproteomic and menu datasets (regardless of menu date?)
```{r}
intersect(row.names(taxtab.protein), row.names(taxtab.menu))
```

What species are in metaproteomic dataset that are never recorded in menu data (on any day?)

This comparison is a bit tricky because some food taxa names aren't fully specified in our transcription of the menu.  Examples:
* 

```{r}
fp <- setdiff(row.names(taxtab.protein), row.names(taxtab.menu))
fp
```

What kingdom do these originate from?  Do we have more false-positive animals than plants, for example?

```{r}
# Pull these rows from the taxonomy table
fp.taxonomy <- taxtab.protein[fp, ]
fp.taxonomy %>% 
     data.frame() %>% 
     group_by(phylum) %>% 
     count() %>% 
     arrange(desc(n))
```
Interesting-- at least in this high-level view, it's mostly plants. 

Try visualizing with metacoder?

### Pairing with 1-2 days prior intake

For this broadest level comparison, let's look only at the two days prior to sampling.

```{r echo = FALSE}
# Pull sample data for stool samples
samdf.protein <- 
     sample_data(ps.protein) %>% 
     as('data.frame') %>% 
     rownames_to_column(var = 'row')

# Label each stool sample with its two preceding days
for (delta in 1:2){
     varname <- paste0('delta', delta)
     samdf.protein <- mutate(samdf.protein, 
                             !!varname := paste(subj, date - delta, 
                                                sep = '_'))
}

# Replace in phyloseq object
sample_data(ps.protein) <- column_to_rownames(samdf.protein, var = 'row')
```

```{r}
# Can't use merge_samples on phyloseq here because some samples make up >1 merge

# Do manually on ASV table
asvtab.menu <- otu_table(ps.menu)@.Data
asvtab.menu.12 <- data.frame()

for (sample in seq(dim(samdf.protein)[1])){
     # Pull menu day -1
     one.before <- asvtab.menu[samdf.protein$delta1[sample], ]
     # Pull menu day -2
     two.before <- asvtab.menu[samdf.protein$delta2[sample], ]
     # Merge and place in new, aggregated OTU table
     asvtab.menu.12 <- rbind(asvtab.menu.12,
                             one.before + two.before)
     # Update food names (only has to be done once)
     if (sample == 1){names(asvtab.menu.12) <- names(one.before + two.before)}
     # Update sample name
     row.names(asvtab.menu.12)[sample] <- samdf.protein$row[sample]
}

# Now rebuild a subsetted phyloseq object
ps.menu.12 <- phyloseq(otu_table(asvtab.menu.12, taxa_are_rows = FALSE),
                       sample_data(ps.protein), # Now this matches
                       tax_table(ps.menu)) # Menu-specific taxonomy

# Remove any taxa that aren't present any longer
ps.menu.12 <- prune_taxa(taxa_sums(ps.menu.12) > 0, ps.menu.12)
ps.menu.12
```

There are `r ntaxa(ps.menu.12)` recorded in the menu in the 2 days prior to each sample, and `r ntaxa(ps.protein)` detected by metaproteomics.

Overlap between the two? Good-- this is pretty similar in overall number to the full dataset (not filtering by menu day-- that gave 69 intersecting taxon names). 

A bit tricky because not all options are compatibly named.  Try grouping to genus level to smooth over at least some of this.

```{r}
ps.menu.12 <- tax_glom(ps.menu.12, 'genus')
taxa_names(ps.menu.12) <- tax_table(ps.menu.12)[, 'genus']

ps.protein <- tax_glom(ps.protein, 'genus')
taxa_names(ps.protein) <- tax_table(ps.protein)[, 'genus']
```

### Paired samples, per-food

```{r}
# Extract OTU tables
asvtab.protein <- otu_table(ps.protein)@.Data
asvtab.menu.12 <- otu_table(ps.menu.12)@.Data

# Synchronize names (not all genera in common between the two)
dim(asvtab.protein)
dim(asvtab.menu.12)
length(union(colnames(asvtab.menu.12), colnames(asvtab.protein)))
```

```{r}
# Helper function to pad out columns
pad_columns <- function(x, y, fill = 0){
     # Takes two matrices and synchronizes columns across them, filling added 
     # cols with a set value
     # Returns a list of the two updated matrices
     
     # Find missing columns in both matrices
     missing.x <- setdiff(colnames(y), colnames(x))
     missing.y <- setdiff(colnames(x), colnames(y))
     
     # Pad out columns of x
     fill.x <- matrix(fill, 
                      nrow = dim(x)[1],
                      ncol = length(missing.x))
     colnames(fill.x) <- missing.x 
     x <- cbind(x, fill.x)
     
     # Pad out columns of y
     fill.y <- matrix(fill, 
                      nrow = dim(y)[1],
                      ncol = length(missing.y))
     colnames(fill.y) <- missing.y
     y <- cbind(y, fill.y)
     
     # Arrange the columns so they appear in identical order
     u <- sort(union(colnames(x), colnames(y)))
     x <- x[, u]
     y <- y[, u]
     
     list(x, y)
}
```

```{r}
padded <- pad_columns(asvtab.protein, asvtab.menu.12)
asvtab.protein <- padded[[1]]
asvtab.menu.12 <- padded[[2]]

rm(padded)
```

```{r}
# Confirm row and column names are equal before proceeding
all(row.names(asvtab.protein) == row.names(asvtab.menu.12))
all(colnames(asvtab.protein) == colnames(asvtab.menu.12))
```

```{r}
# Label predictions 
protein.pos <- asvtab.protein > 4
menu.pos <- asvtab.menu.12 > 0

tp <- protein.pos & menu.pos
tn <- !protein.pos & !menu.pos
fp <- protein.pos & !menu.pos
fn <- !protein.pos & menu.pos

pred <- bind_rows(colSums(tp),
                  colSums(tn),
                  colSums(fp),
                  colSums(fn))

pred$prediction <- c('tp', 'tn', 'fp', 'fn')

pred <- select(pred, prediction, everything())
```

```{r}
# Reformat to long for visualization
pred.long <- pivot_longer(pred,
                          cols = -prediction,
                          names_to = 'taxon',
                          values_to = 'count')

# Factor data for visualization
# Calculate F-measure to be used downstream
pred.summary <- 
     pred.long %>% 
     pivot_wider(names_from = prediction, values_from = count) %>% 
     mutate(recall = tp/(tp + fn),
            precision = tp/(tp + fp),
            f_measure = (2 * recall * precision)/(recall + precision)) %>% 
     arrange(desc(f_measure), fn, fp)

# Assign label for faceting plot
# Done by manually inspecting pred.summary
breaks <- c(rep(1, 30), rep(2, 33), rep(3, 37), rep(4, 33))
pred.summary$facet <- breaks

pred.long <- 
     pred.summary %>% 
     select(taxon, facet) %>% 
     right_join(pred.long) %>% 
     select(facet, taxon, everything())

pred.long <- 
     pred.long %>% 
     mutate(taxon = factor(taxon, levels = pred.summary$taxon),
            prediction = factor(prediction, levels = c('tp', 'tn', 'fp', 'fn'),
                                labels = c('True positive', 'True negative',
                                           'False positive', 'False negative')))
```

```{r}
# Save for comparison to metabarcoding data
saveRDS(pred.long,
        here('data', 'processed', 'performance-metrics', 
             'Metaproteomic v. menu predictions by food, 1-2d prior, >4 counts per taxon filter.rds'))
```


```{r}
ggplot(pred.long, aes(x = taxon, y = count, fill = prediction)) +
     geom_bar(position = 'stack', stat = 'identity') +
     scale_fill_manual(values = c('#1A9641','#A6D96A', # TP, TN
                                  '#D7191C', '#FDAE61')) + # FP, FN
     facet_wrap(~facet, nrow = 4, scale = 'free_x') +
     theme_bw() +
     theme(axis.text.x = element_text(angle = 45, hjust = 1),
           axis.text.y = element_blank(),
           axis.title = element_blank(),
           legend.position = 'top',
           legend.title = element_blank(), 
           plot.margin = margin(0, 0.5, 0.5, 0.5, 'in'),
           strip.text.x = element_blank()) 

plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results', paste0(plotID, '_By-food predictions compared to menu, >4 PSMs per taxon filter.pdf')),
       height = 8, width = 8)
# 2021-02-22.677: First pass
# 2021-03-10.682: Re-colored
# 2021-04-09.985: Database v2, no PSM filter
# 2021-04-09.831: Database v2, >4 PSMs/taxon/sample filter
```

## Metaproteomics x metabarcoding 

### Paired samples, per-food

```{r}
# Extract OTU tables
asvtab.protein <- otu_table(ps.protein)@.Data
asvtab.trnL <- otu_table(ps.trnL)@.Data 
asvtab.12SV5 <- otu_table(ps.12SV5)@.Data 

# Combine metabarcoding (have to do here because I'm having trouble with merge_phyloseq, above)
shared.mb <- intersect(row.names(asvtab.trnL), row.names(asvtab.12SV5))
asvtab.mb <- cbind(asvtab.trnL[shared.mb,],
                   asvtab.12SV5[shared.mb,])

# Synchronize names (not all genera in common between the two)
dim(asvtab.protein)
dim(asvtab.menu.12)
length(union(colnames(asvtab.menu.12), colnames(asvtab.protein)))
```

```{r}
# Helper function to pad out columns
pad_columns <- function(x, y, fill = 0){
     # Takes two matrices and synchronizes columns across them, filling added 
     # cols with a set value
     # Returns a list of the two updated matrices
     
     # Find missing columns in both matrices
     missing.x <- setdiff(colnames(y), colnames(x))
     missing.y <- setdiff(colnames(x), colnames(y))
     
     # Pad out columns of x
     fill.x <- matrix(fill, 
                      nrow = dim(x)[1],
                      ncol = length(missing.x))
     colnames(fill.x) <- missing.x 
     x <- cbind(x, fill.x)
     
     # Pad out columns of y
     fill.y <- matrix(fill, 
                      nrow = dim(y)[1],
                      ncol = length(missing.y))
     colnames(fill.y) <- missing.y
     y <- cbind(y, fill.y)
     
     # Arrange the columns so they appear in identical order
     u <- sort(union(colnames(x), colnames(y)))
     x <- x[, u]
     y <- y[, u]
     
     list(x, y)
}
```

```{r}
padded <- pad_columns(asvtab.protein, asvtab.menu.12)
asvtab.protein <- padded[[1]]
asvtab.menu.12 <- padded[[2]]

rm(padded)
```

```{r}
# Confirm row and column names are equal before proceeding
all(row.names(asvtab.protein) == row.names(asvtab.menu.12))
all(colnames(asvtab.protein) == colnames(asvtab.menu.12))
```

```{r}
# Label predictions 
protein.pos <- asvtab.protein > 4
menu.pos <- asvtab.menu.12 > 0

tp <- protein.pos & menu.pos
tn <- !protein.pos & !menu.pos
fp <- protein.pos & !menu.pos
fn <- !protein.pos & menu.pos

pred <- bind_rows(colSums(tp),
                  colSums(tn),
                  colSums(fp),
                  colSums(fn))

pred$prediction <- c('tp', 'tn', 'fp', 'fn')

pred <- select(pred, prediction, everything())
```

```{r}
# Reformat to long for visualization
pred.long <- pivot_longer(pred,
                          cols = -prediction,
                          names_to = 'taxon',
                          values_to = 'count')

# Factor data for visualization
# Calculate F-measure to be used downstream
pred.summary <- 
     pred.long %>% 
     pivot_wider(names_from = prediction, values_from = count) %>% 
     mutate(recall = tp/(tp + fn),
            precision = tp/(tp + fp),
            f_measure = (2 * recall * precision)/(recall + precision)) %>% 
     arrange(desc(f_measure), fn, fp)

# Assign label for faceting plot
# Done by manually inspecting pred.summary
breaks <- c(rep(1, 30), rep(2, 33), rep(3, 37), rep(4, 33))
pred.summary$facet <- breaks

pred.long <- 
     pred.summary %>% 
     select(taxon, facet) %>% 
     right_join(pred.long) %>% 
     select(facet, taxon, everything())

pred.long <- 
     pred.long %>% 
     mutate(taxon = factor(taxon, levels = pred.summary$taxon),
            prediction = factor(prediction, levels = c('tp', 'tn', 'fp', 'fn'),
                                labels = c('True positive', 'True negative',
                                           'False positive', 'False negative')))
```

```{r}
ggplot(pred.long, aes(x = taxon, y = count, fill = prediction)) +
     geom_bar(position = 'stack', stat = 'identity') +
     scale_fill_manual(values = c('#1A9641','#A6D96A', # TP, TN
                                  '#D7191C', '#FDAE61')) + # FP, FN
     facet_wrap(~facet, nrow = 4, scale = 'free_x') +
     theme_bw() +
     theme(axis.text.x = element_text(angle = 45, hjust = 1),
           axis.text.y = element_blank(),
           axis.title = element_blank(),
           legend.position = 'top',
           legend.title = element_blank(), 
           plot.margin = margin(0, 0.5, 0.5, 0.5, 'in'),
           strip.text.x = element_blank()) 

plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results', paste0(plotID, '_By-food predictions compared to menu, >4 PSMs per taxon filter.pdf')),
       height = 8, width = 8)
# 2021-02-22.677: First pass
# 2021-03-10.682: Re-colored
# 2021-04-09.985: Database v2, no PSM filter
# 2021-04-09.831: Database v2, >4 PSMs/taxon/sample filter
```



