---
title: "Estimating individualized transit times"
author: "Brianna Petrone"
output: html_notebook
---

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = file.path(getwd(), '..'))
```

```{r}
library(here)
library(phyloseq)
library(tidyverse)
```

## Read in files

### Metaproteomic phyloseq

```{r}
ps.protein <- readRDS(here('data', 'processed', 'phyloseq',
                           '20210212_ps_mp.rds'))
ps.protein
```

### DFC menu data phyloseq

```{r}
ps.menu <- readRDS(here('data', 'processed', 'phyloseq',
                        '20210220_ps_menu.rds'))

ps.menu
```

### Metabarcoding phyloseqs

#### trnL

```{r}
# This experiment contained PCRs amplified with an annealing temp of 55 or 63C; filter to only those at 63.
ps.trnL <- 
     readRDS(here('data', 'processed', 'phyloseq',
                        '20200225_ps_trnL.rds')) %>% 
     subset_samples(anneal_t == '63') %>% # Subset to 63C samples
     prune_taxa(taxa_sums(.) > 0, .) %>% # Remove taxa now with 0 reads
     subset_taxa(!is.na(superkingdom)) # Remove SVs not identified to a food

ps.trnL
```

```{r}
# Rename samples by subject_date for compatibility
sample_names(ps.trnL) <- paste(sample_data(ps.trnL)$subject,
                               sample_data(ps.trnL)$date, sep = '_')
```

#### 12SV5

```{r}
ps.12SV5 <- readRDS(here('data', 'processed', 'phyloseq',
                         '20201014_ps_12SV5.rds'))

ps.12SV5
```

## Find uniquely detected foods

```{r}
subjects <- unique(sample_data(ps.protein)$subj)
subjects
```

```{r}
# Helper function to find uniquely detected foods for a given subject
get_unique <- function(ps, subject){
     # Input: a phyloseq object and the subject of interest
     # Returns: a taxonomy table containing only those taxa detected in
     # a single sample
     
     # Subset to only the subject desired
     ps <- subset_samples(ps, subj == subject)
     # Get their ASV table
     asvtab <- otu_table(ps)@.Data

     uniques.protein <- names(which(colSums(asvtab > 0) == 1))
     uniques.protein
}
```

### Metaproteomic

```{r}
# For storing results
uniques.protein.protein <- data.frame()

# Find unique foods for each subject
for (subject in subjects){
     uniques.protein.protein <- 
          # Get uniques.protein
          data.frame(subj = subject, 
                     food = get_unique(ps.protein, subject)) %>% 
          # Add to results dataframe
          bind_rows(uniques.protein.protein, .)
}
```

```{r}
# Do any of these come up in all four subjects?
uniques.protein %>% 
     group_by(food) %>% 
     count() %>% 
     filter(n > 1)
```

```{r}
# Add PSM counts
melt.protein <- psmelt(ps.protein)

uniques.protein <- 
     melt.protein %>% 
     filter(Abundance != 0) %>% 
     select(subj, Sample, food = OTU, Abundance) %>% 
     right_join(uniques.protein, by = c('subj', 'food')) %>% 
     arrange(subj, desc(Abundance))
```

Follow-on thoughts: 
* How believable is the signal in each case? 
     + Is it abundant? (High number of PSMs)?
     + Is it backed up by menu data/trnL? (Menu data would be the necessary feature to get at transit time).
     
```{r}
# How to connect to menu data? Visually?
# Refactor food by count
foods.desc <- 
     uniques.protein %>% 
     group_by(food) %>% 
     summarize(tot.psms = sum(Abundance)) %>% 
     arrange(desc(tot.psms))

uniques.protein$food <- factor(uniques.protein$food,
                       levels = foods.desc$food)

rm(foods.desc)
```

```{r}
# Only foods with >5 counts, per input from Ali
uniques.protein %>% 
     filter(Abundance > 5) %>% 
     ggplot(aes(x = food, y = Abundance)) +
     geom_bar(stat = 'identity') +
     facet_wrap(~subj, scales = 'free') +
     labs(y = 'Total PSMs') + 
     theme_bw() +
     theme(axis.text.x = element_text(angle = 45,
                                      hjust = 1),
           axis.title.x = element_blank())
```

```{r}
# plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

# 2021-03-12.132

# ggsave(here('results', 'transit time', 
#             paste0(plotID, '_Unique foods per-participant, with PSMs.png')),
#        height = 7, width = 8)
```

### Metabarcoding

#### trnL

#### 12SV5

### Menu

```{r}
# Update ID to subj to be compatible with function
samdf.menu <- as(sample_data(ps.menu), 'data.frame')
samdf.menu <- rename(samdf.menu, subj = ID)

sample_data(ps.menu) <- samdf.menu
```

```{r}
# For storing results
uniques.protein.menu <- data.frame()

# Find unique foods for each subject
for (subject in subjects){
     uniques.protein.menu <- 
          # Get uniques.protein
          data.frame(subj = subject, 
                     food = get_unique(ps.menu, subject)) %>% 
          # Add to results dataframe
          bind_rows(uniques.protein.menu, .)
}
```

```{r}
# Add amount consumed (dry weight here)
melt.menu <- psmelt(ps.menu)

uniques.protein.menu <- 
     melt.menu %>% 
     filter(Abundance != 0) %>% 
     select(subj, Sample, food = OTU, Abundance) %>% 
     right_join(uniques.protein.menu, by = c('subj', 'food')) %>% 
     arrange(subj, desc(Abundance))
```

# Intersection

## Unique in metaproteomics, menu

