---
title: "Ordinations"
output: html_notebook
---

# R setup

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = file.path(getwd(), '..'))
```

```{r libraries, echo=FALSE}
library(here) 
library(phyloseq) 
library(tidyverse) 
```

# Read in data

## Shared samples

```{r}
# Read in and name comparably to phyloseqs (SUBJ_DATE)
samples <- 
     here('data', 'metadata', '20200723_MTA DFC samples shared.csv') %>% 
     read_csv() %>% 
     mutate(sample.date = as.Date(sample.date, format = '%m/%d/%y'),
            name = paste(subj, sample.date, sep = '_')) %>%
     pull(name)
               
samples %>% sort()
```

## Metaproteomic phyloseq

```{r}
ps.protein <- readRDS(here('data', 'processed', 'phyloseq',
                           # '20210408_ps_mp.rds')) # 1PUP analysis
                           # '20210912_ps_mp.rds')) # 1UP analysis
                           '20211102_ps_mp_joint.rds')) #1UP w/908 (1PUP)
ps.protein
```

## DFC menu data phyloseq

```{r}
ps.menu <- readRDS(here('data', 'processed', 'phyloseq',
                        '20210220_ps_menu.rds'))
ps.menu
```

## Metabarcoding phyloseq

```{r}
# Combined 12SV5 and trnL, made in combined MiniSeq notebook in DFC project folder
ps.mb <- 
     readRDS(here('data', 'processed', 'phyloseq',
                  '20210727_ps_mb.rds'))
     
ps.mb
```

```{r}
# Filter to only the samples under consideration here
ps.mb <- 
     ps.mb %>% 
     prune_samples(sample_names(.) %in% samples, .) %>% 
     prune_taxa(taxa_sums(.) > 0, .)

ps.mb
```

```{r}
# Subset to only those taxa identified to some position in the plant phylogeny
unlabeled <- 
     data.frame(ps.mb@tax_table) %>% 
     filter(is.na(superkingdom)) %>% 
     row.names()

asvtab.mb <- ps.mb@otu_table

cat('Percent of reads not assigned to a taxon:', 
    sum(asvtab.mb[, unlabeled])/sum(asvtab.mb) * 100)

ps.mb <- subset_taxa(ps.mb, !is.na(superkingdom))
```

# Pre-process

## Metaproteomic phyloseq

```{r}
# Impose >= 5 PSM/taxon filter
asvtab.protein <- otu_table(ps.protein)@.Data
sum(asvtab.protein)

asvtab.protein[asvtab.protein < 5] <- 0
sum(asvtab.protein)

# Replace in phyloseq object
otu_table(ps.protein) <- otu_table(asvtab.protein,
                                   taxa_are_rows = FALSE)
```

## Metabarcoding phyloseq

```{r}
# Impose >= 5 count/taxon filter
asvtab.mb <- otu_table(ps.mb)@.Data
sum(asvtab.mb)

asvtab.mb[asvtab.mb < 5] <- 0
sum(asvtab.mb)

# Replace in phyloseq object
otu_table(ps.mb) <- otu_table(asvtab.mb,
                                   taxa_are_rows = FALSE)
```

# Analyze

## Ordinations

### Metaproteomics

```{r}
ord.protein <- 
     ordinate(ps.protein,
              method = 'PCoA',
              distance = 'bray')
```

```{r}
p <-  
     plot_ordination(ps.protein,
                     ord.protein,
                     type = 'samples',
                     color = 'subj')
```

```{r}
p
```

```{r}
subj.colors <- 
     palette.colors(n = 6,
                    palette = 'Okabe-Ito') %>% 
     unname()
```


```{r}
# Customize plot 
data <- data.frame(ord.protein$vectors)
samdf.protein <- data.frame(ps.protein@sam_data)

data <- bind_cols(data, samdf.protein)

ggplot(data, aes(x = Axis.1, y = Axis.2, color = subj)) +
     geom_point(alpha = 0.8,
                size = 3) +
     labs(x = 'PCo1 (40.6%)', 
          y = 'PCo2 (11.4%)') +
     scale_color_manual(values = subj.colors) +
     coord_equal() +
     theme_bw() +
     theme(legend.title = element_blank()) +
     theme(axis.text = element_text(size = 12,
                                    face = 'bold',
                                    color = 'gray80'), 
           axis.title = element_text(size = 14,
                                    face = 'bold',
                                    color = 'gray80'), 
           axis.ticks = element_line(size = 2, 
                                     color = 'gray80'),
           legend.text = element_text(size = 10,
                                      face = 'bold',
                                      color = 'gray40'),
           panel.border = element_rect(size = 3,
                                       color = 'gray80'),
           panel.grid.major = element_blank())
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results',
            'manuscript',
            '2',
            paste0(plotID, '_Metaproteomic ordination.pdf')))
```

```{r}
vegan::adonis(distance(ps.protein, method = 'bray') ~ subj,
              data = samdf.protein)
```

### Metabarcoding



### Menu