---
title: "Combined phyloseq objects"
output: html_notebook
---

# Setup

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = file.path(getwd(), '..'))
```

```{r}
library(here)
library(phyloseq)
library(tidyverse)
```

# Raw data

## Menu

```{r}
menu.dfc <- 
     here('data', 
          'processed', 
          'phyloseq',
          'ps_menu_DFC.rds') %>% 
     readRDS()

menu.dfc
```

```{r}
menu.908 <- 
     here('data', 
          'processed', 
          'phyloseq',
          'ps_menu_908.rds') %>% 
     readRDS()

menu.908
```

## Metabarcoding

Note that this is the full dataset, whereas metaproteomics is only a subset of these samples.

```{r}
# DFC:
# 12SV5
mb.dfc.animal <- 
     here('data',
          'processed',
          'phyloseq',
          'ps_12SV5_DFC.rds') %>% 
     readRDS()

# trnL
mb.dfc.plant <- 
     here('data',
          'processed',
          'phyloseq',
          'ps_trnL_DFC.rds') %>% 
     readRDS()
     
mb.dfc.plant
mb.dfc.animal
```

```{r}
# Healthy donor
# 12SV5
mb.908.animal <- 
     here('data',
          'processed',
          'phyloseq',
          'ps_12SV5_908.rds') %>% 
     readRDS()

# trnL
mb.908.plant <- 
     here('data',
          'processed',
          'phyloseq',
          'ps_trnL_908.rds') %>% 
     readRDS()

mb.908.animal
mb.908.plant
```

## Metaproteomics

```{r}
# 1UP analysis, master proteins only, both datasets
mp.up <- 
     here('data',
          'processed',
          'phyloseq',
          'ps_mp_1UP.rds') %>%  # Unique accessions within species preserved
     readRDS()

mp.up
```

```{r}
# 1PUP analysis, master proteins only, both datasets
mp.pup <- 
     here('data',
          'processed',
          'phyloseq',
          'ps_mp_1PUP.rds') %>%  # Unique accessions within species preserved
     readRDS()

mp.pup
```

## Metadata

```{r}
# Sample collection times across the dataset
times <- 
     here('data',
          'metadata',
          'Sample collection times.csv') %>% 
     read_csv(col_types = cols(date = col_date("%m/%d/%y")))

head(times)
```

# Pre-process

## Menu

Note difference in entries in OTU table:
     - DFC: grams of intake
     - Healthy Donor: # of instances of intake
     
```{r}
head(sample_names(menu.dfc))
head(sample_names(menu.908))
```

```{r}
sample_variables(menu.dfc)
sample_variables(menu.908)
```

```{r}
# Correct difference in naming and state of variables (908 subj is character, DFC is factor)
sample_data(menu.dfc) <- 
     menu.dfc@sam_data %>% 
     data.frame() %>% 
     rename(subj = ID) %>% 
     mutate(subj = as.character(subj))

sample_data(menu.908) <- 
     menu.908@sam_data %>% 
     data.frame() %>% 
     rename(subj = ID,
            date = day)
```

```{r}
# Add study variable
sample_data(menu.dfc)$study <- 'Intervention'
sample_data(menu.908)$study <- 'Habitual Diet'
```

## Metaproteomics

Missing sample neds to be added after split by kingdom, otherwise it won't be preserved.

### Tidy

```{r}
head(sample_names(mp.pup))
head(sample_names(mp.up))
```

```{r}
sample_variables(mp.pup)
sample_variables(mp.up)
```

```{r}
# Add study in line with other datasets
sample_data(mp.pup)$study <- 
     ifelse(sample_data(mp.pup)$subj == '908',
            yes = 'Habitual Diet',
            no = 'Intervention')

sample_data(mp.up)$study <- 
     ifelse(sample_data(mp.up)$subj == '908',
            yes = 'Habitual Diet',
            no = 'Intervention')
```

```{r}
# Add sample time
sample_data(mp.pup) <- 
     mp.pup@sam_data %>% 
     data.frame() %>% 
     rownames_to_column(var = 'row') %>% 
     left_join(times) %>% 
     column_to_rownames(var = 'row')

sample_data(mp.up) <- 
     mp.up@sam_data %>% 
     data.frame() %>% 
     rownames_to_column(var = 'row') %>% 
     left_join(times) %>% 
     column_to_rownames(var = 'row')
```

## Metabarcoding

### Tidy metadata

Check overlap between variables between studies and markers
```{r}
setdiff(sample_variables(mb.dfc.animal),
        sample_variables(mb.908.animal)) %>% 
     sort()
```

```{r}
setdiff(sample_variables(mb.908.animal),
        sample_variables(mb.dfc.animal)) %>% 
     sort()
```

```{r}
## DFC 12SV5
# All are Power Soil tube extracted but are listed as "tube"; synchronize
sample_data(mb.dfc.animal) <- 
     mb.dfc.animal@sam_data %>% 
     data.frame() %>% 
     rownames_to_column(var = 'row') %>% 
     select(-c(extraction_type,
               well)) %>% 
     mutate(extraction_kit = 'PowerSoil tube',
            subj = as.character(subj)) %>% 
     left_join(times) %>% 
     column_to_rownames(var = 'row')
```

```{r}
# 908 12SV5
# Can remove the following variables
# Replicate (only one preserved), evaporation (not applicable to any of the samples kept)
sample_data(mb.908.animal) <- 
     mb.908.animal@sam_data %>% 
     data.frame() %>% 
     rownames_to_column(var = 'row') %>% 
     mutate(subj = as.character(subj)) %>% 
     select(-c(starts_with('replicate_'),
               evaporation,
               preprocess,
               reads)) %>%
     rename(reads_12SV5 = reads_animal) %>% 
     left_join(times) %>% 
     column_to_rownames(var = 'row')
```

```{r}
setdiff(sample_variables(mb.dfc.plant),
        sample_variables(mb.908.plant)) %>% 
     sort()
```

```{r}
setdiff(sample_variables(mb.908.plant),
        sample_variables(mb.dfc.plant)) %>% 
     sort()
```

```{r}
sample_data(mb.dfc.plant) <- 
     mb.dfc.plant@sam_data %>% 
     data.frame() %>% 
     rownames_to_column(var = 'row') %>% 
     select(-c(starts_with('extraction'),
               slurry, 
               aliquot,
               sample,
               well,
               is_neg,
               starts_with('CH_'))) %>% 
     mutate(extraction_kit = 'PowerSoil tube',
            subj = as.character(subj)) %>% 
     rename(original_order = orig_order,
            random_order = rand_order) %>% 
     left_join(times) %>% 
     column_to_rownames(var = 'row')
```

```{r}
## 908
# Can remove the following variables
# Replicate (only one preserved), evaporation (not applicable to any of the samples kept)
sample_data(mb.908.plant) <- 
     mb.908.plant@sam_data %>% 
     data.frame() %>% 
     rownames_to_column(var = 'row') %>% 
     mutate(subj = as.character(subj)) %>% 
     select(-c(starts_with('replicate_'),
               preprocess)) %>%
     rename(reads_trnL = reads_plant) %>% 
     left_join(times) %>% 
     column_to_rownames(var = 'row')
```

```{r}
# Add study in line with other datasets
sample_data(mb.dfc.animal)$study <- 'Intervention'
sample_data(mb.dfc.plant)$study <- 'Intervention'
sample_data(mb.908.animal)$study <- 'Habitual Diet'
sample_data(mb.908.plant)$study <- 'Habitual Diet'
```

### Subset samples

We sequenced all DFC samples, but only applied metaproteomics to a subset from consecutive days.
```{r}
length(intersect(sample_names(mp.up), 
                 c(sample_names(mb.dfc.plant), sample_names(mb.908.plant))))

nsamples(mb.dfc.plant) + nsamples(mb.908.plant)
```

```{r}
# Prune only needs to be done for DFC (using all 5 908 samples)
mb.dfc.plant <- 
     prune_samples(sample_names(mp.up),
                   mb.dfc.plant) %>% 
     prune_taxa(taxa_sums(.) > 0, .)

mb.dfc.animal <- 
     prune_samples(sample_names(mp.up),
                   mb.dfc.animal) %>% 
     prune_taxa(taxa_sums(.) > 0, .)

mb.dfc.plant
mb.dfc.animal
```

# Combine, refactor subject

## Menu

```{r}
nsamples(menu.908)
nsamples(menu.dfc)
```

```{r}
menu <- 
     merge_phyloseq(menu.908,
                    menu.dfc)

menu
```

```{r}
sample_data(menu)$subj <- 
     factor(sample_data(menu)$subj,
            labels = c('HD', 'ID1', 'ID2', 'ID3', 'ID4'))
```

## Metabarcoding

```{r}
mb.plant <- 
     merge_phyloseq(mb.908.plant,
                    mb.dfc.plant)

mb.plant
```

```{r}
sample_data(mb.plant)$subj <- 
     factor(sample_data(mb.plant)$subj,
            labels = c('HD', 'ID1', 'ID2', 'ID3', 'ID4'))
```

```{r}
mb.animal <- 
     merge_phyloseq(mb.908.animal,
                    mb.dfc.animal)

mb.animal
```

```{r}
sample_data(mb.animal)$subj <- 
     factor(sample_data(mb.animal)$subj,
            labels = c('HD', 'ID1', 'ID2', 'ID3', 'ID4'))
```

### Update assignment

```{r}
ref.trnL <- 
     here('..',
          'food-dbs',
          'data', 
          'processed', 
          'dada2-compatible',
          'trnL',
          'trnLGH.fasta')

ref.12SV5 <- 
     here('..',
          'food-dbs',
          'data', 
          'processed', 
          'dada2-compatible',
          '12SV5',
          '12SV5_taxonomy.fasta')
```

```{r}
mb.plant <- MButils::update_assignment(mb.plant,
                                       ref.trnL,
                                       use_function = 'species')

mb.animal <- MButils::update_assignment(mb.animal,
                                        ref.12SV5,
                                        use_function = 'taxonomy')
```

### Glom taxa

#### 12SV5

```{r}
# Get lowest level name
taxtab.animal <- 
     mb.animal@tax_table@.Data %>% 
     data.frame() %>% 
     MButils::lowest_level()
```

```{r}
dups <- taxtab.animal$name[duplicated(taxtab.animal$name)]
```

```{r}
sort(dups)
```

```{r}
# Replace taxonomy table in phyloseq object
# Column indexing puts name on leftmost side
tax_table(mb.animal) <- as.matrix(taxtab.animal)

# Now, these and all others can be merged by name
mb.animal <- tax_glom(mb.animal, 
                      taxrank = 'name',
                      NArm = FALSE)

mb.animal
```

#### trnL

```{r}
# Get lowest level name
taxtab.plant <- 
     mb.plant@tax_table@.Data %>% 
     data.frame() %>% 
     MButils::lowest_level()
```

```{r}
dups <- taxtab.plant$name[duplicated(taxtab.plant$name)]
```

```{r}
sort(dups)
```

This is a bit fraught at higher levels, because depending on exact ASVs these families may actually map to different species. 

Procedure for these is to run the below:
```{r eval = FALSE}
# taxtab.plant %>%
#      filter(name == 'NAME') %>%
#      row.names()

# Test if they're subsequences
# grepl(asv1, asv2, fixed = TRUE)
```

```{r}
# Update names at higher levels to reflect distinction

# Brassica oleracea:
taxtab.plant['ATCCTGGGTTACGCGAACAAAACAGAGTTTAGAAAGCGG',
       'name'] <-
     'Brassica oleracea (cabbage, broccoli, brussels sprouts, kohlrabi)'

taxtab.plant['ATCCTGGGTTACGCGAACAAACCAAAGTTTAGAAAGCGG',
       'name'] <-
     'Brassica oleracea (gai lan)'

taxtab.plant['ATCCTAGGTTACGCGAACAAACCAAAGTTTAGAAAGCGG',
       'name'] <-
     'Brassica oleracea (broccoli)'

# # Lamiaceae
taxtab.plant['ATCCTGTTTTCTCAAAACAAAAGTTCAAAAAACGAAAAAAAAAAAAAG',
       'name'] <-
     'Lamiaceae (basil, holy basil, lemon basil, catnip)'

taxtab.plant['ATCCTGTTTTCTCAAAACAAAGGTTCAAAAAACGAAAAAAAAAG',
       'name'] <-
     'Lamiaceae (rosemary, lemon balm)'

# Rosaceae
taxtab.plant['ATCCCGTTTTATGAAAACAAACAAGGGTTTCAGAAAGCGAGAATAAATAAAG',
       'name'] <-
     'Rosaceae (arctic raspberry, strawberry)'

taxtab.plant['ATCCTGTTTTATGAAAATAAACAAGGGTTTCATAAACCGAAAATAAAAAAG',
       'name'] <-
     'Rosaceae (apple, pear)'

# Oryza sativa, Solanum lycopersicum can be merged
```

```{r}
# Replace taxonomy table in phyloseq object
# Column indexing puts name on leftmost side
tax_table(mb.plant) <- as.matrix(taxtab.plant)

# # Now, these and all others can be merged by name
mb.plant <- tax_glom(mb.plant, 
                     taxrank = 'name',
                     NArm = FALSE)

mb.plant
```

## Metaproteomic

```{r}
sample_data(mp.pup)$subj <- 
     factor(sample_data(mp.pup)$subj,
            labels = c('HD', 'ID1', 'ID2', 'ID3', 'ID4'))
```

```{r}
sample_data(mp.up)$subj <- 
     factor(sample_data(mp.up)$subj,
            labels = c('HD', 'ID1', 'ID2', 'ID3', 'ID4'))
```

# Save

```{r}
# Menu
saveRDS(menu,
        here('data',
             'processed',
             'phyloseq',
             'combined',
             'ps_menu.rds')
)

# Metabarcoding
saveRDS(mb.plant,
        here('data',
             'processed',
             'phyloseq',
             'combined',
             'ps_mb_trnL.rds') 
)

saveRDS(mb.animal,
        here('data',
             'processed',
             'phyloseq',
             'combined',
             'ps_mb_12SV5.rds') 
)

# Metaproteomic
saveRDS(mp.pup,
        here('data',
             'processed',
             'phyloseq',
             'combined',
             'ps_mp_1PUP.rds')
)

saveRDS(mp.up,
        here('data',
             'processed',
             'phyloseq',
             'combined',
             'ps_mp_1UP.rds')
)
```
