---
title: "Data description"
output: html_notebook
---

# Setup

```{r setup}
require(knitr)
knitr::opts_knit$set(root.dir = file.path(getwd(), '..'))
```

```{r}
library(Biostrings)
library(cowplot)
library(eulerr) # For Euler plots
library(here)
library(phyloseq)
library(tidyverse)
library(scales) # For comma in plot labels
```

```{r}
# Plotting themes
theme_set(theme_bw() +
               theme(
                    axis.text = element_text(size = 12),
                    axis.title.x = ggtext::element_markdown(size = 14,
                                                            face = 'bold'),
                    axis.title.y = ggtext::element_markdown(size = 14,
                                                            face = 'bold'),
                    legend.title = element_text(size = 12,
                                                face = 'bold'),
                    strip.text = element_text(size = 12,
                                              face = 'bold')
               )
)
```

```{r}
study.color <- c('#4e79a7', # Intervention
                 '#f28e2b') # Habitual Diet
```

# Read in data

## Taxon name mapping

## Metabarcoding
### trnL
```{r}
ps.mb.plant <- 
     here('data',
          'processed',
          'phyloseq',
          'combined',
          'filtered',
          'ps_mb_trnL.rds') %>% 
     readRDS()

ps.mb.plant
```

### 12SV5
```{r}
ps.mb.animal <- 
     here('data',
          'processed',
          'phyloseq',
          'combined',
          'filtered',
          'ps_mb_12SV5.rds') %>% 
     readRDS()

ps.mb.animal
```

## Metaproteomics

Filtering steps done in notebook "Filtered phyloseq objects":
- Subset animal proteins to only those certainly of muscle, dairy, or egg origin
- Filter taxa identified to only those with cumulatively >= 5 PSMs

```{r}
# Tax-glommed
ps.protein.taxon <- 
     here('data',
          'processed',
          'phyloseq',
          'combined',
          'filtered',
          'ps_mp_taxa.rds') %>%
     readRDS()

# By protein
ps.protein <- 
     here('data',
          'processed',
          'phyloseq',
          'combined',
          'filtered',
          'ps_mp_proteins.rds') %>%
     readRDS()
```

```{r}
# Unfiltered data
proteins.df <- 
     here('data',
          'processed',
          '5FDR_1UP_both datasets.rds') %>% 
     readRDS()
```

## Menu

```{r}
ps.menu <- 
     here('data', 
          'processed', 
          'phyloseq',
          'combined',
          'filtered',
          'ps_menu.rds') %>% 
     readRDS()
```

## References

### Metaproteomic database

```{r }
# Read in the first sheet of the file using read_excel function
db <- readxl::read_excel(here('data', 
                              'processed',
                              'DietaryDB_ForDavidLab.xlsx'), 
                         sheet = 1, 
                         n_max = 547) # Exclude annotations at end of file
head(db)
```

### trnL

```{r}
ref.trnL <- 
     here('..',
          'food-dbs',
          'data',
          'processed',
          'dada2-compatible',
          'trnL',
          'trnLGH.fasta') %>% 
     readDNAStringSet()

ref.trnL
```

### 12SV5

```{r}
ref.12SV5 <- 
     here('..',
          'food-dbs',
          'data',
          'processed',
          'dada2-compatible',
          '12SV5',
          '12SV5_taxonomy.fasta') %>% 
     readDNAStringSet()

ref.12SV5
```

# Pre-process

## Metaproteomic reference

```{r }
# Manual updates
# Specify scientific name for tilapia
db$scientific_name[db$scientific_name == 'Tilapia'] <- 
     'Oreochromis niloticus'
# Correct misspelling for mallard duck
db$scientific_name[db$scientific_name == 'Anas platyrhynchose'] <- 
     'Anas platyrhynchos'
# Switch var. to subsp. for durum
db$scientific_name[db$scientific_name == 'Triticum turgidum var. durum'] <-
     'Triticum turgidum subsp. durum'
db$scientific_name[db$scientific_name == 'Triticum dicoccon'] <- 
     'Triticum dicoccum'
# Correct spelling of water buffalo
db$scientific_name[db$scientific_name == 'Bubalis bubalis'] <- 
     'Bubalus bubalis'
# Think something weird about formatting of "x" in Musa x paradisiaca, not being recognized
db$scientific_name[db$common_name == 'plantain'] <- 
     'Musa x paradisiaca'
# Same for Mentha x piperita
db$scientific_name[db$common_name == 'peppermint'] <- 
     'Mentha x piperita'
# Add hybrid designation for grapefruit
db$scientific_name[db$scientific_name == 'Citrus paradisi'] <- 
     'Citrus x paradisi'
# Remove zante currant, which is redundant with V. vinifera
db <- filter(db, common_name != 'zante currant')
db$scientific_name[db$scientific_name == 'Mercenaria mercenaria, others'] <-
     'Mercenaria mercenaria'
db$scientific_name[db$scientific_name == 'Polystichum munitum and others'] <-
     'Polystichum munitum'
# Query resulted in hits for var. capitata only
db$scientific_name[db$scientific_name == 'Brassica oleracea var. capitata f. alba'] <- 
     'Brassica oleracea var. capitata'
# Switch fennel to its synonym used by NCBI taxonomy
db$scientific_name[db$scientific_name == 'Foeniculum vulgare'] <- 
     'Anethum foeniculum'
# Correct name for dragonfruit
db$scientific_name[db$scientific_name == 'Hylocereus undatus'] <- 
     'Selenicereus undatus'
# Update name for nori
db$scientific_name[db$scientific_name == 'Pyropia yezoensis'] <- 
     'Neopyropia yezoensis'
# Subspecies designation for maple (syrup)
db$scientific_name[db$scientific_name == 'Acer nigrum'] <- 
     'Acer saccharum subsp. nigrum'
``` 

## Metaproteomic

### Separate by kingdom

```{r}
ps.protein.plant <- subset_taxa(ps.protein.taxon, 
                                kingdom == 'Viridiplantae')

ps.protein.animal <- subset_taxa(ps.protein.taxon, 
                                 kingdom == 'Metazoa')

ps.protein.other <- subset_taxa(ps.protein.taxon, 
                                !(kingdom %in% c('Viridiplantae', 'Metazoa')))
```

## Menu

### Separate by kingdom

```{r}
ps.menu.plant <- subset_taxa(ps.menu, kingdom == 'Viridiplantae')
ps.menu.animal <- subset_taxa(ps.menu, kingdom == 'Metazoa')
ps.menu.other <- subset_taxa(ps.menu, !(kingdom %in% c('Viridiplantae',
                                                       'Metazoa')))
```

# Describe

```{r}
melt.mp <- psmelt(ps.protein)
melt.mb.plant <- psmelt(ps.mb.plant)
melt.mb.animal <- psmelt(ps.mb.animal)
```

## Success-failure

```{r}
# Metabarcoding: trnL
melt.mb.plant %>% 
     group_by(Sample) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     group_by(reads_tot > 0) %>% 
     count()
```

```{r}
# Metabarcoding: 12SV5
melt.mb.animal %>% 
     group_by(Sample) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     group_by(reads_tot > 0) %>% 
     count()
```
```{r}
# Metaproteomic
melt.mp %>% 
     group_by(Sample) %>% 
     summarize(psms_tot = sum(Abundance)) %>% 
     group_by(psms_tot > 0) %>% 
     count()
```
## Counts per sample

### Metaproteomic spectra

```{r}
melt.mp %>% 
     group_by(Sample, study) %>% 
     summarize(psms_tot = sum(Abundance)) %>% 
     filter(psms_tot > 0) %>% 
     ggplot(aes(x = psms_tot, fill = study)) +
     geom_histogram(stat = 'bin',
                    binwidth = 1000,
                    boundary = 0) +
     scale_y_continuous(breaks = seq(2, 10, by = 2),
                        limits = c(0, 10)) +
     scale_x_continuous(labels = comma) +
     scale_fill_manual(values = study.color) +
     labs(x = 'PSMs', y = 'Samples (*n*)') +
     theme(legend.position = 'none',
           panel.grid.minor = element_blank())
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results',
            'manuscript',
            'supplemental',
            'description',
            paste0(plotID, '_PSM histogram.pdf')),
       height = 3, width = 3)
```

```{r}
melt.mp %>% 
     group_by(Sample) %>% 
     summarize(psms_tot = sum(Abundance)) %>% 
     filter(psms_tot > 0) %>% 
     summarize(median(psms_tot),
               range(psms_tot))
```

### Metabarcoding reads

#### trnL

```{r}
melt.mb.plant %>% 
     group_by(Sample, study) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     filter(reads_tot > 0) %>% 
     ggplot(aes(x = reads_tot, fill = study)) +
     geom_histogram(stat = 'bin',
                    binwidth = 5000,
                    boundary = 0) +
     scale_y_continuous(breaks = seq(2, 10, by = 2),
                        limits = c(0, 10)) +
     scale_x_continuous(labels = comma,
                        breaks = seq(0, 90000, 30000)) +
     scale_fill_manual(values = study.color) +
     labs(x = '*trnL* reads', y = 'Samples (*n*)') +
     theme(legend.position = 'none',
           panel.grid.minor = element_blank())
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results',
            'manuscript',
            'supplemental',
            'description',
            paste0(plotID, '_trnL read histogram.pdf')),
       height = 3, width = 3)
```

```{r}
melt.mb.plant %>% 
     group_by(Sample) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     filter(reads_tot > 0) %>% 
     summarize(median(reads_tot),
               range(reads_tot))
```
#### 12SV5

```{r}
melt.mb.animal %>% 
     group_by(Sample, study) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     filter(reads_tot > 0) %>% 
     ggplot(aes(x = reads_tot, fill = study)) +
     geom_histogram(stat = 'bin',
                    binwidth = 5000,
                    boundary = 0) +
     scale_y_continuous(breaks = seq(2, 10, by = 2),
                        limits = c(0, 10)) +
     scale_x_continuous(labels = comma,
                        limits = c(0, 100000),
                        breaks = seq(0, 90000, 30000)) +
     scale_fill_manual(values = study.color) +
     labs(x = '12SV5 reads', y = 'Samples (*n*)') +
     theme(legend.position = 'none',
           panel.grid.minor = element_blank())
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results',
            'manuscript',
            'supplemental',
            'description',
            paste0(plotID, '_12SV5 read histogram.pdf')),
       height = 3, width = 3)
```

```{r}
melt.mb.animal %>% 
     group_by(Sample) %>% 
     summarize(reads_tot = sum(Abundance)) %>% 
     filter(reads_tot > 0) %>% 
     summarize(median(reads_tot),
               range(reads_tot))
```

## # reference foods

### Metaproteomics

TODO: Should likely go back and more accurately map these names to species based on entries in the "Notes" column

```{r}
db %>% 
     group_by(category) %>% 
     summarize(n_distinct(scientific_name))
```
plant* is Rhodophyta (plant)
NA is hake (animal)

### Metabarcoding

#### 12SV5 

Can just leverage taxonomy here to split apart and reduce
```{r}
# Number of sequences:
length(ref.12SV5)
```

```{r}
ref.12SV5 <- 
     ref.12SV5 %>% 
     names() %>% 
     data.frame(name = .) %>% 
     # Trim terminal semicolon
     mutate(name = gsub(pattern = ';$', 
                        replacement = '',
                        name)) %>% 
     separate(name,
              into = c('kingdom',
                       'phylum',
                       'class',
                       'order',
                       'family',
                       'genus',
                       'species',
                       'subspecies'),
              sep = ';',
              remove = FALSE) %>% 
     MButils::lowest_level()

ref.12SV5.db
```
```{r}
# Number of animals
# Note one of these is human: exclude
n_distinct(ref.12SV5.db$name) - 1
```

#### trnL

```{r}
# Number of sequences:
length(ref.trnL)
```

```{r}
# Number of plants
ref.trnL <- 
     ref.trnL %>% 
     names() %>% 
     data.frame(name = .) %>% 
     separate(name,
              into = c('accession', 
                       'name'),
              sep = ' ',
              extra = 'merge')
```

```{r}
n_distinct(ref.trnL$name)
```

## Detected foods

### Metaproteomics

```{r}
ntaxa(ps.protein.taxon)
```

What about number of unique dietary peptides?  Use intermediate processing stage of data for this (saved in "Metaproteomic phyloseq" notebook):

```{r}
dim(proteins.df)
```

```{r}
n_distinct(proteins.df$Accession)
n_distinct(proteins.df$Description)
```

```{r}
# Most prevalent foods
melt.mp %>% 
     group_by(Sample, kingdom, name) %>% 
     summarize(psms_tot = sum(Abundance)) %>% 
     mutate(prevalence_within = psms_tot >= 5) %>% 
     # Now group by just food, get across-sample prevalence
     group_by(kingdom, name) %>% 
     summarize(prevalence_across = sum(prevalence_within)) %>% 
     arrange(desc(prevalence_across)) %>% 
     filter(kingdom == 'Metazoa')
```

Interesting: there are still some that trigger questions for me here. For example, goat had previously been down to 4 detections, maybe I am being too permissive?  Check which proteins are listed and maybe flag??

```{r}
proteins.animal <- 
     proteins.df %>% 
     select(Accession, Description) %>% 
     mutate(word = gsub(Description,
                        pattern = ' OS=.*',
                        replacement = '')) %>% 
     distinct() %>% 
     right_join(melt.mp, by = c('Accession'='OTU'))
```

```{r}
proteins.animal %>% 
     filter(name == 'Phoenix dactylifera') %>% 
     select(word, Abundance, Sample) %>% 
     arrange(desc(Abundance))
```
```{r}
# Most prevalent foods
melt.mp %>% 
     filter(kingdom == 'Viridiplantae') %>% 
     group_by(Sample, name) %>%
     summarize(psms_tot = sum(Abundance)) %>% 
     mutate(prevalence_within = psms_tot >= 5) %>% 
     # Now group by just food, get across-sample prevalence
     group_by(name) %>% 
     summarize(prevalence_across = sum(prevalence_within)) %>% 
     arrange(desc(prevalence_across))
```

Oh wow-- with what proportion of total PSMs?

```{r}
all.plants <- c(
     'Arachis hypogaea',
     'Glycine max',
     'Helianthus annuus',
     'Lactuca sativa',
     'Malus domestica',
     'Phoenix dactylifera',
     'Theobroma cacao',
     'Triticum aestivum',
     'Triticum turgidum subsp. durum',
     'Zea mays'
)

all.animals <- c(
     'Bos taurus',
     'Capra hircus',
     'Gallus gallus',
     'Meleagris gallopavo',
     'Oncorhynchus mykiss',
     'Salmo salar',
     'Siluriformes',
     'Sus scrofa'
)

melt.protein.animal <- psmelt(ps.protein.animal)
melt.protein.plant <- psmelt(ps.protein.plant)
```

```{r}
# Write loop to examine
# Same plot ID for all
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

for (plant in all.plants){
     p <- 
          melt.protein.plant %>% 
          group_by(is_plant = OTU == plant,
                   Sample) %>% 
          summarize(Abundance = sum(Abundance)) %>% 
          ggplot(aes(x = Sample,
                     y = Abundance,
                     fill = is_plant)) +
          geom_bar(stat = 'identity',
                   position = 'fill') +
          labs(title = plant) +
          scale_fill_manual(values = c('gray80',
                                       '#638b66')) +
          theme(panel.grid = element_blank(),
                legend.position = 'none',
                axis.text.x = element_blank(),
                axis.title.x = element_blank(),
                plot.title = element_text(face = 'bold.italic'))
     
     ggsave(here('results',
                 'manuscript',
                 'tbd',
                 paste0(plotID, 
                        '_',
                        plant,
                        ' PSMs across all metaproteomic samples.pdf')),
            plot = p,
            height = 1.5, width = 4)
}
```

```{r}
# Write loop to examine
# Same plot ID for all
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

for (animal in all.animals){
     p <- 
          melt.protein.animal %>% 
          group_by(is_animal = OTU == animal,
                   Sample) %>% 
          summarize(Abundance = sum(Abundance)) %>% 
          ggplot(aes(x = Sample,
                     y = Abundance,
                     fill = is_animal)) +
          geom_bar(stat = 'identity',
                   position = 'fill') +
          labs(title = animal) +
          scale_fill_manual(values = c('gray80',
                                       '#b66353')) +
          theme(panel.grid = element_blank(),
                legend.position = 'none',
                axis.text.x = element_blank(),
                axis.title.x = element_blank(),
                plot.title = element_text(face = 'bold.italic'))
     
     ggsave(here('results',
                 'manuscript',
                 'tbd',
                 paste0(plotID, 
                        '_',
                        animal,
                        ' PSMs across all metaproteomic samples.pdf')),
            plot = p,
            height = 1.5, width = 4)
}
```

```{r}
# Pick one for follow up-- look at proteins
proteins.df %>% 
     filter(scientific_name == 'Capra hircus') %>% 
     select(`# PSMs`, Description) %>% 
     arrange(desc(`# PSMs`)) %>% 
     View()
```

This seems like a nice example:
- carboxypeptidase (digestive-- thinking cross-reactive)
- myosins/myoglobins/microtubule
- others

```{r}
goat <- 
     proteins.df %>% 
     filter(scientific_name == 'Capra hircus') %>%
     mutate(Sample = paste(subj,
                           date,
                           sep = '_')) %>%      
     select(Sample, Description, `# PSMs`) %>% 
     mutate(category = case_when(
          grepl(pattern = 'myo',
                ignore.case = TRUE, 
                Description) ~ 'muscle',
          grepl(pattern = 'carboxypeptidase|progastricsin',
                ignore.case = TRUE, 
                Description) ~ 'digestive',
          TRUE ~ 'other'
     ))
```

```{r}
goat %>% 
     group_by(Sample, category) %>% 
     summarize(Abundance = sum(`# PSMs`)) %>% 
     ggplot(aes(x = Sample,
                y = Abundance,
                fill = category)) +
     geom_bar(stat = 'identity',
              position = 'fill') +
     scale_fill_manual(values =
                            c('gray80',
                              '#b66353',
                              '#fbb04e')) +
     theme(panel.grid.minor = element_blank(),
           panel.grid.major.x = element_blank(),
           axis.text.x = element_blank(),
           legend.title = element_blank())
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results',
            'manuscript',
            'tbd',
            paste0(plotID, 
                   '_',
                   'Capra hircus PSMs by category.pdf')),
       height = 1.5, width = 5)
```

```{r}
goat %>% 
     group_by(Sample, category) %>% 
     summarize(Abundance = sum(`# PSMs`)) %>% 
     filter(category == 'muscle' & Abundance >= 5)
```

```{r}
# Most abundant foods
melt.mp %>% 
     filter(Abundance > 5) %>% 
     group_by(OTU) %>% 
     summarize(abundance = sum(Abundance)) %>% 
     arrange(desc(abundance))
```

### Metabarcoding

```{r}
# trnL
ps.mb.plant
```

```{r}
# Most prevalent foods
melt.mb.plant %>% 
     group_by(Sample, name) %>% 
     summarize(prevalence_within = sum(Abundance > 0)) %>% 
     # Now group by just food, get across-sample prevalence
     group_by(name) %>% 
     summarize(prevalence_across = sum(prevalence_within)) %>% 
     arrange(desc(prevalence_across))
```

```{r}
# Most abundant foods
melt.mb.plant %>% 
     group_by(name) %>% 
     summarize(abundance = sum(Abundance)) %>% 
     arrange(desc(abundance))
```

```{r}
# 12SV5
ps.mb.animal
```

```{r}
# Most prevalent foods
melt.mb.animal %>% 
     group_by(Sample, name) %>% 
     summarize(prevalence_within = sum(Abundance > 0)) %>% 
     # Now group by just food, get across-sample prevalence
     group_by(name) %>% 
     summarize(prevalence_across = sum(prevalence_within)) %>% 
     arrange(desc(prevalence_across))
```

### Intersection

What species are shared between metaproteomic, metabarcoding, and (day before sample) menu datasets?

#### Filter to only successful samples in all 3

```{r}
ps.mb.plant
ps.mb.plant <- 
     ps.mb.plant %>% 
     prune_samples(sample_sums(.) > 0, .) %>% 
     prune_taxa(taxa_sums(.) > 0, .)

ps.mb.plant
```

No similar pruning necessary for animal metabarcoding (all successful) or metaproteomic dataset (failed sample falls off when split kingdom-wise). 

```{r}
samples.plant <- 
     intersect(sample_names(ps.mb.plant),
               sample_names(ps.protein))

length(samples.plant)
```

```{r}
samples.animal <- 
     intersect(sample_names(ps.mb.animal),
               sample_names(ps.protein))

length(samples.animal)
```

```{r}
subset_ps <- function(ps, samples){
     ps %>% 
          prune_samples(samples, .) %>% 
          prune_taxa(taxa_sums(.) > 0, .)
}

# Now, subset each dataset (except menu)
ps.mb.plant <- subset_ps(ps.mb.plant, samples.plant)
ps.protein.plant <- 
     ps.protein.taxon %>% 
     subset_taxa(kingdom == 'Viridiplantae') %>% 
     subset_ps(., samples.plant)

ps.mb.animal <- subset_ps(ps.mb.animal, samples.animal)
ps.protein.animal <- 
     ps.protein.taxon %>% 
     subset_taxa(kingdom == 'Metazoa') %>% 
     subset_ps(., samples.animal)
```

```{r}
# Plants
# Get union of taxa detected in the two analyses
all.plants <- 
     union(taxa_names(ps.mb.plant),
           taxa_names(ps.protein.plant)) %>% 
     union(taxa_names(ps.menu.plant)) 

length(all.plants)
```

```{r}
# Animals
all.animals <- 
     union(taxa_names(ps.mb.animal),
           taxa_names(ps.protein.animal)) %>% 
     union(taxa_names(ps.menu.animal)) 

length(all.animals)
```

```{r}
# Other
all.other <- 
     union(taxa_names(ps.menu.other),
           taxa_names(ps.protein.other))

length(all.other)
```

```{r}
# Make dataframe for plot input
# Plants
plant.df <- 
     data.frame(taxon = all.plants) %>% 
     mutate(dna = taxon %in% taxa_names(ps.mb.plant),
            protein = taxon %in% taxa_names(ps.protein.plant),
            menu = taxon %in% taxa_names(ps.menu.plant))

# Quick checks
sum(plant.df$dna) 
sum(plant.df$protein) 
sum(plant.df$menu) 
```

```{r}
# Make dataframe for plot input
# Animals
animal.df <- 
     data.frame(taxon = all.animals) %>% 
     mutate(dna = taxon %in% taxa_names(ps.mb.animal),
            protein = taxon %in% taxa_names(ps.protein.animal),
            menu = taxon %in% taxa_names(ps.menu.animal))

# Quick checks
sum(animal.df$dna) 
sum(animal.df$protein) 
sum(animal.df$menu) 
```

```{r}
# Make dataframe for plot input
# Other
other.df <- 
     data.frame(taxon = all.other) %>% 
     mutate(dna = FALSE,
            protein = taxon %in% taxa_names(ps.protein.other),
            menu = taxon %in% taxa_names(ps.menu.other))

# Quick checks
sum(other.df$dna) 
sum(other.df$protein) 
sum(other.df$menu) 
```

```{r}
a <- 
     euler(plant.df[,c('dna', 'protein', 'menu')]) %>% 
     plot(
          fills = c('#fbb04e', '#849db1', '#7e756d'),
          labels = TRUE,
          shape = "ellipse", 
          quantities = TRUE
          )

b <- 
     euler(animal.df[,c('dna', 'protein', 'menu')]) %>% 
     plot(
          fills = c('#fbb04e', '#849db1', '#7e756d'),
          labels = FALSE,
          shape = "ellipse", 
          quantities = TRUE
     )

c <- 
     euler(other.df[,c('dna', 'protein', 'menu')]) %>% 
     plot(
          fills = c('#fbb04e', '#849db1', '#7e756d'),
          labels = FALSE,
          shape = "ellipse", 
          quantities = TRUE
          )
```

```{r}
# Combined plot
plot_grid(a, b, c, 
          labels = c('a', 'b', 'c'), 
          label_size = 12,
          nrow = 1,
          rel_widths = c(6, 4, 2)) + 
     draw_plot_label(label = c('Plant', 'Animal', 'Other'),
                     x = c(0.21, 0.53, 0.85),
                     y = c(0, 0.17, 0.32),
                     size = 12) +
     theme(plot.margin = unit(c(0.5, 0.5, 1, 0.5), "cm")) 
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
ggsave(here('results',
            'manuscript',
            'tbd',
            paste0(plotID, 
                   '_Plant taxa Euler diagram across methods, with counts, menu lag of 1-2 days.pdf')),
       height = 4, width = 8)
```


## Per-sample details

Overall view:
```{r echo = FALSE}
# How many food taxa consumed daily (menu)?
melt.menu <- psmelt(ps.menu)
counts.menu <- 
     melt.menu %>% 
     filter(Abundance > 0) %>% 
     group_by(subj, date, kingdom) %>% 
     count(name = 'menu_taxa')
```

```{r echo = FALSE}
# How many food taxa detected daily (metaproteomics)?
melt.protein <- psmelt(ps.protein)
counts.protein <- 
     melt.protein %>% 
     filter(Abundance > 4) %>% 
     group_by(subj, date, kingdom) %>% 
     count(name = 'protein_taxa')
```

```{r}
# Note some NA values here: what are they?
melt.protein %>% 
     filter(Abundance > 4) %>% 
     filter(is.na(kingdom)) %>% 
     pull(OTU) %>% 
     unique()
```
If one defines the kingdom Plantae to mean the Archaeplastida, the red algae will be part of that kingdom.
If Plantae are defined more narrowly, to be the Viridiplantae, then the red algae might be considered their own kingdom, or part of the kingdom Protista.

```{r echo = FALSE}
# How many food taxa detected daily (metabarcoding)?
melt.mb <- psmelt(
     merge_phyloseq(ps.mb.plant,
                    ps.mb.animal)
)
counts.mb <- 
     melt.mb %>% 
     filter(Abundance > 4) %>% 
     group_by(subj, date, phylum) %>% 
     count(name = 'dna_taxa')
```

```{r}
# Do this for only samples in common
# Though need to preserve Fungi, which aren't measured by metabarcoding
# This can be done by doing a full join
counts <- 
     counts.protein %>% 
     full_join(counts.mb) %>% 
     filter(!is.na(kingdom)) %>% 
     mutate(delta1 = date - 1) %>% 
     # Join to menu, lagged 1 day 
     # Full join to add kingdom if necessary
     full_join(counts.menu, by = c('kingdom',
                                   'subj', 
                                   'delta1' = 'date')) %>% 
     # Then remove elements that didn't join by the date
     filter(!is.na(date))

# Long version for plotting
counts.long <- pivot_longer(counts, 
                            cols = ends_with('taxa'),
                            names_to = 'dataset',
                            values_to = 'value')
```

```{r}
# Update names for plotting
counts.long$dataset <- factor(counts.long$dataset, 
                                     levels = c('menu_taxa', 
                                                'protein_taxa',
                                                'dna_taxa'), 
                                     labels = c('Recorded menu',
                                                'Metaproteomics',
                                                'Metabarcoding'))
# Refactor for plotting
counts.long$kingdom <- factor(counts.long$kingdom,
                              levels = c('Viridiplantae',
                                         'Metazoa', 
                                         'Fungi'))
```

```{r}
ggplot(counts.long, aes(x = kingdom, y = value, fill = dataset)) +
     geom_boxplot() +
     labs(title = 'Number of food taxa identified by measure',
          subtitle = '>4 PSMs per taxon filter',
          x = 'Kingdom',
          y = 'Number of taxa', fill = 'Measure') +
     ylim(0, 65) + # Leave room to add significance
     # scale_y_log10() +
     scale_fill_manual(values = c('#d7ce9f', '#849db1', '#b66353')) +
     theme_classic() +
     theme()
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
ggsave(
     here(
          'results',
          'QC',
          paste0(plotID,
                 '_Number of food taxa by measure, shared samples only.pdf'
                 )
          ),
     height = 3,
     width = 6)
```

Statistics on above plot: 

```{r}
# Repeated measures ANOVA (Viridiplantae)
# Get only plant detection events, make unique sample ID
viridiplantae <- 
     counts.long %>% 
     mutate(sample = paste(subj, date, sep = '_')) %>% 
     filter(kingdom == 'Viridiplantae') 

model <- lm(value ~ dataset + sample,
            data = viridiplantae)

analysis <- Anova(model, 
                  idata = viridiplantae,
                  idesign = ~sample)

print(analysis)
```

```{r}
# Post-hoc testing

# Unadjusted p values
pairwise.t.test(viridiplantae$value,
                viridiplantae$dataset,
                p.adjust.method = 'BH')
```

```{r}
# Repeated measures ANOVA (Metazoa)
metazoa <- 
     counts.long %>% 
     mutate(sample = paste(subj, date, sep = '_')) %>% 
     filter(kingdom == 'Metazoa') 

model <- lm(value ~ dataset + sample,
            data = metazoa)

analysis <- Anova(model, 
                  idata = metazoa,
                  idesign = ~sample)

print(analysis)
```

```{r}
# Unadjusted p values
pairwise.t.test(metazoa$value,
                metazoa$dataset,
                p.adjust.method = 'BH')
```


```{r}
# Paired t-test (Fungi)
fungi <- 
     filter(counts, kingdom == 'Fungi')

t.test(fungi$menu_taxa, fungi$protein_taxa,
       paired = TRUE)
```

Patterns of recorded foods by day of week: 

```{r echo = FALSE}
counts.long %>% 
     group_by(ID, date, dataset) %>% 
     summarize(value = sum(value, na.rm = TRUE)) %>% 
     filter(dataset == 'Recorded menu') %>% 
     ggplot(aes(x = date, y = value)) +
     geom_point(aes(color = dataset)) +
     facet_wrap(~ ID, nrow = 4, scales = 'free_x') + 
     labs(x = 'Date', y = 'Number of taxa') +
     theme_bw() +
     theme(axis.title.x = element_blank(),
           legend.title = element_blank(),
           legend.position = 'bottom')

plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
# ggsave(here('results', 
#             paste0(plotID, 
#                    '_Number of food taxa in menu by day.pdf')),
#        height = 7, width = 4)
```

