---
title: "Filter phyloseq objects"
output: html_notebook
---

This filters the phyloseq objects created in the "Combined phyloseq objects" notebook.

# Setup

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = file.path(getwd(), '..'))
```

```{r}
library(here)
library(phyloseq)
library(tidyverse)
```

# Read in data
## Metabarcoding
### trnL
```{r}
ps.mb.plant <- 
     here('data',
          'processed',
          'phyloseq',
          'combined',
          'ps_mb_trnL.rds') %>% 
     readRDS()

ps.mb.plant
```

### 12SV5
```{r}
ps.mb.animal <- 
     here('data',
          'processed',
          'phyloseq',
          'combined',
          'ps_mb_12SV5.rds') %>% 
     readRDS()

ps.mb.animal
```

## Metaproteomics

Two versions of this dataset: one with an entry for each protein accession, and another for each species.

However, can go from one to the other, right?  Can always group accession information by taxon and sum up PSMs. 

```{r}
ps.protein <- 
     here('data',
          'processed',
          'phyloseq',
          'combined',
          'ps_mp (all proteins).rds') %>%
     readRDS()

ps.protein
```

#### Detailed data

```{r}
# Protein detail
proteins.df <- 
     here('data',
          'processed',
          '5FDR_1UP_both datasets.rds') %>% 
     readRDS()

proteins.df
```

```{r}
# Animal-based protein mapping
proteins.animal <- 
     here('data',
          'processed',
          'Animal-based proteins (annotated).csv') %>% 
     read_csv()
```

## Menu

```{r}
ps.menu <- 
     here('data', 
          'processed', 
          'phyloseq',
          'combined',
          'ps_menu.rds') %>% 
     readRDS()

ps.menu
```

# Pre-process

## Metabarcoding

### Unidentified reads

#### trnL

```{r}
# What proportion of overall dataset?
MButils::percent_unassigned(ps.mb.plant)
```

```{r}
# Per sample?
trnL.no.id <- 
     MButils::percent_unassigned(ps.mb.plant, by_sample = TRUE) %>% 
     data.frame(pct_no_id = .) %>% 
     rownames_to_column(var = 'sample') %>% 
     separate(sample,
              into = c('subj', 'date'),
              sep = '_') %>% 
     mutate(study = ifelse(subj == '908',
                           yes = 'Habitual Diet',
                           no = 'Intervention'))

trnL.no.id
```
```{r}
median(trnL.no.id$pct_no_id, na.rm = TRUE)
```

```{r}
range(trnL.no.id$pct_no_id, na.rm = TRUE)
```

Small enough to safely drop: remove these from object.

```{r}
ps.mb.plant <- subset_taxa(ps.mb.plant, !is.na(phylum))
ps.mb.plant

rm(trnL.no.id)
```

#### 12SV5

```{r}
# What proportion of each sample?
MButils::percent_unassigned(ps.mb.animal)
```

Here, may want to instead ask what we can't assign to the family level.

```{r}
ps.mb.animal@tax_table@.Data %>% 
     data.frame() %>% 
     rownames_to_column(var = 'asv') %>% 
     select(family:species)
```
Nothing: everything has a genus or species assignment. 

### Filter to >=5 counts 

Impose >= 5 count/taxon filter. 

```{r}
# trnL
asvtab.mb.plant <- otu_table(ps.mb.plant)@.Data
sum(asvtab.mb.plant)

asvtab.mb.plant[asvtab.mb.plant < 5] <- 0
sum(asvtab.mb.plant)

# Replace in phyloseq object
otu_table(ps.mb.plant) <- otu_table(asvtab.mb.plant,
                                    taxa_are_rows = FALSE)
```

```{r}
# 12SV5
asvtab.mb.animal <- otu_table(ps.mb.animal)@.Data
sum(asvtab.mb.animal)

asvtab.mb.animal[asvtab.mb.animal < 5] <- 0
sum(asvtab.mb.animal)

# Replace in phyloseq object
otu_table(ps.mb.animal) <- otu_table(asvtab.mb.animal,
                                    taxa_are_rows = FALSE)
```

### Add lag day

```{r echo = FALSE}
# trnL
samdf.mb.plant <- 
     sample_data(ps.mb.plant) %>% 
     as('data.frame') %>% 
     rownames_to_column(var = 'row')

# Label each stool sample with its two preceding days
for (delta in 1:2){
     varname <- paste0('delta', delta)
     samdf.mb.plant <- mutate(samdf.mb.plant, 
                             !!varname := paste(subj, date - delta, 
                                                sep = '_'))
}

# Replace in phyloseq object
sample_data(ps.mb.plant) <- column_to_rownames(samdf.mb.plant, var = 'row')
```

```{r echo = FALSE}
# 12SV5
samdf.mb.animal <- 
     sample_data(ps.mb.animal) %>% 
     as('data.frame') %>% 
     rownames_to_column(var = 'row')

# Label each stool sample with its two preceding days
for (delta in 1:2){
     varname <- paste0('delta', delta)
     samdf.mb.animal <- mutate(samdf.mb.animal, 
                             !!varname := paste(subj, date - delta, 
                                                sep = '_'))
}

# Replace in phyloseq object
sample_data(ps.mb.animal) <- column_to_rownames(samdf.mb.animal, var = 'row')
```

### Rename

```{r}
taxa_names(ps.mb.plant) <- tax_table(ps.mb.plant)[, 'name']
taxa_names(ps.mb.animal) <- tax_table(ps.mb.animal)[, 'name']
```

## Metaproteomic phyloseq

### Add lag day

```{r echo = FALSE}
# Pull sample data for stool samples
samdf.protein <- 
     sample_data(ps.protein) %>% 
     as('data.frame') %>% 
     rownames_to_column(var = 'row')

# Label each stool sample with its two preceding days
for (delta in 1:2){
     varname <- paste0('delta', delta)
     samdf.protein <- mutate(samdf.protein, 
                             !!varname := paste(subj, date - delta, 
                                                sep = '_'))
}

# Replace in phyloseq object
sample_data(ps.protein) <- column_to_rownames(samdf.protein, var = 'row')
```

### Protein categories

I manually labeled entries with >=5 PSMs. Categories are:
- muscle
- tissue
- egg
- dairy
- none (== 'other')

```{r}
proteins.animal <-
     proteins.animal %>% 
     # Convert empty category to "Other"
     mutate(category = ifelse(category == '',
                              yes = 'other',
                              no = category))

str(proteins.animal)
```
I manually labeled these based on searches of protein names.  How does the grouping break out? 

```{r}
proteins.animal %>% 
     group_by(category, certainty) %>% 
     count()
```

```{r}
# Make "other" label "certain"
proteins.animal$certainty[proteins.animal$category == 'other'] <- 1
```

#### Automated labels (<-5 PSMs)

Scan certain terms from above to make regular expressions:

```{r}
manual.label <- filter(proteins.animal, weight >= 5)
automatic.label <- filter(proteins.animal, weight < 5)
```

```{r}
# Muscle
# Need the [^nt] to filter out hits to "interACTINg" and 'contACTIN'
ex.muscle <- '[^rt][Aa]ctin|[Cc]reatine|[Mm]yo|[Ff]erritin|[Nn]ebulin|[Ss]arco|[Tt]itin|[Tt]ropo'

# Tissue
# Think many of these (had previously had as '[Kk]eratin|[Cc]ollagen|[Ff]ormin') are not cell-type specific and could be reasonably expected to be in human epithelia

# Egg
ex.egg <- '[Vv]itell|^Ovo|[Aa]lbumin'

# Dairy
# Lacto gives other hits (e.g. beta-galactosidase, galactokinase), but nothing resembling lactoglobulin, so removed
ex.dairy <- '[Cc]asein|[Bb]utyrophilin'
```

```{r}
# Label these in data
automatic.label <- 
     automatic.label %>%
     mutate(category = 
                 case_when(grepl(word, pattern = ex.muscle) ~ 'muscle',
                           grepl(word, pattern = ex.dairy) ~ 'dairy',
                           grepl(word, pattern = ex.egg) ~ 'egg',
                           TRUE ~ 'other')) %>% 
     mutate(certainty = 1)
```

```{r}
# How many labels assigned (out of 1170 starting?)
sum(automatic.label$category != 'other')
```

~5%.  May want to go back and refine this, but consider it workable for now.

```{r}
# Join automated to manual assignments
proteins.animal <- 
     bind_rows(manual.label,
               automatic.label)
```

```{r}
# For now, consider only certain assignments:
proteins.animal <- 
     proteins.animal %>% 
     mutate(category = ifelse(certainty == 0,
                              yes = 'other',
                              no = category),
            category = ifelse(is.na(category),
                              yes = 'other',
                              no = category)) 

table(proteins.animal$category)
```

```{r}
sum(proteins.animal$category != 'other')/length(proteins.animal$category)
```

```{r}
proteins.animal %>% 
     group_by(category != 'other') %>% 
     summarize(psms = sum(weight)) %>% 
     mutate(pct = psms/sum(psms))
```
This is 10.6% of peptides, and 42.2% of PSMs.

#### Filter

PSM counts are named with accession-- need to connect this to protein name, and then subset. 

```{r}
head(taxa_names(ps.protein))
```

Join to taxonomy, because I want to preserve all plants while subset animals
```{r}
taxtab.protein <- data.frame(ps.protein@tax_table@.Data)
proteins.df <- 
     taxtab.protein %>% 
     rownames_to_column(var = 'Accession') %>% 
     left_join(proteins.df, .) 
```

```{r}
# Protein names to subset by
meat.dairy.egg <- 
     filter(proteins.animal, category != 'other') %>% 
     pull(word) 

head(meat.dairy.egg)
```

```{r}
cat('Starting entries:', nrow(proteins.df), '\n')

proteins.df <- 
     proteins.df %>% 
     mutate(protein = gsub(Description,
                           pattern = ' OS=.*$',
                           replacement = '')) %>% 
     filter(protein %in% meat.dairy.egg | kingdom != 'Metazoa')

cat('Filtered entries:', nrow(proteins.df))
```

Now, can subset phyloseq object based on accessions.
```{r}
ps.protein

ps.protein <- prune_taxa(proteins.df$Accession, ps.protein)
ps.protein
```

### Group by taxon

```{r}
# Relabel with lowest taxonomic level
tax_table(ps.protein) <- 
     ps.protein@tax_table@.Data %>% 
     data.frame() %>% 
     MButils::lowest_level() %>% 
     as.matrix()

ps.protein.taxon <- tax_glom(ps.protein,
                                    taxrank = 'name')

taxa_names(ps.protein.taxon) <- 
     ps.protein.taxon@tax_table@.Data[, 'name']

ps.protein.taxon
```

### Filter to >=5 counts

Note that I'm applying this filter to the taxon-level analysis only.

```{r}
# Impose >= 5 PSM/taxon filter
asvtab.protein.taxon <- otu_table(ps.protein.taxon)@.Data
sum(asvtab.protein.taxon)

asvtab.protein.taxon[asvtab.protein.taxon < 5] <- 0
sum(asvtab.protein.taxon)

# Replace in phyloseq object
otu_table(ps.protein.taxon) <- otu_table(asvtab.protein.taxon,
                                                taxa_are_rows = FALSE)
```

```{r}
# Does this remove any taxa? 
ntaxa(ps.protein.taxon)

ps.protein.taxon <- 
     ps.protein.taxon %>% 
     prune_taxa(taxa_sums(.) > 0, .) 

ntaxa(ps.protein.taxon)
```

### Add missing sample

ML48 (10/17/19) omitted due to low spectra counts and poor extraction.  Hold off on including for now; not sure if I need.

## Menu phyloseq

Here, what should I plot? Aggregated menu from 1-2 d before intake (to be consistent w/downstream results?)

Why don't I just start with one day before and see how it goes.

### [CONSIDER] Subset to lag -1-2?

```{r}
nsamples(ps.menu)
```

```{r}
# What taxa are removed in this filtering?
ps.menu %>% 
     # Use all possible samples here, rather than filtering for only successes
     # This occurs in either metabarcoding object (full 27), 
     # not yet filtered for success
     prune_samples(sample_names(.) %in% sample_data(ps.mb.animal)$delta1,
                   .) %>% 
     prune_taxa(taxa_sums(.) == 0, .) %>% 
     taxa_names() %>% 
     sort()
```

```{r}
# Hold off on filtering for now
# ps.menu <- 
#      ps.menu %>% 
#      # Use all possible samples here, rather than filtering for only successes
#      # This occurs in either metabarcoding object (full 27), not yet filtered for success
#      prune_samples(sample_names(.) %in% sample_data(ps.mb.animal)$delta1,
#                    .) %>% 
#      prune_taxa(taxa_sums(.) > 0, .)
# 
# ps.menu
```

# Save

## Metabarcoding

```{r}
saveRDS(ps.mb.plant,
        here('data',
             'processed',
             'phyloseq',
             'combined',
             'filtered',
             'ps_mb_trnL.rds'))

saveRDS(ps.mb.animal,
        here('data',
             'processed',
             'phyloseq',
             'combined',
             'filtered',
             'ps_mb_12SV5.rds'))
```

## Metaproteomics

```{r}
# Tax-glommed
saveRDS(ps.protein.taxon,
        here('data',
             'processed',
             'phyloseq',
             'combined',
             'filtered',
             'ps_mp_taxa.rds'))
```

```{r}
# By proteins
saveRDS(ps.protein,
        here('data',
             'processed',
             'phyloseq',
             'combined',
             'filtered',
             'ps_mp_proteins.rds'))
```

## Menu

```{r}
saveRDS(ps.menu,
        here('data',
             'processed',
             'phyloseq',
             'combined',
             'filtered',
             'ps_menu.rds'))
```

