---
title: "Kleiner lab database testing"
output:
  html_document:
    df_print: paged
---

Here we are evaluating the preliminary data shared by the Kleiner lab at NCSU to determine if their metaproteomic database should be adjusted (e.g. filtering out homologs to human proteins, other ideas?) before continuing with analysis of the DFC samples.

# Setup

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = file.path(getwd(), '..'))
```

```{r libraries, echo=FALSE}
library(here) 
library(phyloseq) 
library(RColorBrewer)
library(tidyverse) 
```

# Read in data

## Synchronized names

## Taxon name mapping

```{r}
name.sync <- 
     here('data',
          'processed',
          'Naming discrepancies.xlsx') %>% 
     readxl::read_excel()

name.sync
```

## Metabarcoding phyloseq

```{r}
ps.animal <- 
     here('data', 
          'processed', 
          'phyloseq',
          'combined',
          'filtered',
          'ps_mb_12SV5.rds') %>% 
     readRDS()
ps.animal

ps.plant <- 
     here('data', 
          'processed', 
          'phyloseq',
          'combined',
          'filtered',
          'ps_mb_trnL.rds') %>% 
     readRDS()

ps.plant
```

## Menu data phyloseq

```{r}
ps.menu <- 
     here('data', 
          'processed', 
          'phyloseq',
          'combined',
          'filtered',
          'ps_menu.rds') %>% 
     readRDS()

ps.menu
```

# Pre-process

## Metabarcoding

```{r}
# The 12SV5 data have a shifted taxonomy, I think I need to go back and figure 
# this out. Maybe due to assignTaxonomy()?
# Below, kingdom should be superkingdom, phylum should be kingdom.
# Think it's class that's throwing this off, for some should be Mammalia
ps.animal@tax_table@.Data

# Update accordingly
tax_table(ps.animal) <- 
     tax_table(ps.animal)@.Data %>% 
     data.frame() %>% 
     rownames_to_column(var = 'row') %>% 
     mutate(superkingdom = kingdom,
            kingdom = phylum,
            phylum = class) %>% 
     select(-class) %>% 
     select(superkingdom, kingdom, phylum, everything()) %>% 
     column_to_rownames(var = 'row') %>% 
     as.matrix()
```

```{r}
# Note failed trnL samples and relabel with NAs prior to merge
failed <- 
     which(sample_sums(ps.plant)==0) %>% 
     names()

asvtab.plant <- ps.plant@otu_table@.Data
asvtab.plant[failed, ] <- NA
otu_table(ps.plant) <- otu_table(asvtab.plant,
                                 taxa_are_rows = FALSE)

# Combine into a single object
ps.mb <- merge_phyloseq(ps.plant, ps.animal)
ps.mb

# Remove separated objects
rm(ps.plant, ps.animal)
```

## Menu

```{r}
# Group by taxon, omitting food group info
ps.menu <- 
     ps.menu %>% 
     tax_glom(taxrank = 'name')

# Update names, which previously included food group
taxa_names(ps.menu) <- tax_table(ps.menu)@.Data[, 'name']
```

## Pair with 1-2 days prior intake

For this broadest level comparison, let's look only at the two days prior to sampling.

```{r}
# These already exist in sample data:
sample_variables(ps.mb)
samdf.mb <- 
     sample_data(ps.mb) %>% 
     data.frame() %>% 
     rownames_to_column(var = 'row')
```

```{r}
# Can't use merge_samples on phyloseq here because some samples make up >1 merge

# Do manually on ASV table
asvtab.menu <- otu_table(ps.menu)@.Data
asvtab.menu.12 <- data.frame()

for (sample in seq(dim(samdf.mb)[1])){
     # Pull menu day -1
     one.before <- asvtab.menu[samdf.mb$delta1[sample], ]
     # Pull menu day -2
     two.before <- asvtab.menu[samdf.mb$delta2[sample], ]
     # Merge and place in new, aggregated OTU table
     asvtab.menu.12 <- rbind(asvtab.menu.12,
                             one.before + two.before)
     # Update food names (only has to be done once)
     if (sample == 1){names(asvtab.menu.12) <- names(one.before + two.before)}
     # Update sample name
     row.names(asvtab.menu.12)[sample] <- samdf.mb$row[sample]
}

# Now rebuild a subsetted phyloseq object
ps.menu.12 <- phyloseq(otu_table(asvtab.menu.12, taxa_are_rows = FALSE),
                       sample_data(ps.mb), # Now this matches
                       tax_table(ps.menu)) # Menu-specific taxonomy

# Remove any taxa that aren't present any longer
ps.menu.12 <- prune_taxa(taxa_sums(ps.menu.12) > 0, ps.menu.12)
ps.menu.12
```

There are `r ntaxa(ps.menu.12)` recorded in the menu in the 2 days prior to each sample, and `r ntaxa(ps.mb)` detected by metaproteomics.

## Reference taxa by their shared index

Will need to glom by the shared taxon name, because otherwise phyloseq object would have multiple instances of that taxon. 

So what I'll want to do here is join the taxon indices to the names to the taxonomy, and then glom there and rename in the object.

Here, use "mb_tomp_to_menu" column, which maps specifically between these two datasets.
```{r}
# Metabarcoding
tax_table(ps.mb) <- 
     name.sync %>% 
     select(mb_to_mp_to_menu, mb_taxa) %>% 
     right_join(data.frame(ps.mb@tax_table@.Data),
               by = c('mb_taxa' = 'name')) %>%
     distinct() %>% 
     column_to_rownames(var = 'mb_taxa') %>% 
     as.matrix()

ps.mb.id <- ps.mb
taxa_names(ps.mb.id) <- ps.mb.id@tax_table@.Data[, 'mb_to_mp_to_menu']
```
`
```{r}
# Synchronize menu to metabarcoding names
# Needs name column added
tax_table(ps.menu.12) <- 
     MButils::lowest_level(data.frame(ps.menu.12@tax_table@.Data)) %>% 
     as.matrix()

tax_table(ps.menu.12) <- 
     name.sync %>% 
     select(mb_to_mp_to_menu, menu_taxa) %>% 
     right_join(data.frame(ps.menu.12@tax_table@.Data), 
               by = c('menu_taxa' = 'name')) %>% 
     distinct() %>% 
     column_to_rownames(var = 'menu_taxa') %>% 
     as.matrix()
```

```{r}
# Get a list of these duplicates
dups <- 
     ps.menu.12@tax_table@.Data %>% 
     data.frame() %>% 
     filter(duplicated(mb_to_mp_to_menu)) %>% 
     pull(mb_to_mp_to_menu) %>% 
     unique()

dups
```

```{r}
ps.menu.12
merges = list()
for (taxon in dups){
     # Get the list of equivalent taxa at the species level for that food
     eq_taxa <- 
          ps.menu.12@tax_table@.Data %>% 
          data.frame() %>% 
          filter(mb_to_mp_to_menu == taxon) %>% 
          select(-mb_to_mp_to_menu) %>% 
          MButils::lowest_level() %>% 
          pull(name)
     
     merges <- append(merges, list(eq_taxa))
}

if (length(merges) == 1){
     # Just do one merge
      ps.menu.id <- merge_taxa(ps.menu.12, merges[[1]])
} else {
     # Merge once to make new phyloseq
     ps.menu.id <- merge_taxa(ps.menu.12, merges[[1]])
     # Then loop over remainder
     for (i in 2:length(merges)){
          ps.menu.id <- merge_taxa(ps.menu.id, merges[[i]])
     }
}

taxa_names(ps.menu.id) <- tax_table(ps.menu.id)[, 'mb_to_mp_to_menu']
ps.menu.id
```

# Analyze 

## Metaproteomics x menu

### All taxa

```{r}
# Extract OTU tables
asvtab.mb <- otu_table(ps.mb.id)@.Data
asvtab.menu.12 <- otu_table(ps.menu.id)@.Data

# Synchronize names
dim(asvtab.mb)
dim(asvtab.menu.12)
length(union(colnames(asvtab.menu.12), colnames(asvtab.mb)))
```

Want to make these tables the same same size (taking the union of their columns, and populating the new additions to each with zeros since the food wasn't detected in that dataset).

```{r}
# Helper function to pad out columns
pad_columns <- function(x, y, fill = 0){
     # Takes two matrices and synchronizes columns across them
     # Returns a list of the two updated matrices
     
     # Find missing columns in both matrices
     missing.x <- setdiff(colnames(y), colnames(x))
     missing.y <- setdiff(colnames(x), colnames(y))
     
     # Pad out columns of x
     fill.x <- matrix(fill, 
                      nrow = dim(x)[1],
                      ncol = length(missing.x))
     colnames(fill.x) <- missing.x 
     x <- cbind(x, fill.x)
     
     # Pad out columns of y
     fill.y <- matrix(fill, 
                      nrow = dim(y)[1],
                      ncol = length(missing.y))
     colnames(fill.y) <- missing.y
     y <- cbind(y, fill.y)
     
     # Arrange the columns so they appear in identical order
     u <- sort(union(colnames(x), colnames(y)))
     x <- x[, u]
     y <- y[, u]
     
     list(x, y)
}
```

```{r}
padded <- pad_columns(asvtab.mb, asvtab.menu.12)
asvtab.mb <- padded[[1]]
asvtab.menu.12 <- padded[[2]]

rm(padded)
```

Now, we've padded out these columns, but the ones that came from the menu and added to the metabarcoding data may contain plant info for the failed samples. 
```{r}
# Find these columns
different <- setdiff(taxa_names(ps.menu.id), taxa_names(ps.mb.id)) %>% sort()
different

# However, want to select only those that are plant-based (only failed samples
# from trnL, not 12SV5). Use taxonomy to judge:
menu.plants <- 
     ps.menu.id@tax_table@.Data %>% 
     data.frame() %>% 
     filter(kingdom == 'Viridiplantae') %>% 
     row.names()

different <- intersect(different, menu.plants)
```

```{r}
# Change the failed sample x padded plant column pairs to be NA in the 
# metabarcoding dataset
asvtab.mb[failed, different] <- NA
```

```{r}
# Confirm row and column names are equal before proceeding
all(row.names(asvtab.mb) == row.names(asvtab.menu.12))
all(colnames(asvtab.mb) == colnames(asvtab.menu.12))
```

```{r}
# Label predictions 
# If failed samples present (OTU table entries are NAs), need to define a 
# version of "&" that does not return FALSE for NA & FALSE 
and_na <- function(x, y){
     
     # Instead returns NA
     return <- NULL
     if (any(is.na(c(x,y)))){
          # If there's also FALSE, change the output
          if (FALSE %in% c(x, y)){
               return <- NA
          } else {
          # Otherwise keep as-is
               return <- x & y
          }
     } else {
          return <- x & y
     }
     
     return
}

dna.pos <- asvtab.mb > 0
menu.pos <- asvtab.menu.12 > 0

tp <- 
     mapply(and_na, dna.pos, menu.pos) %>% 
     matrix(nrow = nrow(dna.pos),
            dimnames = list(row.names(dna.pos),
                            colnames(dna.pos)))

tn <- 
     mapply(and_na, !dna.pos, !menu.pos) %>% 
     matrix(nrow = nrow(dna.pos),
            dimnames = list(row.names(dna.pos),
                            colnames(dna.pos)))
fp <- 
     mapply(and_na, dna.pos, !menu.pos) %>% 
     matrix(nrow = nrow(dna.pos),
            dimnames = list(row.names(dna.pos),
                            colnames(dna.pos)))

fn <- 
     mapply(and_na, !dna.pos, menu.pos) %>% 
     matrix(nrow = nrow(dna.pos),
            dimnames = list(row.names(dna.pos),
                            colnames(dna.pos)))
```

```{r}
# Group results by study and reorganize
reformat <- function(pred, label){
     # Relabel results with study (based on sample) and prediction type
     pred <- 
          data.frame(pred, check.names = FALSE) %>% 
          rownames_to_column(var = 'study') %>% 
          mutate(study = substr(study, 1, 2)) %>%  # 1st 2 chars: HD/ID
          group_by(study) %>% 
          summarize(across(.cols = everything(),
                           ~sum(.x, na.rm = TRUE))) %>% 
          mutate(prediction = label) %>% 
          select(study, prediction, everything())
     
     pred
}

tp <- reformat(tp, 'tp')
tn <- reformat(tn, 'tn')
fp <- reformat(fp, 'fp')
fn <- reformat(fn, 'fn')

pred <- bind_rows(tp, tn, fp, fn)
```

```{r}
# Reformat to long for visualization
pred.long <- pivot_longer(pred,
                          cols = -c(prediction, study),
                          names_to = 'common_name',
                          values_to = 'count')

# Factor data for visualization
# Calculate F-measure to be used downstream
# For this ranking, combine results from both cohorts
pred.summary <- 
     pred.long %>% 
     select(-study) %>% 
     group_by(prediction, common_name) %>% 
     summarize(count = sum(count)) %>% 
     pivot_wider(names_from = prediction, values_from = count) %>% 
     mutate(recall = tp/(tp + fn),
            precision = tp/(tp + fp),
            f_measure = (2 * recall * precision)/(recall + precision),
            pos_n = tp + fp) %>% 
     arrange(desc(f_measure), fn, fp)

pred.summary
```
```{r}
# Join to taxonomy to get kingdom-level info
# Make overall taxonomy matrix (combined menu and metaproteomics)
taxtab <- 
     bind_rows(
          data.frame(ps.menu.id@tax_table),
          data.frame(ps.mb.id@tax_table)
          ) %>% 
     select(kingdom, common_name = mb_to_mp_to_menu) %>% 
     mutate(kingdom = ifelse(kingdom %in% c('Metazoa',
                                            'Viridiplantae'),
                             yes = kingdom,
                             no = 'Other')) %>% 
     distinct()

pred.summary <- left_join(pred.summary, taxtab)
```

```{r}
# Incorporate into object
pred.long <- 
     pred.summary %>% 
     select(common_name, kingdom) %>% 
     right_join(pred.long) %>% 
     select(kingdom, common_name, everything())
```

```{r}
# Now that joins complete, add factor levels
pred.long <- 
     pred.long %>% 
     mutate(
          study = factor(study,
                         levels = c('ID', 'HD'),
                         labels = c('Intervention', 'Habitual')),
          common_name = factor(common_name, 
                               levels = pred.summary$common_name),
          prediction = factor(prediction, levels = c('tp',
                                                     'tn',
                                                     'fp',
                                                     'fn'),
                              labels = c('True positive',
                                         'True negative',
                                         'False positive',
                                         'False negative')),
          kingdom = factor(kingdom,
                           levels = c('Viridiplantae',
                                      'Metazoa',
                                      'Other')))
```

```{r}
# Save for comparison to metabarcoding data
# saveRDS(pred.long,
#         here('data', 
#              'processed', 
#              'performance-metrics', 
#              'Metabarcoding v. menu predictions by shared food taxa, 1-2d prior, no intake filter.rds'))
```

#### Plot

```{r}
pred.long %>%
     ggplot(aes(x = common_name, y = count, fill = prediction)) +
     geom_bar(position = 'fill', stat = 'identity') +
     scale_fill_manual(values = c('#309343','#8AC373', # TP, TN
                                  '#D7191C', '#FDAE61')) + # FP, FN
     facet_grid(cols = vars(kingdom),
                # rows = vars(study),
                scales = 'free',
                space = 'free') +
     theme_classic() +
     theme(axis.text.x = element_text(angle = 40, hjust = 1),
           axis.text.y = element_blank(), 
           axis.ticks = element_blank(),
           axis.line = element_blank(),
           axis.title = element_blank(),
           legend.position = 'none',
           legend.title = element_blank(), 
           plot.margin = margin(0, 0.5, 0.5, 0.5, 'in'),
           strip.background = element_blank()) 
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

# ggsave(here('results',
#             'manuscript',
#             '3',
#             paste0(plotID,
#                    '_Metabarcoding by-food predictions compared to menu.pdf')),
#        height = 4, width = 24)
```

### Grouped: to species, family, genera

```{r}
rank_names(ps.menu.12)
rank_names(ps.mb)
```

```{r}
# Glom taxa at these levels (removing common-name labels first)
# Menu
tax_table(ps.menu.12) <- 
     ps.menu.12@tax_table@.Data %>% 
     data.frame() %>% 
     select(-mb_to_mp_to_menu) %>% 
     as.matrix()

menus <- 
     lapply(list('species', 'genus', 'family'),
            tax_glom,
            physeq = ps.menu.12)

# Update names
taxa_names(menus[[1]]) <- tax_table(menus[[1]])@.Data[, 'species']
taxa_names(menus[[2]]) <- tax_table(menus[[2]])@.Data[, 'genus']
taxa_names(menus[[3]]) <- tax_table(menus[[3]])@.Data[, 'family']
```

```{r}
# Metabarcoding
tax_table(ps.mb) <- 
     ps.mb@tax_table@.Data %>% 
     data.frame() %>% 
     select(-mb_to_mp_to_menu) %>% 
     as.matrix()

# Note that tax_glom doesn't like NAs, so revert to zeros for now:
otu_table(ps.mb)[is.na(otu_table(ps.mb))] <- 0

asvs <- 
     lapply(list('species', 'genus', 'family'),
            tax_glom,
            physeq = ps.mb) 

# Update names
taxa_names(asvs[[1]]) <- tax_table(asvs[[1]])@.Data[, 'species']
taxa_names(asvs[[2]]) <- tax_table(asvs[[2]])@.Data[, 'genus']
taxa_names(asvs[[3]]) <- tax_table(asvs[[3]])@.Data[, 'family']
```

```{r}
# Replace full phyloseq object with its OTU table
menus <- 
     lapply(menus, otu_table) %>% # otu_table-class object
     lapply(., slot, name = '.Data') # matrix

asvs <- 
     lapply(asvs, otu_table) %>% # otu_table-class object
     lapply(., slot, name = '.Data') # matrix
```

```{r}
# Check their dimensions
lapply(menus, dim)
lapply(asvs, dim)
```

Want to make these tables the same same size (taking the union of their columns, and populating the new additions to each with zeros since the food wasn't detected in that dataset). Use `pad_columns()` helper function defined above, but first, list plant columns, which will need to be populated with NAs for the failed samples.

```{r}
# Here, will actually want to just identify *all* plant columns, since I had
# to replace NAs with zeros above.
plants <- 
     mapply(union, 
            lapply(menus, colnames), 
            lapply(asvs, colnames))

# However, want to select only those that are plant-based (only failed samples
# from trnL, not 12SV5). Use taxonomy to judge:
taxtab.plants <- 
     merge_phyloseq(ps.menu.12, ps.mb) %>% 
     tax_table() %>% 
     slot(name = '.Data') %>% 
     data.frame() %>% 
     select(superkingdom,
            kingdom,
            phylum,
            order,
            family,
            genus,
            species,
            subspecies,
            varietas, 
            forma) %>% 
     filter(kingdom == 'Viridiplantae') %>% 
     as.matrix()

# Get indices of plant entries, then subset
plants.i <- lapply(plants, function(x){x %in% taxtab.plants})
for (i in 1:3){
     plants[[i]] <- plants[[i]][plants.i[[i]]]
}
```

```{r}
padded <- 
     mapply(
          pad_columns,
          menus,
          asvs
     )

menus <- list(padded[[1]], padded[[3]], padded[[5]])
asvs <- list(padded[[2]], padded[[4]], padded[[6]])
```

```{r}
# Label predictions 
# Recode detections or records as binary
menus.pos <- lapply(menus, function(x){x > 0})
asvs.pos <- lapply(asvs, function(x){x > 0})

# When counting TP, TN, FP, FN, need to remember that FALSE & NA returns FALSE,
# but I want it to return NA (the case for missing samples). Simply relabel
# everything that is [failed, plant] in each afterwards:

tp <- mapply(function(record, detect){record & detect},
             menus.pos, 
             asvs.pos)

tn <- mapply(function(record, detect){!record & !detect},
             menus.pos, 
             asvs.pos)

fp <- mapply(function(record, detect){!record & detect},
             menus.pos, 
             asvs.pos)

fn <- mapply(function(record, detect){record & !detect},
             menus.pos, 
             asvs.pos)

# Change the failed sample x padded plant column pairs to be NA in the 
# metabarcoding dataset
for (i in 1:3){
     tp[[i]][failed, plants[[i]]] <- NA
     tn[[i]][failed, plants[[i]]] <- NA
     fp[[i]][failed, plants[[i]]] <- NA
     fn[[i]][failed, plants[[i]]] <- NA
}
```

```{r}
# Group results by study and reorganize
reformat <- function(pred, label){
     # Relabel results with study (based on sample) and prediction type
     pred <- 
          data.frame(pred, check.names = FALSE) %>% 
          rownames_to_column(var = 'study') %>% 
          mutate(study = substr(study, 1, 2)) %>%  # 1st 2 chars: HD/ID
          group_by(study) %>% 
          summarize(across(.cols = everything(),
                           ~sum(.x, na.rm = TRUE))) %>% 
          mutate(prediction = label) %>% 
          select(study, prediction, everything())
     
     pred
}

tp <- lapply(tp, reformat, label = 'tp')
tn <- lapply(tn, reformat, label = 'tn')
fp <- lapply(fp, reformat, label = 'fp')
fn <- lapply(fn, reformat, label = 'fn')

pred <- 
     mapply(bind_rows, 
            tp, 
            tn, 
            fp,
            fn)
```

```{r}
# Reformat to long for visualization
pred.long <- 
     lapply(pred,
            pivot_longer,
            cols = -c(prediction, study),
            names_to = 'name',
            values_to = 'count')
```

```{r}
# Factor data for visualization
# Calculate F-measure to be used downstream
# For this ranking, combine results from both cohorts
summary_helper <- function(p){
     # p: pred.long: Long-form summary of per-taxon predictions
     p %>% 
     select(-study) %>% 
     group_by(prediction, name) %>% 
     summarize(count = sum(count)) %>% 
     pivot_wider(names_from = prediction, values_from = count) %>% 
     mutate(recall = tp/(tp + fn),
            precision = tp/(tp + fp),
            f_measure = (2 * recall * precision)/(recall + precision),
            pos_n = tp + fp) %>% 
     arrange(desc(f_measure), fn, fp)
     }

pred.summary <- lapply(pred.long, summary_helper)
```

```{r}
# Join to taxonomy to get kingdom-level info
# Make overall taxonomy matrix (combined menu and metabarcoding)
taxtabs <- 
     bind_rows(
          data.frame(ps.menu.12@tax_table),
          data.frame(ps.mb@tax_table)
          ) %>% 
     # Pair kingdom with either species, genus, or family label
     # for downstream join
     lapply(list('species', 'genus', 'family'),
            select, 
            .data = .,
            kingdom) %>% 
     lapply(.,
            distinct) %>% 
     # "Other": fungi and bacteria
     lapply(.,
            mutate,
            kingdom = ifelse(!(kingdom %in% c('Viridiplantae',
                                              'Metazoa')),
                             yes = 'Other',
                             no = kingdom)) 
```

```{r}
# Incorporate kingdom into objects
# Rename name to the specific taxonomic level relevant to that dataframe
pred.long[[1]] <- rename(pred.long[[1]], species = name)
pred.long[[2]] <- rename(pred.long[[2]], genus = name)
pred.long[[3]] <- rename(pred.long[[3]], family = name)

pred.long <- 
     mapply(left_join,
            pred.long,
            taxtabs) %>% 
     # Matrix output, reorganize into 3 dataframes
     apply(2, data.frame) %>% 
     # Rename taxon column, which was reset to "species" by prior step
     lapply(.,
            rename,
            name = species) %>% 
     # Remove "Other" kingdom -- not detected by metabarcoding
     lapply(.,
            filter,
            kingdom != 'Other')
```

```{r}
# Now that joins complete, add factor levels
pred.long <- 
     lapply(pred.long, 
            mutate,
            study = factor(study,
                         levels = c('ID', 'HD'),
                         labels = c('Intervention', 'Habitual')),
            prediction = factor(prediction, levels = c('tp',
                                                       'tn',
                                                       'fp',
                                                       'fn'),
                                labels = c('True positive',
                                           'True negative',
                                           'False positive',
                                           'False negative')),
            kingdom = factor(kingdom,
                             levels = c('Viridiplantae',
                                        'Metazoa'))
     ) 
```
      
```{r}
# Refactor taxa to rank them by descending F measure (info from pred.summary)
for (i in 1:3){
     pred.long[[i]]$name <- factor(pred.long[[i]]$name,
                                   levels = pred.summary[[i]]$name)
}

names(pred.long) <- c('species', 'genera', 'families')
```

```{r}
# Save results for comparison to metabarcoding data
ranks <- c('species', 'genus', 'family')
for (i in 1:3){
     saveRDS(pred.long[[i]],
             here('data', 
                  'processed', 
                  'performance-metrics', 
                  paste0('Metabarcoding v. menu predictions by shared food ',
                         ranks[[i]],
                         ', 1-2d prior, no intake filter.rds')
                  )
             )
}
```

Okay, so now I should be able to iterate over these objects to plot them.

#### Plot

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")
for (i in 1:3){
 ggplot(pred.long[[i]],
        aes(x = name, y = count, fill = prediction)) +
          geom_bar(position = 'fill', 
                   stat = 'identity',
                   width = 0.9) +
          scale_fill_manual(values = c('#309343','#8AC373', # TP, TN
                                       '#D7191C', '#FDAE61')) + # FP, FN
          facet_grid(cols = vars(kingdom),
                     scales = 'free',
                     space = 'free') +
          theme_classic() +
          theme(axis.text.x = element_text(angle = 40, hjust = 1),
                axis.text.y = element_blank(), 
                axis.ticks = element_blank(),
                axis.line = element_blank(),
                axis.title = element_blank(),
                legend.position = 'none',
                legend.title = element_blank(), 
                plot.margin = margin(0, 0.5, 0.5, 0.5, 'in'),
                strip.background = element_blank(),
                strip.text.x = element_blank())    
     
     ggsave(here('results',
                 'manuscript',
                 '3',
                 paste0(plotID,
                        '_Metabarcoding ',
                        names(pred.long)[i],
                        ' predictions compared to menu.pdf')),
            height = 4, width = 14)
}
```
