---
title: "Kleiner lab database testing"
output:
  html_document:
    df_print: paged
---

Here we are evaluating the preliminary data shared by the Kleiner lab at NCSU to determine if their metaproteomic database should be adjusted (e.g. filtering out homologs to human proteins, other ideas?) before continuing with analysis of the DFC samples.

# Setup

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = file.path(getwd(), '..'))
```

```{r libraries, echo=FALSE}
library(here) 
library(phyloseq) 
library(RColorBrewer)
library(tidyverse) 
```

# Read in data

## Metaproteomic phyloseq

```{r}
ps.protein <- 
     here('data', 
          'processed', 
          'phyloseq',
          'combined',
          'filtered',
          # 'ps_mp_1PUP_taxa.rds') %>% # >=1 protein unique peptide
          'ps_mp_1UP_taxa.rds') %>% # >=1 unique peptide
     readRDS()
          
ps.protein
```

## Menu data phyloseq

```{r}
ps.menu <- 
     here('data', 
          'processed', 
          'phyloseq',
          'combined',
          'filtered',
          'ps_menu.rds') %>% 
     readRDS()

ps.menu
```
# Pre-process

## Menu

```{r}
# Group by taxon, omitting food group info
ps.menu <- 
     ps.menu %>% 
     tax_glom(taxrank = 'name')

# Update names, which previously included food group
taxa_names(ps.menu) <- tax_table(ps.menu)@.Data[, 'name']
```

## Pair with 1-2 days prior intake

For this broadest level comparison, let's look only at the two days prior to sampling.

```{r}
# These already exist in sample data:
sample_variables(ps.protein)
samdf.protein <- 
     sample_data(ps.protein) %>% 
     data.frame() %>% 
     rownames_to_column(var = 'row')
```

```{r}
# Can't use merge_samples on phyloseq here because some samples make up >1 merge

# Do manually on ASV table
asvtab.menu <- otu_table(ps.menu)@.Data
asvtab.menu.12 <- data.frame()

for (sample in seq(dim(samdf.protein)[1])){
     # Pull menu day -1
     one.before <- asvtab.menu[samdf.protein$delta1[sample], ]
     # Pull menu day -2
     two.before <- asvtab.menu[samdf.protein$delta2[sample], ]
     # Merge and place in new, aggregated OTU table
     asvtab.menu.12 <- rbind(asvtab.menu.12,
                             one.before + two.before)
     # Update food names (only has to be done once)
     if (sample == 1){names(asvtab.menu.12) <- names(one.before + two.before)}
     # Update sample name
     row.names(asvtab.menu.12)[sample] <- samdf.protein$row[sample]
}

# Now rebuild a subsetted phyloseq object
ps.menu.12 <- phyloseq(otu_table(asvtab.menu.12, taxa_are_rows = FALSE),
                       sample_data(ps.protein), # Now this matches
                       tax_table(ps.menu)) # Menu-specific taxonomy

# Remove any taxa that aren't present any longer
ps.menu.12 <- prune_taxa(taxa_sums(ps.menu.12) > 0, ps.menu.12)
ps.menu.12
```

There are `r ntaxa(ps.menu.12)` recorded in the menu in the 2 days prior to each sample, and `r ntaxa(ps.protein)` detected by metaproteomics.

# Analyze 

## Metaproteomics x menu

### All taxa

This is now handled in the `Figure 3` notebook.

### Grouped: species, genus, family

#### Define helper functions

```{r}
# Helper function to pad out columns of two ASV tables that contain different taxa as members
pad_columns <- function(x, y, fill = 0){
     # Takes two matrices and synchronizes columns across them, filling added 
     # cols with a set value
     # Returns a list of the two updated matrices
     
     # Find missing columns in both matrices
     missing.x <- setdiff(colnames(y), colnames(x))
     missing.y <- setdiff(colnames(x), colnames(y))
     
     # Pad out columns of x
     fill.x <- matrix(fill, 
                      nrow = dim(x)[1],
                      ncol = length(missing.x))
     colnames(fill.x) <- missing.x 
     x <- cbind(x, fill.x)
     
     # Pad out columns of y
     fill.y <- matrix(fill, 
                      nrow = dim(y)[1],
                      ncol = length(missing.y))
     colnames(fill.y) <- missing.y
     y <- cbind(y, fill.y)
     
     # Arrange the columns so they appear in identical order
     u <- sort(union(colnames(x), colnames(y)))
     x <- x[, u]
     y <- y[, u]
     
     list(x, y)
}
```

```{r}
# Helper function to group results by study and reorganize
reformat <- function(pred, label){
     # Relabel results with prediction type
     pred <- 
          data.frame(pred, check.names = FALSE) %>% 
          summarize(across(.cols = everything(),
                           ~sum(.x))) %>% 
          mutate(prediction = label) %>% 
          select(prediction, everything())
     
     pred
}
```

Calculate performance vectorized over a list of objects, which contain the menu/protein information grouped to either the species, genus, or family level.

```{r}
rank_names(ps.menu.12)
rank_names(ps.protein)
```

```{r}
# Glom taxa at these levels 
menus <- 
     lapply(list('species', 'genus', 'family'),
            tax_glom,
            physeq = ps.menu.12) 

# Rename taxa to the glommed level
taxa_names(menus[[1]]) <- tax_table(menus[[1]])@.Data[, 'species']
taxa_names(menus[[2]]) <- tax_table(menus[[2]])@.Data[, 'genus']
taxa_names(menus[[3]]) <- tax_table(menus[[3]])@.Data[, 'family']
```

```{r}
# Metaproteomic
proteins <- 
     lapply(list('species', 'genus', 'family'),
            tax_glom,
            physeq = ps.protein)

taxa_names(proteins[[1]]) <- tax_table(proteins[[1]])@.Data[, 'species']
taxa_names(proteins[[2]]) <- tax_table(proteins[[2]])@.Data[, 'genus']
taxa_names(proteins[[3]]) <- tax_table(proteins[[3]])@.Data[, 'family']
```

```{r}
# Replace full phyloseq object with its OTU table
menus <- 
     lapply(menus, otu_table) %>% # otu_table-class object
     lapply(., slot, name = '.Data') # matrix

proteins <- 
     lapply(proteins, otu_table) %>% # otu_table-class object
     lapply(., slot, name = '.Data') # matrix
```

```{r}
# Check their dimensions
lapply(menus, dim)
lapply(proteins, dim)
```

Want to make these tables the same same size (taking the union of their columns, and populating the new additions to each with zeros since the food wasn't detected in that dataset). Use `pad_columns()` helper function defined above. 

```{r}
padded <- 
     mapply(
          pad_columns,
          menus,
          proteins
     )

menus <- list(padded[[1]], padded[[3]], padded[[5]])
proteins <- list(padded[[2]], padded[[4]], padded[[6]])
```

```{r}
# Confirm row and column names are equal before proceeding
for (i in 1:3){
     print(all(row.names(menus[[i]]) == row.names(proteins[[i]])))
     print(all(colnames(menus[[i]]) == colnames(proteins[[i]])))
}
```

```{r}
# Label predictions 
# Recode detections or records as binary
menus.pos <- lapply(menus, function(x){x > 0})
proteins.pos <- lapply(proteins, function(x){x > 0})

tp <- mapply(function(record, detect){record & detect},
             menus.pos, 
             proteins.pos)

tn <- mapply(function(record, detect){!record & !detect},
             menus.pos, 
             proteins.pos)

fp <- mapply(function(record, detect){!record & detect},
             menus.pos, 
             proteins.pos)

fn <- mapply(function(record, detect){record & !detect},
             menus.pos, 
             proteins.pos)
```

```{r}
# Group results by study and reorganize
tp <- lapply(tp, reformat, label = 'tp')
tn <- lapply(tn, reformat, label = 'tn')
fp <- lapply(fp, reformat, label = 'fp')
fn <- lapply(fn, reformat, label = 'fn')

pred <- 
     mapply(bind_rows, 
            tp, 
            tn, 
            fp,
            fn)
```

```{r}
# Reformat to long for visualization
pred.long <- 
     lapply(pred,
            pivot_longer,
            cols = -prediction,
            names_to = 'name',
            values_to = 'count')
```

```{r}
# Factor data for visualization
# Calculate F-measure to be used downstream
# For this ranking, combine results from both cohorts
summary_helper <- function(p){
     # p: pred.long: Long-form summary of per-taxon predictions
     p %>% 
     group_by(prediction, name) %>% 
     summarize(count = sum(count)) %>% 
     pivot_wider(names_from = prediction, values_from = count) %>% 
     mutate(recall = tp/(tp + fn),
            precision = tp/(tp + fp),
            f_measure = (2 * recall * precision)/(recall + precision),
            pos_n = tp + fp) %>% 
     arrange(desc(f_measure), desc(tn), desc(fp))
     }

pred.summary <- lapply(pred.long, summary_helper)
```

```{r}
# Join to taxonomy to get kingdom-level info
# Make overall taxonomy matrix (combined menu and metaproteomics)
taxtabs <- 
     bind_rows(
          data.frame(ps.menu.12@tax_table),
          data.frame(ps.protein@tax_table)
          ) %>% 
     # Pair kingdom with either species, genus, or family label
     # for downstream join
     lapply(list('species', 'genus', 'family'),
            select, 
            .data = .,
            kingdom) %>% 
     lapply(.,
            distinct) %>% 
     # "Other": fungi and bacteria
     lapply(.,
            mutate,
            kingdom = case_when(kingdom == 'Viridiplantae' ~ 'Plants',
                                kingdom == 'Metazoa' ~ 'Animals',
                                kingdom == 'Fungi' ~ 'Other',
                                is.na(kingdom) ~ 'Other'))
```

```{r}
# Incorporate kingdom into objects
# Rename name to the specific taxonomic level relevant to that dataframe
pred.long[[1]] <- rename(pred.long[[1]], species = name)
pred.long[[2]] <- rename(pred.long[[2]], genus = name)
pred.long[[3]] <- rename(pred.long[[3]], family = name)

pred.long <- 
     mapply(left_join,
            pred.long,
            taxtabs) %>% 
     # Matrix output, reorganize into 3 dataframes
     apply(2, data.frame) %>% 
     # Rename taxon column, which was reset to "species" by prior step
     lapply(.,
            rename,
            name = species)
```

```{r}
# Now that joins complete, add factor levels
pred.long <- 
     lapply(pred.long, 
            mutate,
            prediction = factor(prediction, levels = c('tp',
                                                       'tn',
                                                       'fp',
                                                       'fn'),
                                labels = c('True positive',
                                           'True negative',
                                           'False positive',
                                           'False negative')),
            kingdom = factor(kingdom,
                             levels = c('Plants',
                                        'Animals',
                                        'Other'))
     )
```
      
```{r}
# Refactor taxa to rank them by descending F measure (info from pred.summary)
for (i in 1:3){
     pred.long[[i]]$name <- factor(pred.long[[i]]$name,
                                   levels = pred.summary[[i]]$name)
}

names(pred.long) <- c('species', 'genera', 'families')
```

```{r}
# Save results for comparison to metabarcoding data
ranks <- c('species', 'genus', 'family')
# for (i in 1:3){
#      saveRDS(pred.long[[i]],
#              here('data',
#                   'processed',
#                   'performance-metrics',
#                   paste0('NEW_Metaproteomic 1UP v. menu predictions by shared food ',
#                          ranks[[i]],
#                          ', 1-2d prior, >4 counts per taxon filter, no intake filter.rds')
#                   )
#              )
# }
```

Now these objects can be combined and plotted for a side-by-side comparison with DNA-based data.

Okay, so now I should be able to iterate over these objects to plot them.
